{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running tblastx in both directions\n",
      "^C\n"
     ]
    }
   ],
   "source": [
    "file1 = 'planarian_transcriptome.fasta'\n",
    "type1 = 'nucl' #or 'prot' if file1 is a proteome\n",
    "id1 = 'pl' #2-character ID (e.g. 'hu' for human)\n",
    "\n",
    "file2 = 'schisto_transcriptome.fasta'\n",
    "\n",
    "type2 = 'nucl'\n",
    "id2 = 'sc' #2-character ID (e.g. 'mo' for mouse)\n",
    "!bash map_genes.sh --tr1 {file1} --t1 {type1} --n1 {id1} --tr2 {file2} --t2 {type2} --n2 {id2}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run SAMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from samap.mapping import SAMAP\n",
    "from samap.analysis import get_mapping_scores, GenePairFinder\n",
    "from samalg import SAM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SAMap accepts file paths to unprocessed, raw `.h5ad` files. Alternatively, if you already have a processed `SAM` object, you can load them in directly.\n",
    "\n",
    "\n",
    "Prior to running SAMap, you should have run `map_genes.sh`, which expects 2-character identifiers describing each species. For example, planarians and schistosomes might get `pl` and `sc` identifiers, respectively. `map_genes.sh` generates a `maps/` directory with the transcriptome mapping BLAST results deposited. The input species IDs and path to the `maps/` directory should be input into SAMap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "id1 = 'pl'\n",
    "id2 = 'sc'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing data 1 for SAMap.\n",
      "Preparing data 2 for SAMap.\n",
      "11630 `pl` genes and 7427 `sc` gene symbols match between the datasets and the BLAST graph.\n",
      "Stitching SAM 0 and SAM 1\n",
      "Found 88729 gene pairs\n",
      "Recomputing PC projections with gene pair subsets...\n",
      "Running hsnwlib\n",
      "Using leiden_clusters and leiden_clusters cluster labels.\n",
      "Out-neighbor smart expansion 1\n",
      "Out-neighbor smart expansion 2\n",
      "Indegree coarsening\n",
      "0/3 (0, 7657) True\n",
      "1/3 (20000, 7657) True\n",
      "2/3 (40000, 7657) True\n",
      "Concatenating SAM objects...\n",
      "ITERATION: 0 \n",
      "Average alignment score (A.S.):  0.4494758679567957 \n",
      "Max A.S. improvement: 0.5895449249249848 \n",
      "Min A.S. improvement: 0.0\n",
      "Calculating gene-gene correlations in the homology graph...\n",
      "Stitching SAM 0 and SAM 1\n",
      "Found 36515 gene pairs\n",
      "Recomputing PC projections with gene pair subsets...\n",
      "Running hsnwlib\n",
      "Using leiden_clusters and leiden_clusters cluster labels.\n",
      "Out-neighbor smart expansion 1\n",
      "Out-neighbor smart expansion 2\n",
      "Indegree coarsening\n",
      "0/3 (0, 7657) True\n",
      "1/3 (20000, 7657) True\n",
      "2/3 (40000, 7657) True\n",
      "Concatenating SAM objects...\n",
      "ITERATION: 1 \n",
      "Average alignment score (A.S.):  0.49477829364303205 \n",
      "Max A.S. improvement: 0.19229825114120402 \n",
      "Min A.S. improvement: -0.10815229774162562\n",
      "Calculating gene-gene correlations in the homology graph...\n",
      "Stitching SAM 0 and SAM 1\n",
      "Found 36892 gene pairs\n",
      "Recomputing PC projections with gene pair subsets...\n",
      "Running hsnwlib\n",
      "Using leiden_clusters and leiden_clusters cluster labels.\n",
      "Out-neighbor smart expansion 1\n",
      "Out-neighbor smart expansion 2\n",
      "Indegree coarsening\n",
      "0/3 (0, 7657) True\n",
      "1/3 (20000, 7657) True\n",
      "2/3 (40000, 7657) True\n",
      "Concatenating SAM objects...\n",
      "ITERATION: 2 \n",
      "Average alignment score (A.S.):  0.49727650758122877 \n",
      "Max A.S. improvement: 0.08542242917621143 \n",
      "Min A.S. improvement: -0.06833507920164617\n",
      "Alignment score --- 0.49727650758122877\n",
      "Running UMAP on the stitched manifolds.\n",
      "Elapsed time: 6.887594564755758 minutes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"\\n# passing in already-processed SAM objects\\nsam1=SAM()\\nsam2=SAM()\\nsam1.load_data('/path/to/file1_pr.h5ad')\\nsam2.load_data('/path/to/file2_pr.h5ad')\\n\\nsm = SAMAP(sam1,sam2,id1,id2,f_maps = 'maps/')\\nsamap = sm.run()\\n\""
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# passing in file names (SAMap will process the data with SAM and save the resulting objects to two `.h5ad` files.)\n",
    "fn1 = '/media/storage/Dropbox/scrna/notebooks/data/planarian.h5ad' #processed data will be automatically saved to `/path/to/file/file1_pr.h5ad`\n",
    "fn2 = '/media/storage/Dropbox/scrna/notebooks/data/schisto.h5ad' #processed data will be automatically saved to `/path/to/file/file2_pr.h5ad`\n",
    "# runs SAMAP (f_maps should be the path to the 'maps' directory generated in the BLAST step above)\n",
    "sm = SAMAP(fn1,fn2,id1,id2,f_maps = '/media/storage/Dropbox/scrna/notebooks/maps/')\n",
    "samap = sm.run()\n",
    "\n",
    "\"\"\"\n",
    "# passing in already-processed SAM objects\n",
    "sam1=SAM()\n",
    "sam2=SAM()\n",
    "sam1.load_data('/path/to/file1_pr.h5ad')\n",
    "sam2.load_data('/path/to/file2_pr.h5ad')\n",
    "\n",
    "sm = SAMAP(sam1,sam2,id1,id2,f_maps = 'maps/')\n",
    "samap = sm.run()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To calculate alignment scores between cell types, we can use `get_mapping_scores`. This function will use the combined SAM object produced by SAMap to calculate alignment scores between cell types in the provided cell type annotation columns of `sam.adata.obs`. If no cell type annotations exist, the leiden clusters generated by SAM can be used (`k1=k2='leiden_clusters'`).\n",
    "\n",
    "The resulting tables show the highest-scoring alignment scores for each cell type in organism 1 (`D1`) and organism 2 (`D2`), respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "k1 = 'cluster' #cell types annotation key in `sm.sam1.adata.obs`\n",
    "k2 = 'tissue' #cell types annotation key in `sm.sam2.adata.obs`\n",
    "D1,D2,MappingTable = get_mapping_scores(sm,k1,k2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neoblast: 0</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 10</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Muscle: 14</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Epidermal: 2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Intestine: 30</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Protonephridia: 29</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Muscle: 13</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Epidermal: 11</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 8</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Epidermal: 3</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 4</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neoblast: 5</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 20</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Epidermal: 35</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Pharynx: 37</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Muscle: 7</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 28</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 23</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Intestine: 6</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Epidermal: 24</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 18</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neoblast: 22</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Parapharyngeal: 38</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Muscle: 16</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Pharynx: 25</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 9</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Parapharyngeal: 12</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Protonephridia: 26</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 21</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 15</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 17</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 33</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Protonephridia: 40</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Intestine: 19</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 32</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 36</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Neural: 34</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 39</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Pharynx: 27</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 42</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 31</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Intestine: 43</th>\n",
       "      <th colspan=\"2\" halign=\"left\">pl_Cathepsin+ cells: 41</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.598764</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.58303</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.551705</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.529709</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.423742</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.384437</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.377412</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.36336</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.343862</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.329127</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.284357</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.271645</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.260733</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.258121</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.250161</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.242052</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.223917</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.208507</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.185195</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.173964</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.163489</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.154325</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.153925</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.152423</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.146065</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.139406</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.122546</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.122045</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.12063</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.115261</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.075856</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0687788</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.066575</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.0475718</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0449842</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.0432449</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0417364</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.0311426</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0252688</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.0234031</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00965553</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00865499</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00612058</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00399417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000372316</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00652464</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.0221045</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.0110845</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0359549</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.010683</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000674606</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0916215</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.0103679</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0589594</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.0461106</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0172562</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00115135</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0616554</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0163983</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0222642</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00262725</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00138224</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0875216</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0990904</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.0451302</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00538424</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00415196</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00494203</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000538733</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0406316</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00489054</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0854291</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0207327</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00364086</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00650258</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00632711</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000355507</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.0358028</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0194232</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0143484</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00310874</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00211209</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00266928</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000546309</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000454047</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000191134</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.00286722</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>5.30182e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.000255347</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.00312608</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.007516</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00218263</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.0266434</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00523998</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.000261405</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0410767</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00518293</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.0427878</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.03626</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00284452</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.000850726</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000743582</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000910168</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.0219356</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00112905</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.0304383</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.03546</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0114</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00112168</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00214021</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00336135</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.00044142</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0112741</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000518203</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.0841057</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00655298</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00083906</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00389578</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00360345</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>9.39531e-05</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00383181</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00822886</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.011094</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00310838</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00079361</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000513047</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000439206</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000213625</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>6.51859e-05</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000684832</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000238063</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00255907</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00177825</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000533285</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.0072613</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000344565</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000218814</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00633025</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00344101</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00991286</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00465219</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.0026979</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.000820837</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.000651279</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>8.10503e-06</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.00316284</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00109694</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000897407</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00274775</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.0280967</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00821673</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00104577</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.000489673</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000534204</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.000230289</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00608962</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000317393</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.0135867</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00590324</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.000298878</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00249855</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000159807</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>8.14172e-05</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00212236</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00796742</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00330107</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.00269645</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.000312165</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00022178</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>4.2193e-05</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>1.72094e-05</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>2.85698e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000312964</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000176026</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.00213583</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000827803</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.000321612</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00262176</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000295734</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>2.26788e-05</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00209702</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00178197</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.00331855</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00271026</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000482277</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000635242</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000296811</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00115635</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000477838</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.000601598</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00244264</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.0017991</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.00458293</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000812372</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000200495</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000185859</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000174243</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00606807</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000193865</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.0124812</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00570168</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000269765</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00157465</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>5.2646e-05</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>7.26552e-06</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000954054</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.00132846</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000518912</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00129922</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00025914</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>5.27963e-05</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>2.61002e-05</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>2.56224e-05</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000154783</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00204314</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000182522</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>6.92936e-05</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00165449</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000211041</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>1.7636e-05</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00124177</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000681193</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.00273008</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00210427</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000207564</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000576514</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.000232921</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000619894</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000336652</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000547261</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000435221</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000804331</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0.00366316</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000790871</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000161616</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>8.4048e-05</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000163775</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00453688</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>8.80583e-05</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.0121177</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00110823</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000169934</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.00144615</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>4.25988e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000557493</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00105023</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000474062</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.00124665</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000123139</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>1.97785e-05</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>2.38216e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000148423</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.00149881</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000160643</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>6.42681e-05</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>3.29212e-05</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00011042</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>1.39472e-05</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000308751</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.000459836</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00139009</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.0002527</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000157151</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.00040614</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>5.98466e-05</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000446071</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000223497</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0.0003554</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.00031435</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.00030405</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000482012</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000716543</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>4.1107e-05</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>6.57768e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000145394</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.0027424</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>2.80718e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.00746565</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000860428</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000118023</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000157659</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000211234</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000992771</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>1.49269e-05</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000252654</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>9.06908e-05</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>4.86255e-06</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>3.65333e-06</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000367092</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>5.64936e-05</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>3.25973e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000249147</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000417833</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000281335</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000123119</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>4.71704e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000102242</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>4.23274e-05</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.000398095</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>2.35555e-05</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000296857</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000122504</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000178186</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>7.97385e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>6.71567e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0.000144754</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0.000313065</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>2.69076e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.00472098</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000615579</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000116282</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000118806</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>9.85793e-05</td>\n",
       "      <td>sc_Neural</td>\n",
       "      <td>0.000773393</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>1.38991e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0.000121057</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>2.41592e-06</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0.000176834</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>2.62418e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>3.57892e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000322435</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0.000229421</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>2.17256e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>1.05724e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>6.38496e-06</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000100789</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000166187</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>3.31701e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000107772</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>5.34048e-05</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>0.000119982</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0.000138008</td>\n",
       "      <td>sc_Tegument_prog</td>\n",
       "      <td>1.62621e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>6.91089e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000151646</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>6.91974e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>4.321e-05</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>4.63048e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0.000312033</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>5.31115e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Parenchymal</td>\n",
       "      <td>9.91723e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>2.56126e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>2.04767e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>1.6122e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Muscle</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>5.40327e-05</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>3.03452e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>5.68548e-05</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>5.01361e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Tegument</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>9.5971e-06</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>8.55421e-06</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>3.12583e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>2.54723e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>1.86445e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neoblast</td>\n",
       "      <td>3.80496e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>5.22123e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>1.23473e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>1.66273e-05</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>1.31374e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Intestine</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>9.82237e-06</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>2.17989e-05</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>3.65757e-05</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Gland</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Neural_KK7</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Cathepsin</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "      <td>sc_Flame cells</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      pl_Neoblast: 0                 pl_Cathepsin+ cells: 10                  \\\n",
       "             Cluster Alignment score                 Cluster Alignment score   \n",
       "0        sc_Neoblast        0.598764            sc_Cathepsin         0.58303   \n",
       "1   sc_Tegument_prog     0.000372316        sc_Tegument_prog      0.00652464   \n",
       "2      sc_Neural_KK7     0.000255347             sc_Tegument      0.00312608   \n",
       "3     sc_Parenchymal     0.000238063             sc_Neoblast      0.00255907   \n",
       "4       sc_Cathepsin     0.000176026            sc_Intestine      0.00213583   \n",
       "5          sc_Neural     0.000154783               sc_Muscle      0.00204314   \n",
       "6          sc_Muscle     0.000148423               sc_Neural      0.00149881   \n",
       "7       sc_Intestine     3.65333e-06                sc_Gland     0.000367092   \n",
       "8     sc_Flame cells     2.41592e-06          sc_Flame cells     0.000176834   \n",
       "9        sc_Tegument               0          sc_Parenchymal     9.91723e-05   \n",
       "10          sc_Gland               0           sc_Neural_KK7     5.22123e-05   \n",
       "\n",
       "       pl_Muscle: 14                   pl_Epidermal: 2                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0          sc_Muscle        0.551705  sc_Tegument_prog        0.529709   \n",
       "1       sc_Intestine       0.0221045       sc_Tegument       0.0110845   \n",
       "2      sc_Neural_KK7        0.007516    sc_Parenchymal      0.00218263   \n",
       "3        sc_Neoblast      0.00177825         sc_Neural     0.000533285   \n",
       "4          sc_Neural     0.000827803       sc_Neoblast     0.000321612   \n",
       "5       sc_Cathepsin     0.000182522      sc_Cathepsin     6.92936e-05   \n",
       "6   sc_Tegument_prog     0.000160643         sc_Muscle     6.42681e-05   \n",
       "7        sc_Tegument               0          sc_Gland     5.64936e-05   \n",
       "8     sc_Parenchymal               0      sc_Intestine     2.62418e-05   \n",
       "9           sc_Gland               0     sc_Neural_KK7               0   \n",
       "10    sc_Flame cells               0    sc_Flame cells               0   \n",
       "\n",
       "    pl_Intestine: 30                 pl_Protonephridia: 29                  \\\n",
       "             Cluster Alignment score               Cluster Alignment score   \n",
       "0       sc_Intestine        0.423742        sc_Flame cells        0.384437   \n",
       "1       sc_Cathepsin       0.0359549          sc_Cathepsin        0.010683   \n",
       "2           sc_Gland       0.0266434             sc_Neural      0.00523998   \n",
       "3        sc_Tegument       0.0072613           sc_Tegument     0.000344565   \n",
       "4          sc_Muscle      0.00262176      sc_Tegument_prog     0.000295734   \n",
       "5   sc_Tegument_prog      0.00165449        sc_Parenchymal     0.000211041   \n",
       "6          sc_Neural     3.29212e-05             sc_Muscle      0.00011042   \n",
       "7     sc_Parenchymal               0           sc_Neoblast     3.25973e-05   \n",
       "8      sc_Neural_KK7               0         sc_Neural_KK7               0   \n",
       "9        sc_Neoblast               0          sc_Intestine               0   \n",
       "10    sc_Flame cells               0              sc_Gland               0   \n",
       "\n",
       "       pl_Muscle: 13                  pl_Epidermal: 11                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0          sc_Muscle        0.377412       sc_Tegument         0.36336   \n",
       "1          sc_Neural     0.000674606  sc_Tegument_prog       0.0916215   \n",
       "2      sc_Neural_KK7     0.000261405      sc_Cathepsin       0.0410767   \n",
       "3       sc_Intestine     0.000218814    sc_Parenchymal      0.00633025   \n",
       "4       sc_Cathepsin     2.26788e-05         sc_Neural      0.00209702   \n",
       "5   sc_Tegument_prog      1.7636e-05         sc_Muscle      0.00124177   \n",
       "6        sc_Neoblast     1.39472e-05      sc_Intestine     0.000308751   \n",
       "7        sc_Tegument               0          sc_Gland     0.000249147   \n",
       "8     sc_Parenchymal               0       sc_Neoblast     3.57892e-05   \n",
       "9           sc_Gland               0     sc_Neural_KK7               0   \n",
       "10    sc_Flame cells               0    sc_Flame cells               0   \n",
       "\n",
       "        pl_Neural: 8                   pl_Epidermal: 3                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0          sc_Neural        0.343862  sc_Tegument_prog        0.329127   \n",
       "1      sc_Neural_KK7       0.0103679    sc_Parenchymal       0.0589594   \n",
       "2          sc_Muscle      0.00518293       sc_Neoblast       0.0427878   \n",
       "3     sc_Parenchymal      0.00344101         sc_Neural      0.00991286   \n",
       "4       sc_Cathepsin      0.00178197       sc_Tegument      0.00331855   \n",
       "5       sc_Intestine     0.000681193      sc_Cathepsin      0.00273008   \n",
       "6        sc_Neoblast     0.000459836         sc_Muscle      0.00139009   \n",
       "7   sc_Tegument_prog     0.000417833          sc_Gland     0.000281335   \n",
       "8           sc_Gland     0.000322435      sc_Intestine     0.000229421   \n",
       "9        sc_Tegument     2.56126e-05     sc_Neural_KK7     2.04767e-05   \n",
       "10    sc_Flame cells     1.23473e-05    sc_Flame cells     1.66273e-05   \n",
       "\n",
       "   pl_Cathepsin+ cells: 4                    pl_Neoblast: 5                  \\\n",
       "                  Cluster Alignment score           Cluster Alignment score   \n",
       "0            sc_Cathepsin        0.284357       sc_Neoblast        0.271645   \n",
       "1             sc_Neoblast       0.0461106    sc_Parenchymal       0.0172562   \n",
       "2          sc_Parenchymal         0.03626         sc_Neural      0.00284452   \n",
       "3        sc_Tegument_prog      0.00465219      sc_Cathepsin       0.0026979   \n",
       "4               sc_Neural      0.00271026         sc_Muscle     0.000482277   \n",
       "5               sc_Muscle      0.00210427  sc_Tegument_prog     0.000207564   \n",
       "6             sc_Tegument       0.0002527      sc_Intestine     0.000157151   \n",
       "7            sc_Intestine     0.000123119       sc_Tegument     4.71704e-05   \n",
       "8                sc_Gland     2.17256e-05     sc_Neural_KK7     1.05724e-05   \n",
       "9          sc_Flame cells      1.6122e-05          sc_Gland               0   \n",
       "10          sc_Neural_KK7     1.31374e-05    sc_Flame cells               0   \n",
       "\n",
       "       pl_Neural: 20                  pl_Epidermal: 35                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0          sc_Neural        0.260733       sc_Tegument        0.258121   \n",
       "1   sc_Tegument_prog      0.00115135  sc_Tegument_prog       0.0616554   \n",
       "2        sc_Neoblast     0.000850726      sc_Intestine     0.000743582   \n",
       "3     sc_Flame cells     0.000820837       sc_Neoblast     0.000651279   \n",
       "4     sc_Parenchymal     0.000635242          sc_Gland     0.000296811   \n",
       "5       sc_Cathepsin     0.000576514    sc_Parenchymal     0.000232921   \n",
       "6          sc_Muscle      0.00040614      sc_Cathepsin     5.98466e-05   \n",
       "7           sc_Gland     0.000102242         sc_Neural     4.23274e-05   \n",
       "8      sc_Neural_KK7     6.38496e-06     sc_Neural_KK7               0   \n",
       "9        sc_Tegument               0         sc_Muscle               0   \n",
       "10      sc_Intestine               0    sc_Flame cells               0   \n",
       "\n",
       "      pl_Pharynx: 37                      pl_Muscle: 7                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0        sc_Tegument        0.250161         sc_Muscle        0.242052   \n",
       "1   sc_Tegument_prog       0.0163983    sc_Parenchymal       0.0222642   \n",
       "2          sc_Neural     0.000910168       sc_Neoblast       0.0219356   \n",
       "3       sc_Cathepsin     8.10503e-06      sc_Intestine      0.00316284   \n",
       "4     sc_Parenchymal               0         sc_Neural      0.00115635   \n",
       "5      sc_Neural_KK7               0  sc_Tegument_prog     0.000619894   \n",
       "6        sc_Neoblast               0      sc_Cathepsin     0.000446071   \n",
       "7          sc_Muscle               0     sc_Neural_KK7     0.000398095   \n",
       "8       sc_Intestine               0          sc_Gland     0.000100789   \n",
       "9           sc_Gland               0       sc_Tegument               0   \n",
       "10    sc_Flame cells               0    sc_Flame cells               0   \n",
       "\n",
       "   pl_Cathepsin+ cells: 28                     pl_Neural: 23                  \\\n",
       "                   Cluster Alignment score           Cluster Alignment score   \n",
       "0             sc_Cathepsin        0.223917         sc_Neural        0.208507   \n",
       "1         sc_Tegument_prog      0.00262725      sc_Cathepsin      0.00138224   \n",
       "2              sc_Neoblast        0.001225         sc_Muscle      0.00112905   \n",
       "3                sc_Muscle      0.00109694    sc_Parenchymal     0.000897407   \n",
       "4                sc_Neural     0.000477838    sc_Flame cells     0.000601598   \n",
       "5             sc_Intestine     0.000336652  sc_Tegument_prog     0.000547261   \n",
       "6              sc_Tegument     0.000223497       sc_Neoblast       0.0003554   \n",
       "7           sc_Flame cells     2.35555e-05      sc_Intestine     0.000296857   \n",
       "8           sc_Parenchymal               0          sc_Gland     0.000166187   \n",
       "9            sc_Neural_KK7               0       sc_Tegument     5.40327e-05   \n",
       "10                sc_Gland               0     sc_Neural_KK7     9.82237e-06   \n",
       "\n",
       "        pl_Neural: 1                   pl_Intestine: 6                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0          sc_Neural        0.185195  sc_Tegument_prog        0.173964   \n",
       "1     sc_Parenchymal       0.0875216      sc_Cathepsin       0.0990904   \n",
       "2        sc_Neoblast       0.0304383       sc_Neoblast         0.03546   \n",
       "3       sc_Cathepsin      0.00274775    sc_Parenchymal       0.0280967   \n",
       "4          sc_Muscle      0.00244264         sc_Muscle       0.0017991   \n",
       "5   sc_Tegument_prog     0.000435221         sc_Neural     0.000804331   \n",
       "6      sc_Neural_KK7      0.00031435       sc_Tegument      0.00030405   \n",
       "7       sc_Intestine     0.000122504      sc_Intestine     0.000178186   \n",
       "8           sc_Gland     3.31701e-05          sc_Gland     0.000107772   \n",
       "9        sc_Tegument     3.03452e-05    sc_Flame cells     5.68548e-05   \n",
       "10    sc_Flame cells               0     sc_Neural_KK7     2.17989e-05   \n",
       "\n",
       "    pl_Epidermal: 24                     pl_Neural: 18                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0        sc_Tegument        0.163489         sc_Neural        0.154325   \n",
       "1          sc_Neural       0.0451302    sc_Parenchymal      0.00538424   \n",
       "2   sc_Tegument_prog          0.0114       sc_Neoblast      0.00112168   \n",
       "3       sc_Cathepsin      0.00821673  sc_Tegument_prog      0.00104577   \n",
       "4     sc_Flame cells      0.00458293      sc_Intestine     0.000812372   \n",
       "5     sc_Parenchymal      0.00366316         sc_Muscle     0.000790871   \n",
       "6          sc_Muscle     0.000482012      sc_Cathepsin     0.000716543   \n",
       "7       sc_Intestine     7.97385e-05     sc_Neural_KK7     6.71567e-05   \n",
       "8      sc_Neural_KK7               0       sc_Tegument     5.34048e-05   \n",
       "9        sc_Neoblast               0          sc_Gland     5.01361e-05   \n",
       "10          sc_Gland               0    sc_Flame cells     3.65757e-05   \n",
       "\n",
       "     pl_Neoblast: 22                 pl_Parapharyngeal: 38                  \\\n",
       "             Cluster Alignment score               Cluster Alignment score   \n",
       "0        sc_Neoblast        0.153925              sc_Gland        0.152423   \n",
       "1     sc_Parenchymal      0.00415196      sc_Tegument_prog      0.00494203   \n",
       "2       sc_Cathepsin      0.00214021           sc_Neoblast      0.00336135   \n",
       "3      sc_Neural_KK7     0.000489673             sc_Muscle     0.000534204   \n",
       "4   sc_Tegument_prog     0.000200495          sc_Cathepsin     0.000185859   \n",
       "5          sc_Neural     0.000161616        sc_Parenchymal      8.4048e-05   \n",
       "6          sc_Muscle      4.1107e-05             sc_Neural     6.57768e-05   \n",
       "7        sc_Tegument               0           sc_Tegument               0   \n",
       "8       sc_Intestine               0         sc_Neural_KK7               0   \n",
       "9           sc_Gland               0          sc_Intestine               0   \n",
       "10    sc_Flame cells               0        sc_Flame cells               0   \n",
       "\n",
       "       pl_Muscle: 16                    pl_Pharynx: 25                  \\\n",
       "             Cluster Alignment score           Cluster Alignment score   \n",
       "0          sc_Muscle        0.146065       sc_Tegument        0.139406   \n",
       "1       sc_Intestine     0.000538733  sc_Tegument_prog       0.0406316   \n",
       "2      sc_Neural_KK7      0.00044142    sc_Parenchymal       0.0112741   \n",
       "3        sc_Neoblast     0.000230289       sc_Neoblast      0.00608962   \n",
       "4     sc_Parenchymal     0.000174243      sc_Cathepsin      0.00606807   \n",
       "5          sc_Neural     0.000163775         sc_Neural      0.00453688   \n",
       "6           sc_Gland     0.000145394      sc_Intestine       0.0027424   \n",
       "7       sc_Cathepsin     0.000144754         sc_Muscle     0.000313065   \n",
       "8   sc_Tegument_prog     0.000119982          sc_Gland     0.000138008   \n",
       "9        sc_Tegument               0     sc_Neural_KK7      9.5971e-06   \n",
       "10    sc_Flame cells               0    sc_Flame cells               0   \n",
       "\n",
       "        pl_Neural: 9                 pl_Parapharyngeal: 12                  \\\n",
       "             Cluster Alignment score               Cluster Alignment score   \n",
       "0          sc_Neural        0.122546        sc_Parenchymal        0.122045   \n",
       "1          sc_Muscle      0.00489054          sc_Cathepsin       0.0854291   \n",
       "2     sc_Parenchymal     0.000518203             sc_Neural       0.0841057   \n",
       "3        sc_Tegument     0.000317393             sc_Muscle       0.0135867   \n",
       "4       sc_Cathepsin     0.000193865      sc_Tegument_prog       0.0124812   \n",
       "5        sc_Neoblast     8.80583e-05           sc_Neoblast       0.0121177   \n",
       "6       sc_Intestine     2.80718e-05           sc_Tegument      0.00746565   \n",
       "7      sc_Neural_KK7     2.69076e-05              sc_Gland      0.00472098   \n",
       "8   sc_Tegument_prog     1.62621e-05         sc_Neural_KK7     6.91089e-05   \n",
       "9           sc_Gland               0          sc_Intestine     8.55421e-06   \n",
       "10    sc_Flame cells               0        sc_Flame cells               0   \n",
       "\n",
       "   pl_Protonephridia: 26                     pl_Neural: 21                  \\\n",
       "                 Cluster Alignment score           Cluster Alignment score   \n",
       "0         sc_Flame cells         0.12063         sc_Neural        0.115261   \n",
       "1         sc_Parenchymal       0.0207327       sc_Neoblast      0.00364086   \n",
       "2           sc_Cathepsin      0.00655298  sc_Tegument_prog      0.00083906   \n",
       "3              sc_Neural      0.00590324    sc_Flame cells     0.000298878   \n",
       "4            sc_Neoblast      0.00570168    sc_Parenchymal     0.000269765   \n",
       "5       sc_Tegument_prog      0.00110823      sc_Cathepsin     0.000169934   \n",
       "6              sc_Muscle     0.000860428      sc_Intestine     0.000118023   \n",
       "7           sc_Intestine     0.000615579       sc_Tegument     0.000116282   \n",
       "8            sc_Tegument     0.000151646         sc_Muscle     6.91974e-05   \n",
       "9               sc_Gland     3.12583e-05     sc_Neural_KK7     2.54723e-05   \n",
       "10         sc_Neural_KK7               0          sc_Gland               0   \n",
       "\n",
       "   pl_Cathepsin+ cells: 15                 pl_Cathepsin+ cells: 17  \\\n",
       "                   Cluster Alignment score                 Cluster   \n",
       "0             sc_Cathepsin        0.075856            sc_Cathepsin   \n",
       "1           sc_Parenchymal      0.00650258               sc_Muscle   \n",
       "2                sc_Muscle      0.00389578               sc_Neural   \n",
       "3              sc_Neoblast      0.00249855          sc_Parenchymal   \n",
       "4                sc_Neural      0.00157465             sc_Neoblast   \n",
       "5         sc_Tegument_prog      0.00144615        sc_Tegument_prog   \n",
       "6              sc_Tegument     0.000157659             sc_Tegument   \n",
       "7             sc_Intestine     0.000118806           sc_Neural_KK7   \n",
       "8            sc_Neural_KK7       4.321e-05            sc_Intestine   \n",
       "9           sc_Flame cells     1.86445e-05                sc_Gland   \n",
       "10                sc_Gland               0          sc_Flame cells   \n",
       "\n",
       "                       pl_Neural: 33                 pl_Protonephridia: 40  \\\n",
       "   Alignment score           Cluster Alignment score               Cluster   \n",
       "0        0.0687788         sc_Neural        0.066575           sc_Tegument   \n",
       "1       0.00632711         sc_Muscle     0.000355507        sc_Flame cells   \n",
       "2       0.00360345    sc_Parenchymal     9.39531e-05      sc_Tegument_prog   \n",
       "3      0.000159807  sc_Tegument_prog     8.14172e-05          sc_Cathepsin   \n",
       "4       5.2646e-05       sc_Neoblast     7.26552e-06          sc_Intestine   \n",
       "5      4.25988e-05       sc_Tegument               0             sc_Muscle   \n",
       "6                0     sc_Neural_KK7               0              sc_Gland   \n",
       "7                0      sc_Intestine               0             sc_Neural   \n",
       "8                0          sc_Gland               0        sc_Parenchymal   \n",
       "9                0    sc_Flame cells               0           sc_Neoblast   \n",
       "10               0      sc_Cathepsin               0         sc_Neural_KK7   \n",
       "\n",
       "                    pl_Intestine: 19                     pl_Neural: 32  \\\n",
       "   Alignment score           Cluster Alignment score           Cluster   \n",
       "0        0.0475718      sc_Cathepsin       0.0449842     sc_Neural_KK7   \n",
       "1        0.0358028    sc_Parenchymal       0.0194232      sc_Cathepsin   \n",
       "2       0.00383181  sc_Tegument_prog      0.00822886       sc_Tegument   \n",
       "3       0.00212236       sc_Neoblast      0.00796742    sc_Parenchymal   \n",
       "4      0.000954054          sc_Gland      0.00132846         sc_Muscle   \n",
       "5      0.000557493         sc_Muscle      0.00105023         sc_Neural   \n",
       "6      0.000211234      sc_Intestine     0.000992771       sc_Neoblast   \n",
       "7      9.85793e-05         sc_Neural     0.000773393  sc_Tegument_prog   \n",
       "8      4.63048e-05       sc_Tegument     0.000312033      sc_Intestine   \n",
       "9      3.80496e-05     sc_Neural_KK7               0          sc_Gland   \n",
       "10               0    sc_Flame cells               0    sc_Flame cells   \n",
       "\n",
       "                       pl_Neural: 36                     pl_Neural: 34  \\\n",
       "   Alignment score           Cluster Alignment score           Cluster   \n",
       "0        0.0432449  sc_Tegument_prog       0.0417364         sc_Neural   \n",
       "1        0.0143484         sc_Muscle      0.00310874         sc_Muscle   \n",
       "2         0.011094    sc_Parenchymal      0.00310838    sc_Parenchymal   \n",
       "3       0.00330107       sc_Tegument      0.00269645     sc_Neural_KK7   \n",
       "4      0.000518912         sc_Neural      0.00129922      sc_Cathepsin   \n",
       "5      0.000474062       sc_Neoblast      0.00124665          sc_Gland   \n",
       "6      1.49269e-05      sc_Cathepsin     0.000252654       sc_Neoblast   \n",
       "7      1.38991e-05     sc_Neural_KK7     0.000121057  sc_Tegument_prog   \n",
       "8                0      sc_Intestine     5.31115e-05       sc_Tegument   \n",
       "9                0          sc_Gland               0      sc_Intestine   \n",
       "10               0    sc_Flame cells               0    sc_Flame cells   \n",
       "\n",
       "                   pl_Cathepsin+ cells: 39                    pl_Pharynx: 27  \\\n",
       "   Alignment score                 Cluster Alignment score           Cluster   \n",
       "0        0.0311426            sc_Cathepsin       0.0252688         sc_Neural   \n",
       "1       0.00211209               sc_Muscle      0.00266928       sc_Tegument   \n",
       "2       0.00079361        sc_Tegument_prog     0.000513047         sc_Muscle   \n",
       "3      0.000312165               sc_Neural      0.00022178    sc_Flame cells   \n",
       "4       0.00025914             sc_Neoblast     5.27963e-05  sc_Tegument_prog   \n",
       "5      0.000123139             sc_Tegument               0    sc_Parenchymal   \n",
       "6      9.06908e-05          sc_Parenchymal               0     sc_Neural_KK7   \n",
       "7                0           sc_Neural_KK7               0       sc_Neoblast   \n",
       "8                0            sc_Intestine               0      sc_Intestine   \n",
       "9                0                sc_Gland               0          sc_Gland   \n",
       "10               0          sc_Flame cells               0      sc_Cathepsin   \n",
       "\n",
       "                   pl_Cathepsin+ cells: 42                  \\\n",
       "   Alignment score                 Cluster Alignment score   \n",
       "0        0.0234031            sc_Cathepsin      0.00965553   \n",
       "1      0.000546309        sc_Tegument_prog     0.000454047   \n",
       "2      0.000439206               sc_Neural     0.000213625   \n",
       "3       4.2193e-05               sc_Muscle     1.72094e-05   \n",
       "4                0             sc_Tegument               0   \n",
       "5                0          sc_Parenchymal               0   \n",
       "6                0           sc_Neural_KK7               0   \n",
       "7                0             sc_Neoblast               0   \n",
       "8                0            sc_Intestine               0   \n",
       "9                0                sc_Gland               0   \n",
       "10               0          sc_Flame cells               0   \n",
       "\n",
       "   pl_Cathepsin+ cells: 31                  pl_Intestine: 43                  \\\n",
       "                   Cluster Alignment score           Cluster Alignment score   \n",
       "0             sc_Cathepsin      0.00865499      sc_Cathepsin      0.00612058   \n",
       "1                sc_Neural     0.000191134      sc_Intestine      0.00286722   \n",
       "2           sc_Parenchymal     6.51859e-05  sc_Tegument_prog     0.000684832   \n",
       "3              sc_Neoblast     2.85698e-05       sc_Tegument     0.000312964   \n",
       "4                sc_Muscle     2.61002e-05         sc_Neural     2.56224e-05   \n",
       "5         sc_Tegument_prog     1.97785e-05         sc_Muscle     2.38216e-05   \n",
       "6              sc_Tegument               0       sc_Neoblast     4.86255e-06   \n",
       "7            sc_Neural_KK7               0    sc_Parenchymal               0   \n",
       "8             sc_Intestine               0     sc_Neural_KK7               0   \n",
       "9                 sc_Gland               0          sc_Gland               0   \n",
       "10          sc_Flame cells               0    sc_Flame cells               0   \n",
       "\n",
       "   pl_Cathepsin+ cells: 41                  \n",
       "                   Cluster Alignment score  \n",
       "0             sc_Cathepsin      0.00399417  \n",
       "1                sc_Muscle     5.30182e-05  \n",
       "2         sc_Tegument_prog               0  \n",
       "3              sc_Tegument               0  \n",
       "4           sc_Parenchymal               0  \n",
       "5            sc_Neural_KK7               0  \n",
       "6                sc_Neural               0  \n",
       "7              sc_Neoblast               0  \n",
       "8             sc_Intestine               0  \n",
       "9                 sc_Gland               0  \n",
       "10          sc_Flame cells               0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "D1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Neoblast</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Cathepsin</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Muscle</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Tegument_prog</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Intestine</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Flame cells</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Tegument</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Neural</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Gland</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Parenchymal</th>\n",
       "      <th colspan=\"2\" halign=\"left\">sc_Neural_KK7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Alignment score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.598764</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.58303</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.551705</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0.529709</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0.423742</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.384437</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.36336</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.343862</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0.152423</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.122045</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0.0432449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.271645</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.284357</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0.377412</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.329127</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.0221045</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.12063</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0.258121</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.260733</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0.0266434</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.0875216</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.0103679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0.153925</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.223917</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.242052</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.173964</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.00316284</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.0358028</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0.250161</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.208507</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.00472098</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.0589594</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.007516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.0461106</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.0990904</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.146065</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.0916215</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0.00286722</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.00458293</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.163489</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.185195</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.00132846</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.03626</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0.000489673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.0427878</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.0854291</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.0135867</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0.0616554</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.0027424</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.000820837</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.139406</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.154325</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.000367092</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.0280967</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.00044142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.03546</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.075856</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0.00632711</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.0417364</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.00213583</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.000601598</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.0475718</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0.122546</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.000322435</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.0222642</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.000398095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.0304383</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0.0687788</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.00518293</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.0406316</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.000992771</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.000298878</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0.011094</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.115261</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0.000296811</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.0207327</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.00031435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.0219356</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.0449842</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0.00489054</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0.0163983</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.000954054</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.000176834</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0.0110845</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.0841057</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.000281335</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.0194232</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0.000312165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.0121177</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.0410767</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.00389578</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.0124812</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.000812372</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>5.68548e-05</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0.00746565</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0.066575</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.000249147</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.0172562</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0.000261405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.00796742</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0.0359549</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.00310874</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.0114</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0.000743582</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>4.2193e-05</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0.0072613</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.0451302</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.000211234</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.0112741</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.000255347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.00608962</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0.0252688</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0.00266928</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.00822886</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.000681193</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>3.65757e-05</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.00331855</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0.0311426</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.000166187</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.00650258</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.000121057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.00570168</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0.0143484</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0.00262176</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.00652464</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.000615579</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>2.35555e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.00312608</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0.0234031</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000145394</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.00633025</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>6.91089e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.00364086</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.010683</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.00244264</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0.00494203</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000538733</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>1.86445e-05</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.00269645</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.00991286</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.000138008</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.00538424</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>6.71567e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0.00336135</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0.00965553</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0.00211209</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.00465219</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.000336652</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>1.66273e-05</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0.000546309</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.00590324</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0.000123139</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0.00415196</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>5.22123e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.00255907</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0.00865499</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.00210427</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.00383181</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.000308751</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>1.6122e-05</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.000344565</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.00523998</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.000107772</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.00366316</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>4.321e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.00249855</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.00821673</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.00204314</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.00262725</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.000296857</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>1.23473e-05</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0.000317393</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.00453688</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.000102242</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.00344101</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>2.69076e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.00177825</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.00655298</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.0017991</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0.00165449</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.000229421</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>2.41592e-06</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0.000312964</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0.00360345</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.000100789</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0.00330107</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>2.54723e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.00124665</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0.00612058</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.00139009</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.00144615</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0.000218814</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.000312033</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.00284452</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>5.64936e-05</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.00310838</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>2.17989e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.00606807</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.00124177</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.00115135</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.000178186</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.00030405</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.00271026</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>5.01361e-05</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0.00218263</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>2.04767e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.00112168</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0.00399417</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.00112905</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.00110823</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.000157151</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.0002527</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0.00209702</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>3.31701e-05</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.000897407</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>1.31374e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.000850726</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.00274775</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.00109694</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.00104577</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>0.000123119</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.000223497</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.00157465</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>3.12583e-05</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0.00079361</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>1.05724e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0.000651279</td>\n",
       "      <td>pl_Epidermal: 3</td>\n",
       "      <td>0.00273008</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.00105023</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.00083906</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.000122504</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.000157659</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>0.00149881</td>\n",
       "      <td>pl_Cathepsin+ cells: 4</td>\n",
       "      <td>2.17256e-05</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.000635242</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>9.82237e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.000459836</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.0026979</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.000860428</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0.000684832</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0.000118806</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0.000151646</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.00129922</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0.000518203</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>9.5971e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.0003554</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0.00214021</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.000790871</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.000619894</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.000118023</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.000116282</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.00115635</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.000269765</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>6.38496e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0.000321612</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.00212236</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0.000557493</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.000547261</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>7.97385e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>5.40327e-05</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0.000910168</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.000238063</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000230289</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.00178197</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0.000534204</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0.000513047</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>5.31115e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>5.34048e-05</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.000827803</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0.000232921</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>9.06908e-05</td>\n",
       "      <td>pl_Neural: 23</td>\n",
       "      <td>0.00138224</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0.000518912</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0.000454047</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>2.80718e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>4.71704e-05</td>\n",
       "      <td>pl_Intestine: 6</td>\n",
       "      <td>0.000804331</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.000211041</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>8.80583e-05</td>\n",
       "      <td>pl_Neural: 18</td>\n",
       "      <td>0.000716543</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.000482277</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0.000435221</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>2.62418e-05</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>3.03452e-05</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0.000773393</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000174243</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>5.27963e-05</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.000576514</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0.000482012</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>0.000417833</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>8.55421e-06</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 8</td>\n",
       "      <td>2.56126e-05</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0.000674606</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0.000159807</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>5.2646e-05</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0.000446071</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0.000439206</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.000372316</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>3.65333e-06</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>0.000533285</td>\n",
       "      <td>pl_Cathepsin+ cells: 15</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 10</td>\n",
       "      <td>9.91723e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>3.80496e-05</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0.00025914</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0.00040614</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.000295734</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0.000477838</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>9.39531e-05</td>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>pl_Epidermal: 11</td>\n",
       "      <td>3.57892e-05</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0.000252654</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0.000355507</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0.000207564</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0.000474062</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>8.4048e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>3.25973e-05</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0.000193865</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0.000313065</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0.000200495</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0.00022178</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>6.51859e-05</td>\n",
       "      <td>pl_Intestine: 19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>2.85698e-05</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0.000185859</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.000148423</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.000160643</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0.000213625</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>4.63048e-05</td>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>1.49269e-05</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0.000182522</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0.00011042</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000119982</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>0.000191134</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>1.39472e-05</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.000176026</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>6.91974e-05</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>8.14172e-05</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 1</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000163775</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>7.26552e-06</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0.000169934</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>6.42681e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 17</td>\n",
       "      <td>4.25988e-05</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 25</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0.000161616</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>4.86255e-06</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0.000144754</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>5.30182e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>1.97785e-05</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0.000154783</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 2</td>\n",
       "      <td>6.92936e-05</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>4.1107e-05</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>1.7636e-05</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 12</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 14</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>9.85793e-05</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>5.98466e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 31</td>\n",
       "      <td>2.61002e-05</td>\n",
       "      <td>pl_Neural: 9</td>\n",
       "      <td>1.62621e-05</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 16</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>6.57768e-05</td>\n",
       "      <td>pl_Neoblast: 5</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pl_Epidermal: 24</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 13</td>\n",
       "      <td>2.26788e-05</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>2.38216e-05</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>1.38991e-05</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 0</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>4.23274e-05</td>\n",
       "      <td>pl_Neural: 21</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 39</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>8.10503e-06</td>\n",
       "      <td>pl_Cathepsin+ cells: 42</td>\n",
       "      <td>1.72094e-05</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 29</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 30</td>\n",
       "      <td>3.29212e-05</td>\n",
       "      <td>pl_Neural: 32</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 28</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 34</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neoblast: 22</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 36</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 20</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>2.56224e-05</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Neural: 33</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Epidermal: 35</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 27</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Parapharyngeal: 38</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Muscle: 7</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Cathepsin+ cells: 41</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Intestine: 43</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Pharynx: 37</td>\n",
       "      <td>0</td>\n",
       "      <td>pl_Protonephridia: 40</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                sc_Neoblast                             sc_Cathepsin  \\\n",
       "                    Cluster Alignment score                  Cluster   \n",
       "0            pl_Neoblast: 0        0.598764  pl_Cathepsin+ cells: 10   \n",
       "1            pl_Neoblast: 5        0.271645   pl_Cathepsin+ cells: 4   \n",
       "2           pl_Neoblast: 22        0.153925  pl_Cathepsin+ cells: 28   \n",
       "3    pl_Cathepsin+ cells: 4       0.0461106          pl_Intestine: 6   \n",
       "4           pl_Epidermal: 3       0.0427878    pl_Parapharyngeal: 12   \n",
       "5           pl_Intestine: 6         0.03546  pl_Cathepsin+ cells: 15   \n",
       "6              pl_Neural: 1       0.0304383  pl_Cathepsin+ cells: 17   \n",
       "7              pl_Muscle: 7       0.0219356         pl_Intestine: 19   \n",
       "8     pl_Parapharyngeal: 12       0.0121177         pl_Epidermal: 11   \n",
       "9          pl_Intestine: 19      0.00796742         pl_Intestine: 30   \n",
       "10           pl_Pharynx: 25      0.00608962  pl_Cathepsin+ cells: 39   \n",
       "11    pl_Protonephridia: 26      0.00570168            pl_Neural: 32   \n",
       "12            pl_Neural: 21      0.00364086    pl_Protonephridia: 29   \n",
       "13    pl_Parapharyngeal: 38      0.00336135  pl_Cathepsin+ cells: 42   \n",
       "14  pl_Cathepsin+ cells: 10      0.00255907  pl_Cathepsin+ cells: 31   \n",
       "15  pl_Cathepsin+ cells: 15      0.00249855         pl_Epidermal: 24   \n",
       "16            pl_Muscle: 14      0.00177825    pl_Protonephridia: 26   \n",
       "17            pl_Neural: 36      0.00124665         pl_Intestine: 43   \n",
       "18  pl_Cathepsin+ cells: 28        0.001225           pl_Pharynx: 25   \n",
       "19            pl_Neural: 18      0.00112168  pl_Cathepsin+ cells: 41   \n",
       "20            pl_Neural: 20     0.000850726             pl_Neural: 1   \n",
       "21         pl_Epidermal: 35     0.000651279          pl_Epidermal: 3   \n",
       "22             pl_Neural: 8     0.000459836           pl_Neoblast: 5   \n",
       "23            pl_Neural: 23       0.0003554          pl_Neoblast: 22   \n",
       "24          pl_Epidermal: 2     0.000321612    pl_Protonephridia: 40   \n",
       "25            pl_Muscle: 16     0.000230289             pl_Neural: 8   \n",
       "26            pl_Neural: 34     9.06908e-05            pl_Neural: 23   \n",
       "27             pl_Neural: 9     8.80583e-05            pl_Neural: 18   \n",
       "28  pl_Cathepsin+ cells: 39     5.27963e-05            pl_Neural: 20   \n",
       "29  pl_Cathepsin+ cells: 17      5.2646e-05             pl_Muscle: 7   \n",
       "30    pl_Protonephridia: 40     3.80496e-05            pl_Neural: 34   \n",
       "31         pl_Epidermal: 11     3.57892e-05            pl_Neural: 36   \n",
       "32    pl_Protonephridia: 29     3.25973e-05             pl_Neural: 9   \n",
       "33  pl_Cathepsin+ cells: 31     2.85698e-05    pl_Parapharyngeal: 38   \n",
       "34            pl_Neural: 32     1.49269e-05            pl_Muscle: 14   \n",
       "35            pl_Muscle: 13     1.39472e-05           pl_Neoblast: 0   \n",
       "36            pl_Neural: 33     7.26552e-06            pl_Neural: 21   \n",
       "37         pl_Intestine: 43     4.86255e-06            pl_Muscle: 16   \n",
       "38         pl_Intestine: 30               0          pl_Epidermal: 2   \n",
       "39           pl_Pharynx: 27               0         pl_Epidermal: 35   \n",
       "40         pl_Epidermal: 24               0            pl_Muscle: 13   \n",
       "41  pl_Cathepsin+ cells: 42               0           pl_Pharynx: 37   \n",
       "42  pl_Cathepsin+ cells: 41               0            pl_Neural: 33   \n",
       "43           pl_Pharynx: 37               0           pl_Pharynx: 27   \n",
       "\n",
       "                                  sc_Muscle                  \\\n",
       "   Alignment score                  Cluster Alignment score   \n",
       "0          0.58303            pl_Muscle: 14        0.551705   \n",
       "1         0.284357            pl_Muscle: 13        0.377412   \n",
       "2         0.223917             pl_Muscle: 7        0.242052   \n",
       "3        0.0990904            pl_Muscle: 16        0.146065   \n",
       "4        0.0854291    pl_Parapharyngeal: 12       0.0135867   \n",
       "5         0.075856  pl_Cathepsin+ cells: 17      0.00632711   \n",
       "6        0.0687788             pl_Neural: 8      0.00518293   \n",
       "7        0.0449842             pl_Neural: 9      0.00489054   \n",
       "8        0.0410767  pl_Cathepsin+ cells: 15      0.00389578   \n",
       "9        0.0359549            pl_Neural: 36      0.00310874   \n",
       "10       0.0252688  pl_Cathepsin+ cells: 39      0.00266928   \n",
       "11       0.0143484         pl_Intestine: 30      0.00262176   \n",
       "12        0.010683             pl_Neural: 1      0.00244264   \n",
       "13      0.00965553            pl_Neural: 34      0.00211209   \n",
       "14      0.00865499   pl_Cathepsin+ cells: 4      0.00210427   \n",
       "15      0.00821673  pl_Cathepsin+ cells: 10      0.00204314   \n",
       "16      0.00655298          pl_Intestine: 6       0.0017991   \n",
       "17      0.00612058          pl_Epidermal: 3      0.00139009   \n",
       "18      0.00606807         pl_Epidermal: 11      0.00124177   \n",
       "19      0.00399417            pl_Neural: 23      0.00112905   \n",
       "20      0.00274775  pl_Cathepsin+ cells: 28      0.00109694   \n",
       "21      0.00273008         pl_Intestine: 19      0.00105023   \n",
       "22       0.0026979    pl_Protonephridia: 26     0.000860428   \n",
       "23      0.00214021            pl_Neural: 18     0.000790871   \n",
       "24      0.00212236    pl_Protonephridia: 40     0.000557493   \n",
       "25      0.00178197    pl_Parapharyngeal: 38     0.000534204   \n",
       "26      0.00138224            pl_Neural: 32     0.000518912   \n",
       "27     0.000716543           pl_Neoblast: 5     0.000482277   \n",
       "28     0.000576514         pl_Epidermal: 24     0.000482012   \n",
       "29     0.000446071           pl_Pharynx: 27     0.000439206   \n",
       "30      0.00025914            pl_Neural: 20      0.00040614   \n",
       "31     0.000252654            pl_Neural: 33     0.000355507   \n",
       "32     0.000193865           pl_Pharynx: 25     0.000313065   \n",
       "33     0.000185859           pl_Neoblast: 0     0.000148423   \n",
       "34     0.000182522    pl_Protonephridia: 29      0.00011042   \n",
       "35     0.000176026            pl_Neural: 21     6.91974e-05   \n",
       "36     0.000169934          pl_Epidermal: 2     6.42681e-05   \n",
       "37     0.000144754  pl_Cathepsin+ cells: 41     5.30182e-05   \n",
       "38     6.92936e-05          pl_Neoblast: 22      4.1107e-05   \n",
       "39     5.98466e-05  pl_Cathepsin+ cells: 31     2.61002e-05   \n",
       "40     2.26788e-05         pl_Intestine: 43     2.38216e-05   \n",
       "41     8.10503e-06  pl_Cathepsin+ cells: 42     1.72094e-05   \n",
       "42               0           pl_Pharynx: 37               0   \n",
       "43               0         pl_Epidermal: 35               0   \n",
       "\n",
       "           sc_Tegument_prog                             sc_Intestine  \\\n",
       "                    Cluster Alignment score                  Cluster   \n",
       "0           pl_Epidermal: 2        0.529709         pl_Intestine: 30   \n",
       "1           pl_Epidermal: 3        0.329127            pl_Muscle: 14   \n",
       "2           pl_Intestine: 6        0.173964             pl_Muscle: 7   \n",
       "3          pl_Epidermal: 11       0.0916215         pl_Intestine: 43   \n",
       "4          pl_Epidermal: 35       0.0616554           pl_Pharynx: 25   \n",
       "5             pl_Neural: 36       0.0417364  pl_Cathepsin+ cells: 10   \n",
       "6            pl_Pharynx: 25       0.0406316         pl_Intestine: 19   \n",
       "7            pl_Pharynx: 37       0.0163983    pl_Protonephridia: 40   \n",
       "8     pl_Parapharyngeal: 12       0.0124812            pl_Neural: 18   \n",
       "9          pl_Epidermal: 24          0.0114         pl_Epidermal: 35   \n",
       "10         pl_Intestine: 19      0.00822886             pl_Neural: 8   \n",
       "11  pl_Cathepsin+ cells: 10      0.00652464    pl_Protonephridia: 26   \n",
       "12    pl_Parapharyngeal: 38      0.00494203            pl_Muscle: 16   \n",
       "13   pl_Cathepsin+ cells: 4      0.00465219  pl_Cathepsin+ cells: 28   \n",
       "14    pl_Protonephridia: 40      0.00383181         pl_Epidermal: 11   \n",
       "15  pl_Cathepsin+ cells: 28      0.00262725            pl_Neural: 23   \n",
       "16         pl_Intestine: 30      0.00165449          pl_Epidermal: 3   \n",
       "17  pl_Cathepsin+ cells: 15      0.00144615            pl_Muscle: 13   \n",
       "18            pl_Neural: 20      0.00115135          pl_Intestine: 6   \n",
       "19    pl_Protonephridia: 26      0.00110823           pl_Neoblast: 5   \n",
       "20            pl_Neural: 18      0.00104577   pl_Cathepsin+ cells: 4   \n",
       "21            pl_Neural: 21      0.00083906             pl_Neural: 1   \n",
       "22         pl_Intestine: 43     0.000684832  pl_Cathepsin+ cells: 15   \n",
       "23             pl_Muscle: 7     0.000619894            pl_Neural: 21   \n",
       "24            pl_Neural: 23     0.000547261         pl_Epidermal: 24   \n",
       "25  pl_Cathepsin+ cells: 39     0.000513047            pl_Neural: 36   \n",
       "26  pl_Cathepsin+ cells: 42     0.000454047             pl_Neural: 9   \n",
       "27             pl_Neural: 1     0.000435221          pl_Epidermal: 2   \n",
       "28             pl_Neural: 8     0.000417833    pl_Parapharyngeal: 12   \n",
       "29           pl_Neoblast: 0     0.000372316           pl_Neoblast: 0   \n",
       "30    pl_Protonephridia: 29     0.000295734  pl_Cathepsin+ cells: 17   \n",
       "31           pl_Neoblast: 5     0.000207564  pl_Cathepsin+ cells: 31   \n",
       "32          pl_Neoblast: 22     0.000200495  pl_Cathepsin+ cells: 42   \n",
       "33            pl_Muscle: 14     0.000160643  pl_Cathepsin+ cells: 39   \n",
       "34            pl_Muscle: 16     0.000119982  pl_Cathepsin+ cells: 41   \n",
       "35            pl_Neural: 33     8.14172e-05            pl_Neural: 34   \n",
       "36  pl_Cathepsin+ cells: 17     4.25988e-05            pl_Neural: 33   \n",
       "37  pl_Cathepsin+ cells: 31     1.97785e-05            pl_Neural: 32   \n",
       "38            pl_Muscle: 13      1.7636e-05           pl_Pharynx: 37   \n",
       "39             pl_Neural: 9     1.62621e-05           pl_Pharynx: 27   \n",
       "40            pl_Neural: 32     1.38991e-05            pl_Neural: 20   \n",
       "41  pl_Cathepsin+ cells: 41               0    pl_Protonephridia: 29   \n",
       "42            pl_Neural: 34               0          pl_Neoblast: 22   \n",
       "43           pl_Pharynx: 27               0    pl_Parapharyngeal: 38   \n",
       "\n",
       "                             sc_Flame cells                  \\\n",
       "   Alignment score                  Cluster Alignment score   \n",
       "0         0.423742    pl_Protonephridia: 29        0.384437   \n",
       "1        0.0221045    pl_Protonephridia: 26         0.12063   \n",
       "2       0.00316284    pl_Protonephridia: 40       0.0358028   \n",
       "3       0.00286722         pl_Epidermal: 24      0.00458293   \n",
       "4        0.0027424            pl_Neural: 20     0.000820837   \n",
       "5       0.00213583            pl_Neural: 23     0.000601598   \n",
       "6      0.000992771            pl_Neural: 21     0.000298878   \n",
       "7      0.000954054  pl_Cathepsin+ cells: 10     0.000176834   \n",
       "8      0.000812372          pl_Intestine: 6     5.68548e-05   \n",
       "9      0.000743582           pl_Pharynx: 27      4.2193e-05   \n",
       "10     0.000681193            pl_Neural: 18     3.65757e-05   \n",
       "11     0.000615579  pl_Cathepsin+ cells: 28     2.35555e-05   \n",
       "12     0.000538733  pl_Cathepsin+ cells: 15     1.86445e-05   \n",
       "13     0.000336652          pl_Epidermal: 3     1.66273e-05   \n",
       "14     0.000308751   pl_Cathepsin+ cells: 4      1.6122e-05   \n",
       "15     0.000296857             pl_Neural: 8     1.23473e-05   \n",
       "16     0.000229421           pl_Neoblast: 0     2.41592e-06   \n",
       "17     0.000218814          pl_Epidermal: 2               0   \n",
       "18     0.000178186         pl_Epidermal: 35               0   \n",
       "19     0.000157151           pl_Pharynx: 37               0   \n",
       "20     0.000123119  pl_Cathepsin+ cells: 42               0   \n",
       "21     0.000122504         pl_Epidermal: 11               0   \n",
       "22     0.000118806         pl_Intestine: 30               0   \n",
       "23     0.000118023  pl_Cathepsin+ cells: 41               0   \n",
       "24     7.97385e-05  pl_Cathepsin+ cells: 39               0   \n",
       "25     5.31115e-05  pl_Cathepsin+ cells: 31               0   \n",
       "26     2.80718e-05  pl_Cathepsin+ cells: 17               0   \n",
       "27     2.62418e-05         pl_Intestine: 19               0   \n",
       "28     8.55421e-06             pl_Neural: 9               0   \n",
       "29     3.65333e-06         pl_Intestine: 43               0   \n",
       "30               0            pl_Muscle: 13               0   \n",
       "31               0            pl_Muscle: 14               0   \n",
       "32               0            pl_Muscle: 16               0   \n",
       "33               0          pl_Neoblast: 22               0   \n",
       "34               0           pl_Neoblast: 5               0   \n",
       "35               0             pl_Neural: 1               0   \n",
       "36               0           pl_Pharynx: 25               0   \n",
       "37               0    pl_Parapharyngeal: 38               0   \n",
       "38               0    pl_Parapharyngeal: 12               0   \n",
       "39               0            pl_Neural: 32               0   \n",
       "40               0            pl_Neural: 33               0   \n",
       "41               0            pl_Neural: 34               0   \n",
       "42               0            pl_Neural: 36               0   \n",
       "43               0             pl_Muscle: 7               0   \n",
       "\n",
       "                sc_Tegument                                sc_Neural  \\\n",
       "                    Cluster Alignment score                  Cluster   \n",
       "0          pl_Epidermal: 11         0.36336             pl_Neural: 8   \n",
       "1          pl_Epidermal: 35        0.258121            pl_Neural: 20   \n",
       "2            pl_Pharynx: 37        0.250161            pl_Neural: 23   \n",
       "3          pl_Epidermal: 24        0.163489             pl_Neural: 1   \n",
       "4            pl_Pharynx: 25        0.139406            pl_Neural: 18   \n",
       "5     pl_Protonephridia: 40       0.0475718             pl_Neural: 9   \n",
       "6             pl_Neural: 32        0.011094            pl_Neural: 21   \n",
       "7           pl_Epidermal: 2       0.0110845    pl_Parapharyngeal: 12   \n",
       "8     pl_Parapharyngeal: 12      0.00746565            pl_Neural: 33   \n",
       "9          pl_Intestine: 30       0.0072613         pl_Epidermal: 24   \n",
       "10          pl_Epidermal: 3      0.00331855            pl_Neural: 34   \n",
       "11  pl_Cathepsin+ cells: 10      0.00312608           pl_Pharynx: 27   \n",
       "12            pl_Neural: 36      0.00269645          pl_Epidermal: 3   \n",
       "13           pl_Pharynx: 27     0.000546309    pl_Protonephridia: 26   \n",
       "14    pl_Protonephridia: 29     0.000344565    pl_Protonephridia: 29   \n",
       "15             pl_Neural: 9     0.000317393           pl_Pharynx: 25   \n",
       "16         pl_Intestine: 43     0.000312964  pl_Cathepsin+ cells: 17   \n",
       "17         pl_Intestine: 19     0.000312033           pl_Neoblast: 5   \n",
       "18          pl_Intestine: 6      0.00030405   pl_Cathepsin+ cells: 4   \n",
       "19   pl_Cathepsin+ cells: 4       0.0002527         pl_Epidermal: 11   \n",
       "20  pl_Cathepsin+ cells: 28     0.000223497  pl_Cathepsin+ cells: 15   \n",
       "21  pl_Cathepsin+ cells: 15     0.000157659  pl_Cathepsin+ cells: 10   \n",
       "22    pl_Protonephridia: 26     0.000151646            pl_Neural: 36   \n",
       "23            pl_Neural: 21     0.000116282             pl_Muscle: 7   \n",
       "24            pl_Neural: 23     5.40327e-05           pl_Pharynx: 37   \n",
       "25            pl_Neural: 18     5.34048e-05            pl_Muscle: 14   \n",
       "26           pl_Neoblast: 5     4.71704e-05          pl_Intestine: 6   \n",
       "27             pl_Neural: 1     3.03452e-05         pl_Intestine: 19   \n",
       "28             pl_Neural: 8     2.56126e-05            pl_Muscle: 13   \n",
       "29  pl_Cathepsin+ cells: 42               0          pl_Epidermal: 2   \n",
       "30  pl_Cathepsin+ cells: 39               0  pl_Cathepsin+ cells: 28   \n",
       "31  pl_Cathepsin+ cells: 41               0            pl_Neural: 32   \n",
       "32  pl_Cathepsin+ cells: 17               0  pl_Cathepsin+ cells: 39   \n",
       "33  pl_Cathepsin+ cells: 31               0  pl_Cathepsin+ cells: 42   \n",
       "34            pl_Neural: 33               0  pl_Cathepsin+ cells: 31   \n",
       "35    pl_Parapharyngeal: 38               0            pl_Muscle: 16   \n",
       "36            pl_Neural: 34               0          pl_Neoblast: 22   \n",
       "37            pl_Muscle: 13               0           pl_Neoblast: 0   \n",
       "38            pl_Muscle: 14               0    pl_Protonephridia: 40   \n",
       "39            pl_Muscle: 16               0    pl_Parapharyngeal: 38   \n",
       "40           pl_Neoblast: 0               0         pl_Epidermal: 35   \n",
       "41          pl_Neoblast: 22               0         pl_Intestine: 30   \n",
       "42            pl_Neural: 20               0         pl_Intestine: 43   \n",
       "43             pl_Muscle: 7               0  pl_Cathepsin+ cells: 41   \n",
       "\n",
       "                                   sc_Gland                  \\\n",
       "   Alignment score                  Cluster Alignment score   \n",
       "0         0.343862    pl_Parapharyngeal: 38        0.152423   \n",
       "1         0.260733         pl_Intestine: 30       0.0266434   \n",
       "2         0.208507    pl_Parapharyngeal: 12      0.00472098   \n",
       "3         0.185195         pl_Intestine: 19      0.00132846   \n",
       "4         0.154325  pl_Cathepsin+ cells: 10     0.000367092   \n",
       "5         0.122546             pl_Neural: 8     0.000322435   \n",
       "6         0.115261         pl_Epidermal: 35     0.000296811   \n",
       "7        0.0841057          pl_Epidermal: 3     0.000281335   \n",
       "8         0.066575         pl_Epidermal: 11     0.000249147   \n",
       "9        0.0451302    pl_Protonephridia: 40     0.000211234   \n",
       "10       0.0311426            pl_Neural: 23     0.000166187   \n",
       "11       0.0234031            pl_Muscle: 16     0.000145394   \n",
       "12      0.00991286           pl_Pharynx: 25     0.000138008   \n",
       "13      0.00590324            pl_Neural: 34     0.000123139   \n",
       "14      0.00523998          pl_Intestine: 6     0.000107772   \n",
       "15      0.00453688            pl_Neural: 20     0.000102242   \n",
       "16      0.00360345             pl_Muscle: 7     0.000100789   \n",
       "17      0.00284452          pl_Epidermal: 2     5.64936e-05   \n",
       "18      0.00271026            pl_Neural: 18     5.01361e-05   \n",
       "19      0.00209702             pl_Neural: 1     3.31701e-05   \n",
       "20      0.00157465    pl_Protonephridia: 26     3.12583e-05   \n",
       "21      0.00149881   pl_Cathepsin+ cells: 4     2.17256e-05   \n",
       "22      0.00129922  pl_Cathepsin+ cells: 42               0   \n",
       "23      0.00115635  pl_Cathepsin+ cells: 39               0   \n",
       "24     0.000910168  pl_Cathepsin+ cells: 41               0   \n",
       "25     0.000827803           pl_Pharynx: 37               0   \n",
       "26     0.000804331  pl_Cathepsin+ cells: 31               0   \n",
       "27     0.000773393  pl_Cathepsin+ cells: 28               0   \n",
       "28     0.000674606  pl_Cathepsin+ cells: 17               0   \n",
       "29     0.000533285  pl_Cathepsin+ cells: 15               0   \n",
       "30     0.000477838         pl_Epidermal: 24               0   \n",
       "31     0.000474062            pl_Neural: 36               0   \n",
       "32      0.00022178           pl_Pharynx: 27               0   \n",
       "33     0.000213625             pl_Neural: 9               0   \n",
       "34     0.000191134            pl_Muscle: 13               0   \n",
       "35     0.000163775            pl_Muscle: 14               0   \n",
       "36     0.000161616    pl_Protonephridia: 29               0   \n",
       "37     0.000154783           pl_Neoblast: 0               0   \n",
       "38     9.85793e-05          pl_Neoblast: 22               0   \n",
       "39     6.57768e-05           pl_Neoblast: 5               0   \n",
       "40     4.23274e-05            pl_Neural: 21               0   \n",
       "41     3.29212e-05            pl_Neural: 32               0   \n",
       "42     2.56224e-05            pl_Neural: 33               0   \n",
       "43               0         pl_Intestine: 43               0   \n",
       "\n",
       "             sc_Parenchymal                            sc_Neural_KK7  \\\n",
       "                    Cluster Alignment score                  Cluster   \n",
       "0     pl_Parapharyngeal: 12        0.122045            pl_Neural: 32   \n",
       "1              pl_Neural: 1       0.0875216             pl_Neural: 8   \n",
       "2           pl_Epidermal: 3       0.0589594            pl_Muscle: 14   \n",
       "3    pl_Cathepsin+ cells: 4         0.03626          pl_Neoblast: 22   \n",
       "4           pl_Intestine: 6       0.0280967            pl_Muscle: 16   \n",
       "5              pl_Muscle: 7       0.0222642             pl_Muscle: 7   \n",
       "6     pl_Protonephridia: 26       0.0207327             pl_Neural: 1   \n",
       "7          pl_Intestine: 19       0.0194232            pl_Neural: 34   \n",
       "8            pl_Neoblast: 5       0.0172562            pl_Muscle: 13   \n",
       "9            pl_Pharynx: 25       0.0112741           pl_Neoblast: 0   \n",
       "10  pl_Cathepsin+ cells: 15      0.00650258            pl_Neural: 36   \n",
       "11         pl_Epidermal: 11      0.00633025    pl_Parapharyngeal: 12   \n",
       "12            pl_Neural: 18      0.00538424            pl_Neural: 18   \n",
       "13          pl_Neoblast: 22      0.00415196  pl_Cathepsin+ cells: 10   \n",
       "14         pl_Epidermal: 24      0.00366316  pl_Cathepsin+ cells: 15   \n",
       "15             pl_Neural: 8      0.00344101             pl_Neural: 9   \n",
       "16            pl_Neural: 32      0.00330107            pl_Neural: 21   \n",
       "17            pl_Neural: 36      0.00310838          pl_Intestine: 6   \n",
       "18          pl_Epidermal: 2      0.00218263          pl_Epidermal: 3   \n",
       "19            pl_Neural: 23     0.000897407   pl_Cathepsin+ cells: 4   \n",
       "20            pl_Neural: 34      0.00079361           pl_Neoblast: 5   \n",
       "21            pl_Neural: 20     0.000635242            pl_Neural: 23   \n",
       "22             pl_Neural: 9     0.000518203           pl_Pharynx: 25   \n",
       "23            pl_Neural: 21     0.000269765            pl_Neural: 20   \n",
       "24           pl_Neoblast: 0     0.000238063  pl_Cathepsin+ cells: 28   \n",
       "25         pl_Epidermal: 35     0.000232921          pl_Epidermal: 2   \n",
       "26    pl_Protonephridia: 29     0.000211041  pl_Cathepsin+ cells: 39   \n",
       "27            pl_Muscle: 16     0.000174243  pl_Cathepsin+ cells: 17   \n",
       "28  pl_Cathepsin+ cells: 17     0.000159807  pl_Cathepsin+ cells: 41   \n",
       "29  pl_Cathepsin+ cells: 10     9.91723e-05  pl_Cathepsin+ cells: 42   \n",
       "30            pl_Neural: 33     9.39531e-05         pl_Epidermal: 11   \n",
       "31    pl_Parapharyngeal: 38      8.4048e-05  pl_Cathepsin+ cells: 31   \n",
       "32  pl_Cathepsin+ cells: 31     6.51859e-05         pl_Intestine: 19   \n",
       "33    pl_Protonephridia: 40     4.63048e-05         pl_Epidermal: 24   \n",
       "34         pl_Intestine: 30               0         pl_Epidermal: 35   \n",
       "35         pl_Intestine: 43               0         pl_Intestine: 30   \n",
       "36            pl_Muscle: 13               0         pl_Intestine: 43   \n",
       "37            pl_Muscle: 14               0    pl_Protonephridia: 26   \n",
       "38  pl_Cathepsin+ cells: 42               0           pl_Pharynx: 37   \n",
       "39  pl_Cathepsin+ cells: 41               0           pl_Pharynx: 27   \n",
       "40  pl_Cathepsin+ cells: 39               0    pl_Protonephridia: 29   \n",
       "41  pl_Cathepsin+ cells: 28               0    pl_Parapharyngeal: 38   \n",
       "42           pl_Pharynx: 27               0            pl_Neural: 33   \n",
       "43           pl_Pharynx: 37               0    pl_Protonephridia: 40   \n",
       "\n",
       "                    \n",
       "   Alignment score  \n",
       "0        0.0432449  \n",
       "1        0.0103679  \n",
       "2         0.007516  \n",
       "3      0.000489673  \n",
       "4       0.00044142  \n",
       "5      0.000398095  \n",
       "6       0.00031435  \n",
       "7      0.000312165  \n",
       "8      0.000261405  \n",
       "9      0.000255347  \n",
       "10     0.000121057  \n",
       "11     6.91089e-05  \n",
       "12     6.71567e-05  \n",
       "13     5.22123e-05  \n",
       "14       4.321e-05  \n",
       "15     2.69076e-05  \n",
       "16     2.54723e-05  \n",
       "17     2.17989e-05  \n",
       "18     2.04767e-05  \n",
       "19     1.31374e-05  \n",
       "20     1.05724e-05  \n",
       "21     9.82237e-06  \n",
       "22      9.5971e-06  \n",
       "23     6.38496e-06  \n",
       "24               0  \n",
       "25               0  \n",
       "26               0  \n",
       "27               0  \n",
       "28               0  \n",
       "29               0  \n",
       "30               0  \n",
       "31               0  \n",
       "32               0  \n",
       "33               0  \n",
       "34               0  \n",
       "35               0  \n",
       "36               0  \n",
       "37               0  \n",
       "38               0  \n",
       "39               0  \n",
       "40               0  \n",
       "41               0  \n",
       "42               0  \n",
       "43               0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "D2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pl_Cathepsin+ cells: 10</th>\n",
       "      <th>pl_Cathepsin+ cells: 15</th>\n",
       "      <th>pl_Cathepsin+ cells: 17</th>\n",
       "      <th>pl_Cathepsin+ cells: 28</th>\n",
       "      <th>pl_Cathepsin+ cells: 31</th>\n",
       "      <th>pl_Cathepsin+ cells: 39</th>\n",
       "      <th>pl_Cathepsin+ cells: 4</th>\n",
       "      <th>pl_Cathepsin+ cells: 41</th>\n",
       "      <th>pl_Cathepsin+ cells: 42</th>\n",
       "      <th>pl_Epidermal: 11</th>\n",
       "      <th>pl_Epidermal: 2</th>\n",
       "      <th>pl_Epidermal: 24</th>\n",
       "      <th>pl_Epidermal: 3</th>\n",
       "      <th>pl_Epidermal: 35</th>\n",
       "      <th>pl_Intestine: 19</th>\n",
       "      <th>pl_Intestine: 30</th>\n",
       "      <th>pl_Intestine: 43</th>\n",
       "      <th>pl_Intestine: 6</th>\n",
       "      <th>pl_Muscle: 13</th>\n",
       "      <th>pl_Muscle: 14</th>\n",
       "      <th>pl_Muscle: 16</th>\n",
       "      <th>pl_Muscle: 7</th>\n",
       "      <th>pl_Neoblast: 0</th>\n",
       "      <th>pl_Neoblast: 22</th>\n",
       "      <th>pl_Neoblast: 5</th>\n",
       "      <th>pl_Neural: 1</th>\n",
       "      <th>pl_Neural: 18</th>\n",
       "      <th>pl_Neural: 20</th>\n",
       "      <th>pl_Neural: 21</th>\n",
       "      <th>pl_Neural: 23</th>\n",
       "      <th>pl_Neural: 32</th>\n",
       "      <th>pl_Neural: 33</th>\n",
       "      <th>pl_Neural: 34</th>\n",
       "      <th>pl_Neural: 36</th>\n",
       "      <th>pl_Neural: 8</th>\n",
       "      <th>pl_Neural: 9</th>\n",
       "      <th>pl_Parapharyngeal: 12</th>\n",
       "      <th>pl_Parapharyngeal: 38</th>\n",
       "      <th>pl_Pharynx: 25</th>\n",
       "      <th>pl_Pharynx: 27</th>\n",
       "      <th>pl_Pharynx: 37</th>\n",
       "      <th>pl_Protonephridia: 26</th>\n",
       "      <th>pl_Protonephridia: 29</th>\n",
       "      <th>pl_Protonephridia: 40</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sc_Cathepsin</th>\n",
       "      <td>0.583030</td>\n",
       "      <td>0.075856</td>\n",
       "      <td>0.068779</td>\n",
       "      <td>0.223917</td>\n",
       "      <td>0.008655</td>\n",
       "      <td>0.025269</td>\n",
       "      <td>0.284357</td>\n",
       "      <td>0.003994</td>\n",
       "      <td>0.009656</td>\n",
       "      <td>0.041077</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.008217</td>\n",
       "      <td>0.002730</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.044984</td>\n",
       "      <td>0.035955</td>\n",
       "      <td>0.006121</td>\n",
       "      <td>0.099090</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.000446</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.002140</td>\n",
       "      <td>0.002698</td>\n",
       "      <td>0.002748</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>0.001382</td>\n",
       "      <td>0.014348</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001782</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.085429</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.006068</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.006553</td>\n",
       "      <td>0.010683</td>\n",
       "      <td>0.002122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Flame cells</th>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004583</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000821</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.000602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.120630</td>\n",
       "      <td>0.384437</td>\n",
       "      <td>0.035803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Gland</th>\n",
       "      <td>0.000367</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>0.001328</td>\n",
       "      <td>0.026643</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000322</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004721</td>\n",
       "      <td>0.152423</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Intestine</th>\n",
       "      <td>0.002136</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.000744</td>\n",
       "      <td>0.000993</td>\n",
       "      <td>0.423742</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.000178</td>\n",
       "      <td>0.000219</td>\n",
       "      <td>0.022104</td>\n",
       "      <td>0.000539</td>\n",
       "      <td>0.003163</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000812</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000681</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002742</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000616</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Muscle</th>\n",
       "      <td>0.002043</td>\n",
       "      <td>0.003896</td>\n",
       "      <td>0.006327</td>\n",
       "      <td>0.001097</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.002669</td>\n",
       "      <td>0.002104</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.001242</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.001390</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001050</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.001799</td>\n",
       "      <td>0.377412</td>\n",
       "      <td>0.551705</td>\n",
       "      <td>0.146065</td>\n",
       "      <td>0.242052</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.002443</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.001129</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>0.000356</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>0.003109</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>0.004891</td>\n",
       "      <td>0.013587</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Neoblast</th>\n",
       "      <td>0.002559</td>\n",
       "      <td>0.002499</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.046111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000322</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042788</td>\n",
       "      <td>0.000651</td>\n",
       "      <td>0.007967</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.035460</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.001778</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.021936</td>\n",
       "      <td>0.598764</td>\n",
       "      <td>0.153925</td>\n",
       "      <td>0.271645</td>\n",
       "      <td>0.030438</td>\n",
       "      <td>0.001122</td>\n",
       "      <td>0.000851</td>\n",
       "      <td>0.003641</td>\n",
       "      <td>0.000355</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.001247</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.012118</td>\n",
       "      <td>0.003361</td>\n",
       "      <td>0.006090</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005702</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Neural</th>\n",
       "      <td>0.001499</td>\n",
       "      <td>0.001575</td>\n",
       "      <td>0.003603</td>\n",
       "      <td>0.000478</td>\n",
       "      <td>0.000191</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.002710</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.002097</td>\n",
       "      <td>0.000533</td>\n",
       "      <td>0.045130</td>\n",
       "      <td>0.009913</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000773</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.000828</td>\n",
       "      <td>0.000164</td>\n",
       "      <td>0.001156</td>\n",
       "      <td>0.000155</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.002845</td>\n",
       "      <td>0.185195</td>\n",
       "      <td>0.154325</td>\n",
       "      <td>0.260733</td>\n",
       "      <td>0.115261</td>\n",
       "      <td>0.208507</td>\n",
       "      <td>0.000474</td>\n",
       "      <td>0.066575</td>\n",
       "      <td>0.031143</td>\n",
       "      <td>0.001299</td>\n",
       "      <td>0.343862</td>\n",
       "      <td>0.122546</td>\n",
       "      <td>0.084106</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.004537</td>\n",
       "      <td>0.023403</td>\n",
       "      <td>0.000910</td>\n",
       "      <td>0.005903</td>\n",
       "      <td>0.005240</td>\n",
       "      <td>0.000099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Neural_KK7</th>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.007516</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000490</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000314</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.043245</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.010368</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Parenchymal</th>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.006503</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036260</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006330</td>\n",
       "      <td>0.002183</td>\n",
       "      <td>0.003663</td>\n",
       "      <td>0.058959</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.019423</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028097</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.022264</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.004152</td>\n",
       "      <td>0.017256</td>\n",
       "      <td>0.087522</td>\n",
       "      <td>0.005384</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.000897</td>\n",
       "      <td>0.003301</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>0.003108</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.000518</td>\n",
       "      <td>0.122045</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.011274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020733</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.000046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Tegument</th>\n",
       "      <td>0.003126</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000223</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.363360</td>\n",
       "      <td>0.011084</td>\n",
       "      <td>0.163489</td>\n",
       "      <td>0.003319</td>\n",
       "      <td>0.258121</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>0.007261</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.011094</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002696</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000317</td>\n",
       "      <td>0.007466</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.139406</td>\n",
       "      <td>0.000546</td>\n",
       "      <td>0.250161</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.047572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sc_Tegument_prog</th>\n",
       "      <td>0.006525</td>\n",
       "      <td>0.001446</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.002627</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.004652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000454</td>\n",
       "      <td>0.091621</td>\n",
       "      <td>0.529709</td>\n",
       "      <td>0.011400</td>\n",
       "      <td>0.329127</td>\n",
       "      <td>0.061655</td>\n",
       "      <td>0.008229</td>\n",
       "      <td>0.001654</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.173964</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000435</td>\n",
       "      <td>0.001046</td>\n",
       "      <td>0.001151</td>\n",
       "      <td>0.000839</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041736</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.012481</td>\n",
       "      <td>0.004942</td>\n",
       "      <td>0.040632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016398</td>\n",
       "      <td>0.001108</td>\n",
       "      <td>0.000296</td>\n",
       "      <td>0.003832</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  pl_Cathepsin+ cells: 10  pl_Cathepsin+ cells: 15  \\\n",
       "sc_Cathepsin                     0.583030                 0.075856   \n",
       "sc_Flame cells                   0.000177                 0.000019   \n",
       "sc_Gland                         0.000367                 0.000000   \n",
       "sc_Intestine                     0.002136                 0.000119   \n",
       "sc_Muscle                        0.002043                 0.003896   \n",
       "sc_Neoblast                      0.002559                 0.002499   \n",
       "sc_Neural                        0.001499                 0.001575   \n",
       "sc_Neural_KK7                    0.000052                 0.000043   \n",
       "sc_Parenchymal                   0.000099                 0.006503   \n",
       "sc_Tegument                      0.003126                 0.000158   \n",
       "sc_Tegument_prog                 0.006525                 0.001446   \n",
       "\n",
       "                  pl_Cathepsin+ cells: 17  pl_Cathepsin+ cells: 28  \\\n",
       "sc_Cathepsin                     0.068779                 0.223917   \n",
       "sc_Flame cells                   0.000000                 0.000024   \n",
       "sc_Gland                         0.000000                 0.000000   \n",
       "sc_Intestine                     0.000000                 0.000337   \n",
       "sc_Muscle                        0.006327                 0.001097   \n",
       "sc_Neoblast                      0.000053                 0.001225   \n",
       "sc_Neural                        0.003603                 0.000478   \n",
       "sc_Neural_KK7                    0.000000                 0.000000   \n",
       "sc_Parenchymal                   0.000160                 0.000000   \n",
       "sc_Tegument                      0.000000                 0.000223   \n",
       "sc_Tegument_prog                 0.000043                 0.002627   \n",
       "\n",
       "                  pl_Cathepsin+ cells: 31  pl_Cathepsin+ cells: 39  \\\n",
       "sc_Cathepsin                     0.008655                 0.025269   \n",
       "sc_Flame cells                   0.000000                 0.000000   \n",
       "sc_Gland                         0.000000                 0.000000   \n",
       "sc_Intestine                     0.000000                 0.000000   \n",
       "sc_Muscle                        0.000026                 0.002669   \n",
       "sc_Neoblast                      0.000029                 0.000053   \n",
       "sc_Neural                        0.000191                 0.000222   \n",
       "sc_Neural_KK7                    0.000000                 0.000000   \n",
       "sc_Parenchymal                   0.000065                 0.000000   \n",
       "sc_Tegument                      0.000000                 0.000000   \n",
       "sc_Tegument_prog                 0.000020                 0.000513   \n",
       "\n",
       "                  pl_Cathepsin+ cells: 4  pl_Cathepsin+ cells: 41  \\\n",
       "sc_Cathepsin                    0.284357                 0.003994   \n",
       "sc_Flame cells                  0.000016                 0.000000   \n",
       "sc_Gland                        0.000022                 0.000000   \n",
       "sc_Intestine                    0.000123                 0.000000   \n",
       "sc_Muscle                       0.002104                 0.000053   \n",
       "sc_Neoblast                     0.046111                 0.000000   \n",
       "sc_Neural                       0.002710                 0.000000   \n",
       "sc_Neural_KK7                   0.000013                 0.000000   \n",
       "sc_Parenchymal                  0.036260                 0.000000   \n",
       "sc_Tegument                     0.000253                 0.000000   \n",
       "sc_Tegument_prog                0.004652                 0.000000   \n",
       "\n",
       "                  pl_Cathepsin+ cells: 42  pl_Epidermal: 11  pl_Epidermal: 2  \\\n",
       "sc_Cathepsin                     0.009656          0.041077         0.000069   \n",
       "sc_Flame cells                   0.000000          0.000000         0.000000   \n",
       "sc_Gland                         0.000000          0.000249         0.000056   \n",
       "sc_Intestine                     0.000000          0.000309         0.000026   \n",
       "sc_Muscle                        0.000017          0.001242         0.000064   \n",
       "sc_Neoblast                      0.000000          0.000036         0.000322   \n",
       "sc_Neural                        0.000214          0.002097         0.000533   \n",
       "sc_Neural_KK7                    0.000000          0.000000         0.000000   \n",
       "sc_Parenchymal                   0.000000          0.006330         0.002183   \n",
       "sc_Tegument                      0.000000          0.363360         0.011084   \n",
       "sc_Tegument_prog                 0.000454          0.091621         0.529709   \n",
       "\n",
       "                  pl_Epidermal: 24  pl_Epidermal: 3  pl_Epidermal: 35  \\\n",
       "sc_Cathepsin              0.008217         0.002730          0.000060   \n",
       "sc_Flame cells            0.004583         0.000017          0.000000   \n",
       "sc_Gland                  0.000000         0.000281          0.000297   \n",
       "sc_Intestine              0.000080         0.000229          0.000744   \n",
       "sc_Muscle                 0.000482         0.001390          0.000000   \n",
       "sc_Neoblast               0.000000         0.042788          0.000651   \n",
       "sc_Neural                 0.045130         0.009913          0.000042   \n",
       "sc_Neural_KK7             0.000000         0.000020          0.000000   \n",
       "sc_Parenchymal            0.003663         0.058959          0.000233   \n",
       "sc_Tegument               0.163489         0.003319          0.258121   \n",
       "sc_Tegument_prog          0.011400         0.329127          0.061655   \n",
       "\n",
       "                  pl_Intestine: 19  pl_Intestine: 30  pl_Intestine: 43  \\\n",
       "sc_Cathepsin              0.044984          0.035955          0.006121   \n",
       "sc_Flame cells            0.000000          0.000000          0.000000   \n",
       "sc_Gland                  0.001328          0.026643          0.000000   \n",
       "sc_Intestine              0.000993          0.423742          0.002867   \n",
       "sc_Muscle                 0.001050          0.002622          0.000024   \n",
       "sc_Neoblast               0.007967          0.000000          0.000005   \n",
       "sc_Neural                 0.000773          0.000033          0.000026   \n",
       "sc_Neural_KK7             0.000000          0.000000          0.000000   \n",
       "sc_Parenchymal            0.019423          0.000000          0.000000   \n",
       "sc_Tegument               0.000312          0.007261          0.000313   \n",
       "sc_Tegument_prog          0.008229          0.001654          0.000685   \n",
       "\n",
       "                  pl_Intestine: 6  pl_Muscle: 13  pl_Muscle: 14  \\\n",
       "sc_Cathepsin             0.099090       0.000023       0.000183   \n",
       "sc_Flame cells           0.000057       0.000000       0.000000   \n",
       "sc_Gland                 0.000108       0.000000       0.000000   \n",
       "sc_Intestine             0.000178       0.000219       0.022104   \n",
       "sc_Muscle                0.001799       0.377412       0.551705   \n",
       "sc_Neoblast              0.035460       0.000014       0.001778   \n",
       "sc_Neural                0.000804       0.000675       0.000828   \n",
       "sc_Neural_KK7            0.000022       0.000261       0.007516   \n",
       "sc_Parenchymal           0.028097       0.000000       0.000000   \n",
       "sc_Tegument              0.000304       0.000000       0.000000   \n",
       "sc_Tegument_prog         0.173964       0.000018       0.000161   \n",
       "\n",
       "                  pl_Muscle: 16  pl_Muscle: 7  pl_Neoblast: 0  \\\n",
       "sc_Cathepsin           0.000145      0.000446        0.000176   \n",
       "sc_Flame cells         0.000000      0.000000        0.000002   \n",
       "sc_Gland               0.000145      0.000101        0.000000   \n",
       "sc_Intestine           0.000539      0.003163        0.000004   \n",
       "sc_Muscle              0.146065      0.242052        0.000148   \n",
       "sc_Neoblast            0.000230      0.021936        0.598764   \n",
       "sc_Neural              0.000164      0.001156        0.000155   \n",
       "sc_Neural_KK7          0.000441      0.000398        0.000255   \n",
       "sc_Parenchymal         0.000174      0.022264        0.000238   \n",
       "sc_Tegument            0.000000      0.000000        0.000000   \n",
       "sc_Tegument_prog       0.000120      0.000620        0.000372   \n",
       "\n",
       "                  pl_Neoblast: 22  pl_Neoblast: 5  pl_Neural: 1  \\\n",
       "sc_Cathepsin             0.002140        0.002698      0.002748   \n",
       "sc_Flame cells           0.000000        0.000000      0.000000   \n",
       "sc_Gland                 0.000000        0.000000      0.000033   \n",
       "sc_Intestine             0.000000        0.000157      0.000123   \n",
       "sc_Muscle                0.000041        0.000482      0.002443   \n",
       "sc_Neoblast              0.153925        0.271645      0.030438   \n",
       "sc_Neural                0.000162        0.002845      0.185195   \n",
       "sc_Neural_KK7            0.000490        0.000011      0.000314   \n",
       "sc_Parenchymal           0.004152        0.017256      0.087522   \n",
       "sc_Tegument              0.000000        0.000047      0.000030   \n",
       "sc_Tegument_prog         0.000200        0.000208      0.000435   \n",
       "\n",
       "                  pl_Neural: 18  pl_Neural: 20  pl_Neural: 21  pl_Neural: 23  \\\n",
       "sc_Cathepsin           0.000717       0.000577       0.000170       0.001382   \n",
       "sc_Flame cells         0.000037       0.000821       0.000299       0.000602   \n",
       "sc_Gland               0.000050       0.000102       0.000000       0.000166   \n",
       "sc_Intestine           0.000812       0.000000       0.000118       0.000297   \n",
       "sc_Muscle              0.000791       0.000406       0.000069       0.001129   \n",
       "sc_Neoblast            0.001122       0.000851       0.003641       0.000355   \n",
       "sc_Neural              0.154325       0.260733       0.115261       0.208507   \n",
       "sc_Neural_KK7          0.000067       0.000006       0.000025       0.000010   \n",
       "sc_Parenchymal         0.005384       0.000635       0.000270       0.000897   \n",
       "sc_Tegument            0.000053       0.000000       0.000116       0.000054   \n",
       "sc_Tegument_prog       0.001046       0.001151       0.000839       0.000547   \n",
       "\n",
       "                  pl_Neural: 32  pl_Neural: 33  pl_Neural: 34  pl_Neural: 36  \\\n",
       "sc_Cathepsin           0.014348       0.000000       0.000259       0.000253   \n",
       "sc_Flame cells         0.000000       0.000000       0.000000       0.000000   \n",
       "sc_Gland               0.000000       0.000000       0.000123       0.000000   \n",
       "sc_Intestine           0.000000       0.000000       0.000000       0.000053   \n",
       "sc_Muscle              0.000519       0.000356       0.002112       0.003109   \n",
       "sc_Neoblast            0.000015       0.000007       0.000091       0.001247   \n",
       "sc_Neural              0.000474       0.066575       0.031143       0.001299   \n",
       "sc_Neural_KK7          0.043245       0.000000       0.000312       0.000121   \n",
       "sc_Parenchymal         0.003301       0.000094       0.000794       0.003108   \n",
       "sc_Tegument            0.011094       0.000000       0.000000       0.002696   \n",
       "sc_Tegument_prog       0.000014       0.000081       0.000000       0.041736   \n",
       "\n",
       "                  pl_Neural: 8  pl_Neural: 9  pl_Parapharyngeal: 12  \\\n",
       "sc_Cathepsin          0.001782      0.000194               0.085429   \n",
       "sc_Flame cells        0.000012      0.000000               0.000000   \n",
       "sc_Gland              0.000322      0.000000               0.004721   \n",
       "sc_Intestine          0.000681      0.000028               0.000009   \n",
       "sc_Muscle             0.005183      0.004891               0.013587   \n",
       "sc_Neoblast           0.000460      0.000088               0.012118   \n",
       "sc_Neural             0.343862      0.122546               0.084106   \n",
       "sc_Neural_KK7         0.010368      0.000027               0.000069   \n",
       "sc_Parenchymal        0.003441      0.000518               0.122045   \n",
       "sc_Tegument           0.000026      0.000317               0.007466   \n",
       "sc_Tegument_prog      0.000418      0.000016               0.012481   \n",
       "\n",
       "                  pl_Parapharyngeal: 38  pl_Pharynx: 25  pl_Pharynx: 27  \\\n",
       "sc_Cathepsin                   0.000186        0.006068        0.000000   \n",
       "sc_Flame cells                 0.000000        0.000000        0.000042   \n",
       "sc_Gland                       0.152423        0.000138        0.000000   \n",
       "sc_Intestine                   0.000000        0.002742        0.000000   \n",
       "sc_Muscle                      0.000534        0.000313        0.000439   \n",
       "sc_Neoblast                    0.003361        0.006090        0.000000   \n",
       "sc_Neural                      0.000066        0.004537        0.023403   \n",
       "sc_Neural_KK7                  0.000000        0.000010        0.000000   \n",
       "sc_Parenchymal                 0.000084        0.011274        0.000000   \n",
       "sc_Tegument                    0.000000        0.139406        0.000546   \n",
       "sc_Tegument_prog               0.004942        0.040632        0.000000   \n",
       "\n",
       "                  pl_Pharynx: 37  pl_Protonephridia: 26  \\\n",
       "sc_Cathepsin            0.000008               0.006553   \n",
       "sc_Flame cells          0.000000               0.120630   \n",
       "sc_Gland                0.000000               0.000031   \n",
       "sc_Intestine            0.000000               0.000616   \n",
       "sc_Muscle               0.000000               0.000860   \n",
       "sc_Neoblast             0.000000               0.005702   \n",
       "sc_Neural               0.000910               0.005903   \n",
       "sc_Neural_KK7           0.000000               0.000000   \n",
       "sc_Parenchymal          0.000000               0.020733   \n",
       "sc_Tegument             0.250161               0.000152   \n",
       "sc_Tegument_prog        0.016398               0.001108   \n",
       "\n",
       "                  pl_Protonephridia: 29  pl_Protonephridia: 40  \n",
       "sc_Cathepsin                   0.010683               0.002122  \n",
       "sc_Flame cells                 0.384437               0.035803  \n",
       "sc_Gland                       0.000000               0.000211  \n",
       "sc_Intestine                   0.000000               0.000954  \n",
       "sc_Muscle                      0.000110               0.000557  \n",
       "sc_Neoblast                    0.000033               0.000038  \n",
       "sc_Neural                      0.005240               0.000099  \n",
       "sc_Neural_KK7                  0.000000               0.000000  \n",
       "sc_Parenchymal                 0.000211               0.000046  \n",
       "sc_Tegument                    0.000345               0.047572  \n",
       "sc_Tegument_prog               0.000296               0.003832  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MappingTable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SAMap provides a class to find gene pairs enriched in different cell type pairs. The method entails finding gene pairs that contribute positively to the cross-species correlation between cell types and are differentially expressed in their respective mapped cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding cluster-specific markers in pl:cluster and sc:tissue.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alec/anaconda3/envs/SAMap/lib/python3.7/site-packages/anndata/compat/__init__.py:215: FutureWarning: During AnnData slicing, found matrix at .uns['nnm'] that happens to be dimensioned at n_obs×n_obs (50562×50562).\n",
      "\n",
      "These matrices should now be stored in the .obsp attribute.\n",
      "This slicing behavior will be removed in anndata 0.8.\n",
      "  FutureWarning,\n",
      "... storing 'cluster' as categorical\n",
      "/home/alec/anaconda3/envs/SAMap/lib/python3.7/site-packages/scanpy/tools/_rank_genes_groups.py:404: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/alec/anaconda3/envs/SAMap/lib/python3.7/site-packages/anndata/compat/__init__.py:215: FutureWarning: During AnnData slicing, found matrix at .uns['nnm'] that happens to be dimensioned at n_obs×n_obs (7657×7657).\n",
      "\n",
      "These matrices should now be stored in the .obsp attribute.\n",
      "This slicing behavior will be removed in anndata 0.8.\n",
      "  FutureWarning,\n",
      "... storing 'tissue' as categorical\n",
      "/home/alec/anaconda3/envs/SAMap/lib/python3.7/site-packages/scanpy/tools/_rank_genes_groups.py:405: RuntimeWarning: overflow encountered in expm1\n",
      "  self.expm1_func(mean_rest) + 1e-9\n",
      "/home/alec/anaconda3/envs/SAMap/lib/python3.7/site-packages/scanpy/tools/_rank_genes_groups.py:405: RuntimeWarning: invalid value encountered in true_divide\n",
      "  self.expm1_func(mean_rest) + 1e-9\n",
      "/home/alec/anaconda3/envs/SAMap/lib/python3.7/site-packages/scanpy/tools/_rank_genes_groups.py:408: RuntimeWarning: divide by zero encountered in log2\n",
      "  foldchanges[global_indices]\n"
     ]
    }
   ],
   "source": [
    "gpf = GenePairFinder(sm,k1=k1,k2=k2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`gpf.find_genes` can now be used to find gene pairs enriched in a cell type mapping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "n1 = 'Neoblast: 0' #cell type ID from organism 1 (must be present in `sam1.adata.obs[k1]`)\n",
    "n2 = 'Neoblast' #cell type ID from organism 2 (must be present in `sam2.adata.obs[k1]`)\n",
    "Gp,G1,G2 = gpf.find_genes(n1,n2)\n",
    "#Gp are the gene pairs, G1 are the genes from organism 1, G2 are the genes from organism 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['pl_dd_Smed_v4_659_0_1;sc_Smp_179320',\n",
       "        'pl_dd_Smed_v4_2787_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_648_0_1;sc_Smp_027920',\n",
       "        'pl_dd_Smed_v4_2614_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_15019_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_032500',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_143490',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_172530',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_094140',\n",
       "        'pl_dd_Smed_v4_7837_0_1;sc_Smp_082490',\n",
       "        'pl_dd_Smed_v4_1484_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_054840',\n",
       "        'pl_dd_Smed_v4_10594_0_1;sc_Smp_162370',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_3708_0_1;sc_Smp_245030',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_037590',\n",
       "        'pl_dd_Smed_v4_11487_0_1;sc_Smp_082490',\n",
       "        'pl_dd_Smed_v4_756_0_1;sc_Smp_179320',\n",
       "        'pl_dd_Smed_v4_12365_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_032500',\n",
       "        'pl_dd_Smed_v4_2787_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_143490',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_172530',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_094140',\n",
       "        'pl_dd_Smed_v4_1651_0_1;sc_Smp_009030',\n",
       "        'pl_dd_Smed_v4_2787_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_054840',\n",
       "        'pl_dd_Smed_v4_1920_0_1;sc_Smp_046500',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_037590',\n",
       "        'pl_dd_Smed_v4_2614_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_15019_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_2614_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_15019_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_032500',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_143490',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_172530',\n",
       "        'pl_dd_Smed_v4_2286_0_1;sc_Smp_055890',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_094140',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_032500',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_143490',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_172530',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_094140',\n",
       "        'pl_dd_Smed_v4_469_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_648_0_1;sc_Smp_034670',\n",
       "        'pl_dd_Smed_v4_1258_0_1;sc_Smp_179320',\n",
       "        'pl_dd_Smed_v4_1484_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_13700_0_1;sc_Smp_345010',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_054840',\n",
       "        'pl_dd_Smed_v4_7837_0_1;sc_Smp_047620',\n",
       "        'pl_dd_Smed_v4_1484_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_8630_0_1;sc_Smp_040020',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_054840',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_037590',\n",
       "        'pl_dd_Smed_v4_8626_0_1;sc_Smp_124230',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_037590',\n",
       "        'pl_dd_Smed_v4_601_0_1;sc_Smp_162370',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_337870',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_032500',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_143490',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_172530',\n",
       "        'pl_dd_Smed_v4_12365_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_094140',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_5862_0_1;sc_Smp_055890',\n",
       "        'pl_dd_Smed_v4_532_0_1;sc_Smp_040020',\n",
       "        'pl_dd_Smed_v4_12365_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_334160',\n",
       "        'pl_dd_Smed_v4_5663_0_1;sc_Smp_058320',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_054840',\n",
       "        'pl_dd_Smed_v4_2921_0_1;sc_Smp_079240',\n",
       "        'pl_dd_Smed_v4_13580_0_1;sc_Smp_167540',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_334160',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_037590',\n",
       "        'pl_dd_Smed_v4_10335_0_1;sc_Smp_004060',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_151560',\n",
       "        'pl_dd_Smed_v4_13347_0_1;sc_Smp_040020',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_128200',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_2312_0_1;sc_Smp_179710',\n",
       "        'pl_dd_Smed_v4_9628_0_1;sc_Smp_137390',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_334160',\n",
       "        'pl_dd_Smed_v4_7575_0_1;sc_Smp_128210',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_128200',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_10217_0_1;sc_Smp_054610',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_128200',\n",
       "        'pl_dd_Smed_v4_928_0_1;sc_Smp_179650',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_309960',\n",
       "        'pl_dd_Smed_v4_1194_0_1;sc_Smp_179320',\n",
       "        'pl_dd_Smed_v4_469_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_311740',\n",
       "        'pl_dd_Smed_v4_152_0_1;sc_Smp_027920',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_309960',\n",
       "        'pl_dd_Smed_v4_10338_0_1;sc_Smp_145810',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_469_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_311740',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_157300',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_4712_0_1;sc_Smp_151560',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_309960',\n",
       "        'pl_dd_Smed_v4_2592_0_1;sc_Smp_342620',\n",
       "        'pl_dd_Smed_v4_12599_0_1;sc_Smp_046910',\n",
       "        'pl_dd_Smed_v4_6154_0_1;sc_Smp_145810',\n",
       "        'pl_dd_Smed_v4_7227_0_1;sc_Smp_076600',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_311740',\n",
       "        'pl_dd_Smed_v4_9624_0_1;sc_Smp_347010',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_1260_0_1;sc_Smp_200410',\n",
       "        'pl_dd_Smed_v4_3061_0_1;sc_Smp_115290',\n",
       "        'pl_dd_Smed_v4_928_0_1;sc_Smp_346320',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_13269_0_1;sc_Smp_128870',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_8732_0_1;sc_Smp_066260',\n",
       "        'pl_dd_Smed_v4_13844_0_1;sc_Smp_040020',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_5816_0_1;sc_Smp_115340',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_12818_0_1;sc_Smp_052390',\n",
       "        'pl_dd_Smed_v4_5120_0_1;sc_Smp_033020',\n",
       "        'pl_dd_Smed_v4_5865_0_1;sc_Smp_173700',\n",
       "        'pl_dd_Smed_v4_2595_0_1;sc_Smp_315010',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_3904_0_1;sc_Smp_148930',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_337870',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_8732_0_1;sc_Smp_140530',\n",
       "        'pl_dd_Smed_v4_976_0_1;sc_Smp_084620',\n",
       "        'pl_dd_Smed_v4_1302_0_1;sc_Smp_029240',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_157300',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_147970',\n",
       "        'pl_dd_Smed_v4_2682_0_1;sc_Smp_151560',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_1646_0_1;sc_Smp_000980',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_309950',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_8206_0_1;sc_Smp_331790',\n",
       "        'pl_dd_Smed_v4_4994_0_1;sc_Smp_048660',\n",
       "        'pl_dd_Smed_v4_2592_0_1;sc_Smp_137330',\n",
       "        'pl_dd_Smed_v4_5956_0_1;sc_Smp_151560',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_5156_0_1;sc_Smp_086860',\n",
       "        'pl_dd_Smed_v4_2811_0_1;sc_Smp_044280',\n",
       "        'pl_dd_Smed_v4_1204_0_1;sc_Smp_074390',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_095920',\n",
       "        'pl_dd_Smed_v4_13140_0_1;sc_Smp_058990',\n",
       "        'pl_dd_Smed_v4_3342_0_1;sc_Smp_121800',\n",
       "        'pl_dd_Smed_v4_2016_0_1;sc_Smp_124300',\n",
       "        'pl_dd_Smed_v4_8630_0_1;sc_Smp_089860',\n",
       "        'pl_dd_Smed_v4_3061_0_1;sc_Smp_016410',\n",
       "        'pl_dd_Smed_v4_8732_0_1;sc_Smp_096360',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_147970',\n",
       "        'pl_dd_Smed_v4_16174_0_1;sc_Smp_137090',\n",
       "        'pl_dd_Smed_v4_11717_0_1;sc_Smp_137660',\n",
       "        'pl_dd_Smed_v4_14547_0_1;sc_Smp_331790',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_309950',\n",
       "        'pl_dd_Smed_v4_3904_0_1;sc_Smp_100090',\n",
       "        'pl_dd_Smed_v4_2297_0_1;sc_Smp_078090',\n",
       "        'pl_dd_Smed_v4_900_0_1;sc_Smp_055800',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_1646_0_1;sc_Smp_100090',\n",
       "        'pl_dd_Smed_v4_2065_0_1;sc_Smp_212260',\n",
       "        'pl_dd_Smed_v4_3776_0_1;sc_Smp_063110',\n",
       "        'pl_dd_Smed_v4_8732_0_1;sc_Smp_332200',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_4994_0_1;sc_Smp_248170',\n",
       "        'pl_dd_Smed_v4_4341_0_1;sc_Smp_080530',\n",
       "        'pl_dd_Smed_v4_7808_0_1;sc_Smp_321180',\n",
       "        'pl_dd_Smed_v4_11417_0_1;sc_Smp_051320',\n",
       "        'pl_dd_Smed_v4_9695_0_1;sc_Smp_066260',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_157300',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_10338_0_1;sc_Smp_129260',\n",
       "        'pl_dd_Smed_v4_673_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_147970',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_211330',\n",
       "        'pl_dd_Smed_v4_9585_0_1;sc_Smp_137660',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_004080',\n",
       "        'pl_dd_Smed_v4_11558_0_1;sc_Smp_160190',\n",
       "        'pl_dd_Smed_v4_4808_0_1;sc_Smp_108390',\n",
       "        'pl_dd_Smed_v4_1807_0_1;sc_Smp_167200',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_309950',\n",
       "        'pl_dd_Smed_v4_1089_0_1;sc_Smp_341240',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_9695_0_1;sc_Smp_050820',\n",
       "        'pl_dd_Smed_v4_532_0_1;sc_Smp_089860',\n",
       "        'pl_dd_Smed_v4_10019_0_1;sc_Smp_145620',\n",
       "        'pl_dd_Smed_v4_2787_0_1;sc_Smp_031720',\n",
       "        'pl_dd_Smed_v4_4087_0_1;sc_Smp_245240',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_802_0_1;sc_Smp_000980',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_4087_0_1;sc_Smp_068920',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_5611_0_1;sc_Smp_157250',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_7612_0_1;sc_Smp_212260',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_057020',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_3239_0_1;sc_Smp_340420',\n",
       "        'pl_dd_Smed_v4_1982_0_1;sc_Smp_048660',\n",
       "        'pl_dd_Smed_v4_13972_0_1;sc_Smp_137660',\n",
       "        'pl_dd_Smed_v4_6154_0_1;sc_Smp_129260',\n",
       "        'pl_dd_Smed_v4_6734_0_1;sc_Smp_151560',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_2787_0_1;sc_Smp_337740',\n",
       "        'pl_dd_Smed_v4_6075_0_1;sc_Smp_098850',\n",
       "        'pl_dd_Smed_v4_5140_0_1;sc_Smp_341530',\n",
       "        'pl_dd_Smed_v4_2086_0_1;sc_Smp_169320',\n",
       "        'pl_dd_Smed_v4_1095_0_1;sc_Smp_000980',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_103950',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_077180',\n",
       "        'pl_dd_Smed_v4_904_0_1;sc_Smp_247570',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_077180',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_802_0_1;sc_Smp_100090',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_157300',\n",
       "        'pl_dd_Smed_v4_1535_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_310770',\n",
       "        'pl_dd_Smed_v4_1609_0_1;sc_Smp_342910',\n",
       "        'pl_dd_Smed_v4_1982_0_1;sc_Smp_248170',\n",
       "        'pl_dd_Smed_v4_5055_0_1;sc_Smp_212260',\n",
       "        'pl_dd_Smed_v4_4679_0_1;sc_Smp_044280',\n",
       "        'pl_dd_Smed_v4_6009_0_1;sc_Smp_066610',\n",
       "        'pl_dd_Smed_v4_7808_0_1;sc_Smp_175230',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_2592_0_1;sc_Smp_064590',\n",
       "        'pl_dd_Smed_v4_6610_0_1;sc_Smp_167810',\n",
       "        'pl_dd_Smed_v4_13347_0_1;sc_Smp_089860',\n",
       "        'pl_dd_Smed_v4_4379_0_1;sc_Smp_124840',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_337870',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_1038_0_1;sc_Smp_341240',\n",
       "        'pl_dd_Smed_v4_4471_0_1;sc_Smp_028440',\n",
       "        'pl_dd_Smed_v4_1134_0_1;sc_Smp_021070',\n",
       "        'pl_dd_Smed_v4_1516_0_1;sc_Smp_061560',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_11475_0_1;sc_Smp_127050',\n",
       "        'pl_dd_Smed_v4_2065_0_1;sc_Smp_068950',\n",
       "        'pl_dd_Smed_v4_5644_0_1;sc_Smp_094920',\n",
       "        'pl_dd_Smed_v4_6301_0_1;sc_Smp_104980',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_1983_0_1;sc_Smp_169340',\n",
       "        'pl_dd_Smed_v4_2614_0_1;sc_Smp_031720',\n",
       "        'pl_dd_Smed_v4_13890_0_1;sc_Smp_172480',\n",
       "        'pl_dd_Smed_v4_1095_0_1;sc_Smp_100090',\n",
       "        'pl_dd_Smed_v4_10739_0_1;sc_Smp_337590',\n",
       "        'pl_dd_Smed_v4_15019_0_1;sc_Smp_031720',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_175590',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_337870',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_131770',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_127120',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_332290',\n",
       "        'pl_dd_Smed_v4_1416_0_1;sc_Smp_050360',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_1928_0_1;sc_Smp_104660',\n",
       "        'pl_dd_Smed_v4_8634_0_1;sc_Smp_179180',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_3517_0_1;sc_Smp_004470',\n",
       "        'pl_dd_Smed_v4_1682_0_1;sc_Smp_151030',\n",
       "        'pl_dd_Smed_v4_12569_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_1089_0_1;sc_Smp_075210',\n",
       "        'pl_dd_Smed_v4_764_0_1;sc_Smp_133350',\n",
       "        'pl_dd_Smed_v4_601_0_1;sc_Smp_003770',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_1222_0_1;sc_Smp_300810',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_2614_0_1;sc_Smp_337740',\n",
       "        'pl_dd_Smed_v4_1302_0_1;sc_Smp_337860',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_131360',\n",
       "        'pl_dd_Smed_v4_887_0_1;sc_Smp_026770',\n",
       "        'pl_dd_Smed_v4_1056_0_1;sc_Smp_060910',\n",
       "        'pl_dd_Smed_v4_13727_0_1;sc_Smp_079050',\n",
       "        'pl_dd_Smed_v4_15019_0_1;sc_Smp_337740',\n",
       "        'pl_dd_Smed_v4_4362_0_1;sc_Smp_074230',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_443_0_1;sc_Smp_087760',\n",
       "        'pl_dd_Smed_v4_997_0_1;sc_Smp_333660',\n",
       "        'pl_dd_Smed_v4_998_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_7612_0_1;sc_Smp_068950',\n",
       "        'pl_dd_Smed_v4_565_0_1;sc_Smp_015530',\n",
       "        'pl_dd_Smed_v4_1192_0_1;sc_Smp_317550',\n",
       "        'pl_dd_Smed_v4_1121_0_1;sc_Smp_136090',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_3584_0_1;sc_Smp_016990',\n",
       "        'pl_dd_Smed_v4_4443_0_1;sc_Smp_035280',\n",
       "        'pl_dd_Smed_v4_1535_0_1;sc_Smp_055960',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_337870',\n",
       "        'pl_dd_Smed_v4_4263_0_1;sc_Smp_178740',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_152_0_1;sc_Smp_034670',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_2166_0_1;sc_Smp_066740',\n",
       "        'pl_dd_Smed_v4_1253_0_1;sc_Smp_045510',\n",
       "        'pl_dd_Smed_v4_3517_0_1;sc_Smp_158110',\n",
       "        'pl_dd_Smed_v4_420_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_420_0_1;sc_Smp_009600',\n",
       "        'pl_dd_Smed_v4_1755_0_1;sc_Smp_029240',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_449_0_1;sc_Smp_163050',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_163110',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_267120',\n",
       "        'pl_dd_Smed_v4_1192_0_1;sc_Smp_321380',\n",
       "        'pl_dd_Smed_v4_2086_0_1;sc_Smp_020040',\n",
       "        'pl_dd_Smed_v4_8491_0_1;sc_Smp_157290',\n",
       "        'pl_dd_Smed_v4_5677_0_1;sc_Smp_027300',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_336830',\n",
       "        'pl_dd_Smed_v4_1484_0_1;sc_Smp_031720',\n",
       "        'pl_dd_Smed_v4_560_0_1;sc_Smp_012800',\n",
       "        'pl_dd_Smed_v4_12608_0_1;sc_Smp_179650',\n",
       "        'pl_dd_Smed_v4_4994_0_1;sc_Smp_053470',\n",
       "        'pl_dd_Smed_v4_5573_0_1;sc_Smp_025130',\n",
       "        'pl_dd_Smed_v4_5055_0_1;sc_Smp_068950',\n",
       "        'pl_dd_Smed_v4_1801_0_1;sc_Smp_017660',\n",
       "        'pl_dd_Smed_v4_5312_0_1;sc_Smp_161160',\n",
       "        'pl_dd_Smed_v4_2102_0_1;sc_Smp_150690',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_038840',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_175590',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_8720_0_1;sc_Smp_145470',\n",
       "        'pl_dd_Smed_v4_805_0_1;sc_Smp_096010',\n",
       "        'pl_dd_Smed_v4_1121_0_1;sc_Smp_330330',\n",
       "        'pl_dd_Smed_v4_3731_0_1;sc_Smp_030180',\n",
       "        'pl_dd_Smed_v4_1134_0_1;sc_Smp_087810',\n",
       "        'pl_dd_Smed_v4_3431_0_1;sc_Smp_317550',\n",
       "        'pl_dd_Smed_v4_3155_0_1;sc_Smp_098330',\n",
       "        'pl_dd_Smed_v4_970_0_1;sc_Smp_078520',\n",
       "        'pl_dd_Smed_v4_1038_0_1;sc_Smp_075210',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_1056_0_1;sc_Smp_331920',\n",
       "        'pl_dd_Smed_v4_5677_0_1;sc_Smp_055310',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_044160',\n",
       "        'pl_dd_Smed_v4_8919_0_1;sc_Smp_169140',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_2680_0_1;sc_Smp_340600',\n",
       "        'pl_dd_Smed_v4_5156_0_1;sc_Smp_301290',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_4701_0_1;sc_Smp_179010',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_1535_0_1;sc_Smp_103710',\n",
       "        'pl_dd_Smed_v4_721_0_1;sc_Smp_006000',\n",
       "        'pl_dd_Smed_v4_860_0_1;sc_Smp_005260',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_1484_0_1;sc_Smp_337740',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_8630_0_1;sc_Smp_074610',\n",
       "        'pl_dd_Smed_v4_7227_0_1;sc_Smp_148110',\n",
       "        'pl_dd_Smed_v4_995_0_1;sc_Smp_008500',\n",
       "        'pl_dd_Smed_v4_12789_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_9007_0_1;sc_Smp_176710',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_032580',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_5890_0_1;sc_Smp_140950',\n",
       "        'pl_dd_Smed_v4_998_0_1;sc_Smp_055960',\n",
       "        'pl_dd_Smed_v4_3431_0_1;sc_Smp_321380',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_317550',\n",
       "        'pl_dd_Smed_v4_3605_0_1;sc_Smp_078240',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_1508_0_1;sc_Smp_209050',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_592_0_1;sc_Smp_036780',\n",
       "        'pl_dd_Smed_v4_1279_0_1;sc_Smp_300810',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_15228_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_2505_0_1;sc_Smp_084740',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_044020',\n",
       "        'pl_dd_Smed_v4_872_0_1;sc_Smp_175610',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_516_0_1;sc_Smp_130170',\n",
       "        'pl_dd_Smed_v4_4240_0_1;sc_Smp_148540',\n",
       "        'pl_dd_Smed_v4_695_0_1;sc_Smp_005210',\n",
       "        'pl_dd_Smed_v4_1121_0_1;sc_Smp_101450',\n",
       "        'pl_dd_Smed_v4_1190_0_1;sc_Smp_331870',\n",
       "        'pl_dd_Smed_v4_3025_0_1;sc_Smp_129720',\n",
       "        'pl_dd_Smed_v4_3681_0_1;sc_Smp_343570',\n",
       "        'pl_dd_Smed_v4_5156_0_1;sc_Smp_210630',\n",
       "        'pl_dd_Smed_v4_1345_0_1;sc_Smp_163430',\n",
       "        'pl_dd_Smed_v4_676_0_1;sc_Smp_021070',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_903_0_1;sc_Smp_046090',\n",
       "        'pl_dd_Smed_v4_4855_0_1;sc_Smp_179650',\n",
       "        'pl_dd_Smed_v4_1336_0_1;sc_Smp_029240',\n",
       "        'pl_dd_Smed_v4_6668_0_1;sc_Smp_158910',\n",
       "        'pl_dd_Smed_v4_514_0_1;sc_Smp_335300',\n",
       "        'pl_dd_Smed_v4_9993_0_1;sc_Smp_047720',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_4005_0_1;sc_Smp_042560',\n",
       "        'pl_dd_Smed_v4_458_0_1;sc_Smp_130170',\n",
       "        'pl_dd_Smed_v4_2274_0_1;sc_Smp_069870',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_321380',\n",
       "        'pl_dd_Smed_v4_12365_0_1;sc_Smp_031720',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_3389_0_1;sc_Smp_141330',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_092280',\n",
       "        'pl_dd_Smed_v4_621_0_1;sc_Smp_099320',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_1445_0_1;sc_Smp_155740',\n",
       "        'pl_dd_Smed_v4_691_0_1;sc_Smp_010400',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_303380',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_079730',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_077180',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_031010',\n",
       "        'pl_dd_Smed_v4_3326_0_1;sc_Smp_108390',\n",
       "        'pl_dd_Smed_v4_12569_0_1;sc_Smp_103710',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_093630',\n",
       "        'pl_dd_Smed_v4_11494_0_1;sc_Smp_069760',\n",
       "        'pl_dd_Smed_v4_14121_0_1;sc_Smp_173880',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_532_0_1;sc_Smp_074610',\n",
       "        'pl_dd_Smed_v4_2086_0_1;sc_Smp_319100',\n",
       "        'pl_dd_Smed_v4_818_0_1;sc_Smp_073500',\n",
       "        'pl_dd_Smed_v4_3675_0_1;sc_Smp_006850',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_032580',\n",
       "        'pl_dd_Smed_v4_7488_0_1;sc_Smp_168800',\n",
       "        'pl_dd_Smed_v4_1982_0_1;sc_Smp_053470',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_207080',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_3781_0_1;sc_Smp_002550',\n",
       "        'pl_dd_Smed_v4_802_0_1;sc_Smp_171660',\n",
       "        'pl_dd_Smed_v4_825_0_1;sc_Smp_080980',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_047110',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_303610',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_157800',\n",
       "        'pl_dd_Smed_v4_1012_0_1;sc_Smp_054120',\n",
       "        'pl_dd_Smed_v4_1331_0_1;sc_Smp_000980',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_213800',\n",
       "        'pl_dd_Smed_v4_5865_0_1;sc_Smp_136970',\n",
       "        'pl_dd_Smed_v4_1125_0_1;sc_Smp_328230',\n",
       "        'pl_dd_Smed_v4_1228_0_1;sc_Smp_077180',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_1169_0_1;sc_Smp_079730',\n",
       "        'pl_dd_Smed_v4_1172_0_1;sc_Smp_062790',\n",
       "        'pl_dd_Smed_v4_998_0_1;sc_Smp_103710',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_2444_0_1;sc_Smp_069870',\n",
       "        'pl_dd_Smed_v4_12365_0_1;sc_Smp_337740',\n",
       "        'pl_dd_Smed_v4_12608_0_1;sc_Smp_346320',\n",
       "        'pl_dd_Smed_v4_13844_0_1;sc_Smp_089860',\n",
       "        'pl_dd_Smed_v4_7488_0_1;sc_Smp_083200',\n",
       "        'pl_dd_Smed_v4_903_0_1;sc_Smp_063420',\n",
       "        'pl_dd_Smed_v4_2720_0_1;sc_Smp_132030',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_034190',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_162710',\n",
       "        'pl_dd_Smed_v4_3605_0_1;sc_Smp_012900',\n",
       "        'pl_dd_Smed_v4_1508_0_1;sc_Smp_133820',\n",
       "        'pl_dd_Smed_v4_668_0_1;sc_Smp_072970',\n",
       "        'pl_dd_Smed_v4_1957_0_1;sc_Smp_179270',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_10784_0_1;sc_Smp_001500',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_346310',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_633_0_1;sc_Smp_013360',\n",
       "        'pl_dd_Smed_v4_4364_0_1;sc_Smp_054600',\n",
       "        'pl_dd_Smed_v4_633_0_1;sc_Smp_209050',\n",
       "        'pl_dd_Smed_v4_5889_0_1;sc_Smp_045750',\n",
       "        'pl_dd_Smed_v4_1535_0_1;sc_Smp_180170',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_1222_0_1;sc_Smp_072340',\n",
       "        'pl_dd_Smed_v4_5268_0_1;sc_Smp_077020',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_1771_0_1;sc_Smp_305800',\n",
       "        'pl_dd_Smed_v4_3849_0_1;sc_Smp_001360',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_092280',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_2139_0_1;sc_Smp_012900',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_179070',\n",
       "        'pl_dd_Smed_v4_6820_1_1;sc_Smp_078240',\n",
       "        'pl_dd_Smed_v4_2057_0_1;sc_Smp_087440',\n",
       "        'pl_dd_Smed_v4_206_0_1;sc_Smp_003970',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_1396_0_1;sc_Smp_010940',\n",
       "        'pl_dd_Smed_v4_667_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_3556_0_1;sc_Smp_129720',\n",
       "        'pl_dd_Smed_v4_19210_0_1;sc_Smp_304720',\n",
       "        'pl_dd_Smed_v4_4136_0_1;sc_Smp_076530',\n",
       "        'pl_dd_Smed_v4_721_0_1;sc_Smp_174390',\n",
       "        'pl_dd_Smed_v4_1535_0_1;sc_Smp_118220',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_237_0_1;sc_Smp_076740',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_095920',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_1459_0_1;sc_Smp_106130',\n",
       "        'pl_dd_Smed_v4_6624_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_1771_0_1;sc_Smp_074500',\n",
       "        'pl_dd_Smed_v4_5987_0_1;sc_Smp_180060',\n",
       "        'pl_dd_Smed_v4_986_0_1;sc_Smp_064640',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_526_0_1;sc_Smp_028440',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_2384_0_1;sc_Smp_197180',\n",
       "        'pl_dd_Smed_v4_1222_0_1;sc_Smp_165550',\n",
       "        'pl_dd_Smed_v4_3872_0_1;sc_Smp_043970',\n",
       "        'pl_dd_Smed_v4_174_0_1;sc_Smp_246300',\n",
       "        'pl_dd_Smed_v4_4442_0_1;sc_Smp_347190',\n",
       "        'pl_dd_Smed_v4_2417_0_1;sc_Smp_133820',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_1331_0_1;sc_Smp_100090',\n",
       "        'pl_dd_Smed_v4_2799_0_1;sc_Smp_006440',\n",
       "        'pl_dd_Smed_v4_1169_0_1;sc_Smp_002550',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_207080',\n",
       "        'pl_dd_Smed_v4_1327_0_1;sc_Smp_006440',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_012900',\n",
       "        'pl_dd_Smed_v4_1044_0_1;sc_Smp_053220',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_2060_0_1;sc_Smp_064600',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_077180',\n",
       "        'pl_dd_Smed_v4_3736_0_1;sc_Smp_020060',\n",
       "        'pl_dd_Smed_v4_1192_0_1;sc_Smp_332240',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_11258_0_1;sc_Smp_055310',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_070930',\n",
       "        'pl_dd_Smed_v4_14261_0_1;sc_Smp_247820',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_860_0_1;sc_Smp_064360',\n",
       "        'pl_dd_Smed_v4_633_0_1;sc_Smp_002550',\n",
       "        'pl_dd_Smed_v4_133_0_1;sc_Smp_209050',\n",
       "        'pl_dd_Smed_v4_7703_0_1;sc_Smp_083450',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_095920',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_212920',\n",
       "        'pl_dd_Smed_v4_424_0_1;sc_Smp_106130',\n",
       "        'pl_dd_Smed_v4_3735_0_1;sc_Smp_085480',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_6352_0_1;sc_Smp_136320',\n",
       "        'pl_dd_Smed_v4_13347_0_1;sc_Smp_074610',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_2592_0_1;sc_Smp_312360',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_146340',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_4443_0_1;sc_Smp_048760',\n",
       "        'pl_dd_Smed_v4_2065_0_1;sc_Smp_022660',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_3605_0_1;sc_Smp_067420',\n",
       "        'pl_dd_Smed_v4_6850_0_1;sc_Smp_106130',\n",
       "        'pl_dd_Smed_v4_980_0_1;sc_Smp_095150',\n",
       "        'pl_dd_Smed_v4_330_0_1;sc_Smp_330250',\n",
       "        'pl_dd_Smed_v4_4172_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_986_0_1;sc_Smp_051230',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_805_0_1;sc_Smp_172510',\n",
       "        'pl_dd_Smed_v4_3346_0_1;sc_Smp_340580',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_076230',\n",
       "        'pl_dd_Smed_v4_3246_0_1;sc_Smp_002010',\n",
       "        'pl_dd_Smed_v4_4855_0_1;sc_Smp_346320',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_1077_0_1;sc_Smp_073080',\n",
       "        'pl_dd_Smed_v4_3715_0_1;sc_Smp_050700',\n",
       "        'pl_dd_Smed_v4_1760_0_1;sc_Smp_013360',\n",
       "        'pl_dd_Smed_v4_530_0_1;sc_Smp_007450',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_099030',\n",
       "        'pl_dd_Smed_v4_1222_0_1;sc_Smp_042270',\n",
       "        'pl_dd_Smed_v4_3556_0_1;sc_Smp_064640',\n",
       "        'pl_dd_Smed_v4_879_0_1;sc_Smp_108390',\n",
       "        'pl_dd_Smed_v4_12569_0_1;sc_Smp_180170',\n",
       "        'pl_dd_Smed_v4_676_0_1;sc_Smp_087810',\n",
       "        'pl_dd_Smed_v4_7132_0_1;sc_Smp_179010',\n",
       "        'pl_dd_Smed_v4_1642_0_1;sc_Smp_078320',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_2139_0_1;sc_Smp_067420',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_211330',\n",
       "        'pl_dd_Smed_v4_1508_0_1;sc_Smp_029900',\n",
       "        'pl_dd_Smed_v4_4873_0_1;sc_Smp_000980',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_094020',\n",
       "        'pl_dd_Smed_v4_435_0_1;sc_Smp_046090',\n",
       "        'pl_dd_Smed_v4_2656_0_1;sc_Smp_011720',\n",
       "        'pl_dd_Smed_v4_4580_0_1;sc_Smp_073690',\n",
       "        'pl_dd_Smed_v4_2857_0_1;sc_Smp_098980',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_099030',\n",
       "        'pl_dd_Smed_v4_1332_0_1;sc_Smp_106130',\n",
       "        'pl_dd_Smed_v4_1000_0_1;sc_Smp_103950',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_158970',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_211330',\n",
       "        'pl_dd_Smed_v4_1675_0_1;sc_Smp_073620',\n",
       "        'pl_dd_Smed_v4_12569_0_1;sc_Smp_118220',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_093690',\n",
       "        'pl_dd_Smed_v4_3809_0_1;sc_Smp_309430',\n",
       "        'pl_dd_Smed_v4_6243_0_1;sc_Smp_002820',\n",
       "        'pl_dd_Smed_v4_421_0_1;sc_Smp_097380',\n",
       "        'pl_dd_Smed_v4_2695_0_1;sc_Smp_072370',\n",
       "        'pl_dd_Smed_v4_1222_0_1;sc_Smp_320620',\n",
       "        'pl_dd_Smed_v4_6828_0_1;sc_Smp_070760',\n",
       "        'pl_dd_Smed_v4_3539_0_1;sc_Smp_073570',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_1985_0_1;sc_Smp_247420',\n",
       "        'pl_dd_Smed_v4_998_0_1;sc_Smp_180170',\n",
       "        'pl_dd_Smed_v4_200_0_1;sc_Smp_063350',\n",
       "        'pl_dd_Smed_v4_13890_0_1;sc_Smp_015840',\n",
       "        'pl_dd_Smed_v4_2924_0_1;sc_Smp_078240',\n",
       "        'pl_dd_Smed_v4_420_0_1;sc_Smp_157300',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_211330',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_103950',\n",
       "        'pl_dd_Smed_v4_2924_0_1;sc_Smp_079730',\n",
       "        'pl_dd_Smed_v4_527_0_1;sc_Smp_063420',\n",
       "        'pl_dd_Smed_v4_1608_0_1;sc_Smp_193440',\n",
       "        'pl_dd_Smed_v4_1755_0_1;sc_Smp_337860',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_9510_0_1;sc_Smp_077590',\n",
       "        'pl_dd_Smed_v4_5312_0_1;sc_Smp_164650',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_070930',\n",
       "        'pl_dd_Smed_v4_1760_0_1;sc_Smp_002550',\n",
       "        'pl_dd_Smed_v4_1025_0_1;sc_Smp_046090',\n",
       "        'pl_dd_Smed_v4_562_0_1;sc_Smp_068420',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_212920',\n",
       "        'pl_dd_Smed_v4_3431_0_1;sc_Smp_332240',\n",
       "        'pl_dd_Smed_v4_296_0_1;sc_Smp_104660',\n",
       "        'pl_dd_Smed_v4_37_0_1;sc_Smp_063040',\n",
       "        'pl_dd_Smed_v4_2417_0_1;sc_Smp_029900',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_310770',\n",
       "        'pl_dd_Smed_v4_296_0_1;sc_Smp_210310',\n",
       "        'pl_dd_Smed_v4_525_0_1;sc_Smp_046090',\n",
       "        'pl_dd_Smed_v4_998_0_1;sc_Smp_118220',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_057020',\n",
       "        'pl_dd_Smed_v4_779_0_1;sc_Smp_067900',\n",
       "        'pl_dd_Smed_v4_856_0_1;sc_Smp_092750',\n",
       "        'pl_dd_Smed_v4_1541_0_1;sc_Smp_067890',\n",
       "        'pl_dd_Smed_v4_330_0_1;sc_Smp_129070',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_5764_0_1;sc_Smp_181870',\n",
       "        'pl_dd_Smed_v4_141_0_1;sc_Smp_008545',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_667_0_1;sc_Smp_055960',\n",
       "        'pl_dd_Smed_v4_1181_0_1;sc_Smp_193440',\n",
       "        'pl_dd_Smed_v4_3866_0_1;sc_Smp_317550',\n",
       "        'pl_dd_Smed_v4_530_0_1;sc_Smp_006560',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_1359_0_1;sc_Smp_092380',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_7612_0_1;sc_Smp_022660',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_076230',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_4126_0_1;sc_Smp_090150',\n",
       "        'pl_dd_Smed_v4_6178_0_1;sc_Smp_080730',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_4830_0_1;sc_Smp_046750',\n",
       "        'pl_dd_Smed_v4_133_0_1;sc_Smp_133820',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_113620',\n",
       "        'pl_dd_Smed_v4_903_0_1;sc_Smp_004600',\n",
       "        'pl_dd_Smed_v4_2118_0_1;sc_Smp_200380',\n",
       "        'pl_dd_Smed_v4_1279_0_1;sc_Smp_072340',\n",
       "        'pl_dd_Smed_v4_3867_0_1;sc_Smp_067090',\n",
       "        'pl_dd_Smed_v4_4873_0_1;sc_Smp_100090',\n",
       "        'pl_dd_Smed_v4_4139_0_1;sc_Smp_166430',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_033710',\n",
       "        'pl_dd_Smed_v4_6624_0_1;sc_Smp_055960',\n",
       "        'pl_dd_Smed_v4_4830_0_1;sc_Smp_151180',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_435_0_1;sc_Smp_063420',\n",
       "        'pl_dd_Smed_v4_2647_0_1;sc_Smp_186560',\n",
       "        'pl_dd_Smed_v4_732_0_1;sc_Smp_213360',\n",
       "        'pl_dd_Smed_v4_11724_0_1;sc_Smp_000190',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_3556_0_1;sc_Smp_012900',\n",
       "        'pl_dd_Smed_v4_8206_0_1;sc_Smp_012480',\n",
       "        'pl_dd_Smed_v4_330_0_1;sc_Smp_084940',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_6148_0_1;sc_Smp_194870',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_057020',\n",
       "        'pl_dd_Smed_v4_6452_0_1;sc_Smp_093100',\n",
       "        'pl_dd_Smed_v4_5055_0_1;sc_Smp_241860',\n",
       "        'pl_dd_Smed_v4_206_0_1;sc_Smp_030690',\n",
       "        'pl_dd_Smed_v4_986_0_1;sc_Smp_051300',\n",
       "        'pl_dd_Smed_v4_892_0_1;sc_Smp_070130',\n",
       "        'pl_dd_Smed_v4_276_0_1;sc_Smp_136090',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_332240',\n",
       "        'pl_dd_Smed_v4_265_0_1;sc_Smp_045510',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_12125_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_1220_0_1;sc_Smp_210570',\n",
       "        'pl_dd_Smed_v4_2395_0_1;sc_Smp_094630',\n",
       "        'pl_dd_Smed_v4_1909_1_1;sc_Smp_056640',\n",
       "        'pl_dd_Smed_v4_5622_0_1;sc_Smp_025340',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_050380',\n",
       "        'pl_dd_Smed_v4_231_0_1;sc_Smp_019050',\n",
       "        'pl_dd_Smed_v4_1059_0_1;sc_Smp_153950',\n",
       "        'pl_dd_Smed_v4_1353_0_1;sc_Smp_105960',\n",
       "        'pl_dd_Smed_v4_3866_0_1;sc_Smp_321380',\n",
       "        'pl_dd_Smed_v4_741_0_1;sc_Smp_205570',\n",
       "        'pl_dd_Smed_v4_15527_0_1;sc_Smp_342520',\n",
       "        'pl_dd_Smed_v4_2145_0_1;sc_Smp_106130',\n",
       "        'pl_dd_Smed_v4_3092_0_1;sc_Smp_093100',\n",
       "        'pl_dd_Smed_v4_1276_0_1;sc_Smp_336360',\n",
       "        'pl_dd_Smed_v4_2383_0_1;sc_Smp_179270',\n",
       "        'pl_dd_Smed_v4_469_0_1;sc_Smp_031720',\n",
       "        'pl_dd_Smed_v4_1594_0_1;sc_Smp_328230',\n",
       "        'pl_dd_Smed_v4_1839_0_1;sc_Smp_133110',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_095920',\n",
       "        'pl_dd_Smed_v4_805_0_1;sc_Smp_157780',\n",
       "        'pl_dd_Smed_v4_420_0_1;sc_Smp_247600',\n",
       "        'pl_dd_Smed_v4_672_0_1;sc_Smp_004780',\n",
       "        'pl_dd_Smed_v4_1763_0_1;sc_Smp_006780',\n",
       "        'pl_dd_Smed_v4_4969_0_1;sc_Smp_048430',\n",
       "        'pl_dd_Smed_v4_1025_0_1;sc_Smp_063420',\n",
       "        'pl_dd_Smed_v4_2118_0_1;sc_Smp_340370',\n",
       "        'pl_dd_Smed_v4_4172_0_1;sc_Smp_171370',\n",
       "        'pl_dd_Smed_v4_1235_0_1;sc_Smp_032580',\n",
       "        'pl_dd_Smed_v4_14547_0_1;sc_Smp_012480',\n",
       "        'pl_dd_Smed_v4_525_0_1;sc_Smp_063420',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_332290',\n",
       "        'pl_dd_Smed_v4_6813_0_1;sc_Smp_085480',\n",
       "        'pl_dd_Smed_v4_147_0_1;sc_Smp_096010',\n",
       "        'pl_dd_Smed_v4_330_0_1;sc_Smp_079430',\n",
       "        'pl_dd_Smed_v4_2558_1_1;sc_Smp_088250',\n",
       "        'pl_dd_Smed_v4_2546_0_1;sc_Smp_058650',\n",
       "        'pl_dd_Smed_v4_695_0_1;sc_Smp_091990',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_1695_0_1;sc_Smp_328180',\n",
       "        'pl_dd_Smed_v4_3186_0_1;sc_Smp_045040',\n",
       "        'pl_dd_Smed_v4_360_0_1;sc_Smp_307540',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_058790',\n",
       "        'pl_dd_Smed_v4_1986_0_1;sc_Smp_072970',\n",
       "        'pl_dd_Smed_v4_3025_0_1;sc_Smp_019690',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_1641_0_1;sc_Smp_078640',\n",
       "        'pl_dd_Smed_v4_1788_0_1;sc_Smp_076880',\n",
       "        'pl_dd_Smed_v4_5055_0_1;sc_Smp_022660',\n",
       "        'pl_dd_Smed_v4_178_0_1;sc_Smp_053130',\n",
       "        'pl_dd_Smed_v4_171_0_1;sc_Smp_179180',\n",
       "        'pl_dd_Smed_v4_308_0_1;sc_Smp_083410',\n",
       "        'pl_dd_Smed_v4_1154_0_1;sc_Smp_067890',\n",
       "        'pl_dd_Smed_v4_12569_0_1;sc_Smp_082050',\n",
       "        'pl_dd_Smed_v4_3155_0_1;sc_Smp_104660',\n",
       "        'pl_dd_Smed_v4_527_0_1;sc_Smp_004990',\n",
       "        'pl_dd_Smed_v4_5027_0_1;sc_Smp_326430',\n",
       "        'pl_dd_Smed_v4_3155_0_1;sc_Smp_210310',\n",
       "        'pl_dd_Smed_v4_611_0_1;sc_Smp_066760',\n",
       "        'pl_dd_Smed_v4_1993_0_1;sc_Smp_329680',\n",
       "        'pl_dd_Smed_v4_2547_0_1;sc_Smp_132690',\n",
       "        'pl_dd_Smed_v4_420_0_1;sc_Smp_018290',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_1872_0_1;sc_Smp_162310',\n",
       "        'pl_dd_Smed_v4_1036_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_986_0_1;sc_Smp_113620',\n",
       "        'pl_dd_Smed_v4_1279_0_1;sc_Smp_042270',\n",
       "        'pl_dd_Smed_v4_1390_0_1;sc_Smp_020800',\n",
       "        'pl_dd_Smed_v4_2799_0_1;sc_Smp_036590',\n",
       "        'pl_dd_Smed_v4_153_0_1;sc_Smp_051010',\n",
       "        'pl_dd_Smed_v4_2547_0_1;sc_Smp_064050',\n",
       "        'pl_dd_Smed_v4_648_0_1;sc_Smp_078040',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_2592_0_1;sc_Smp_024140',\n",
       "        'pl_dd_Smed_v4_3781_0_1;sc_Smp_045950',\n",
       "        'pl_dd_Smed_v4_469_0_1;sc_Smp_337740',\n",
       "        'pl_dd_Smed_v4_3556_0_1;sc_Smp_067420',\n",
       "        'pl_dd_Smed_v4_986_0_1;sc_Smp_307050',\n",
       "        'pl_dd_Smed_v4_3490_0_1;sc_Smp_132690',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_998_0_1;sc_Smp_082050',\n",
       "        'pl_dd_Smed_v4_1336_0_1;sc_Smp_337860',\n",
       "        'pl_dd_Smed_v4_6624_0_1;sc_Smp_103710',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_131770',\n",
       "        'pl_dd_Smed_v4_174_0_1;sc_Smp_046750',\n",
       "        'pl_dd_Smed_v4_420_0_1;sc_Smp_335730',\n",
       "        'pl_dd_Smed_v4_1618_0_1;sc_Smp_113620',\n",
       "        'pl_dd_Smed_v4_2440_0_1;sc_Smp_157680',\n",
       "        'pl_dd_Smed_v4_276_0_1;sc_Smp_330330',\n",
       "        'pl_dd_Smed_v4_133_0_1;sc_Smp_029900',\n",
       "        'pl_dd_Smed_v4_1211_0_1;sc_Smp_315490',\n",
       "        'pl_dd_Smed_v4_1508_0_1;sc_Smp_094810',\n",
       "        'pl_dd_Smed_v4_1235_0_1;sc_Smp_092280',\n",
       "        'pl_dd_Smed_v4_1094_0_1;sc_Smp_210380',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_097660',\n",
       "        'pl_dd_Smed_v4_3490_0_1;sc_Smp_064050',\n",
       "        'pl_dd_Smed_v4_737_0_1;sc_Smp_211330',\n",
       "        'pl_dd_Smed_v4_2312_0_1;sc_Smp_320630',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_020200',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_3836_0_1;sc_Smp_102840',\n",
       "        'pl_dd_Smed_v4_1986_0_1;sc_Smp_150320',\n",
       "        'pl_dd_Smed_v4_201_0_1;sc_Smp_130850',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_457_0_1;sc_Smp_209050',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_127120',\n",
       "        'pl_dd_Smed_v4_12569_0_1;sc_Smp_083030',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_2126_0_1;sc_Smp_213550',\n",
       "        'pl_dd_Smed_v4_45208_0_1;sc_Smp_071890',\n",
       "        'pl_dd_Smed_v4_435_0_1;sc_Smp_004990',\n",
       "        'pl_dd_Smed_v4_1459_0_1;sc_Smp_303400',\n",
       "        'pl_dd_Smed_v4_1279_0_1;sc_Smp_320620',\n",
       "        'pl_dd_Smed_v4_9839_0_1;sc_Smp_145800',\n",
       "        'pl_dd_Smed_v4_2211_0_1;sc_Smp_072640',\n",
       "        'pl_dd_Smed_v4_1681_0_1;sc_Smp_019440',\n",
       "        'pl_dd_Smed_v4_7968_0_1;sc_Smp_053780',\n",
       "        'pl_dd_Smed_v4_1278_0_1;sc_Smp_032580',\n",
       "        'pl_dd_Smed_v4_1459_0_1;sc_Smp_069130',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_131770',\n",
       "        'pl_dd_Smed_v4_93_0_1;sc_Smp_342620',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_403_0_1;sc_Smp_130170',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_93_0_1;sc_Smp_103990',\n",
       "        'pl_dd_Smed_v4_1264_0_1;sc_Smp_074500',\n",
       "        'pl_dd_Smed_v4_349_0_1;sc_Smp_333940',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_10618_0_1;sc_Smp_017010',\n",
       "        'pl_dd_Smed_v4_3094_0_1;sc_Smp_037290',\n",
       "        'pl_dd_Smed_v4_113_0_1;sc_Smp_050280',\n",
       "        'pl_dd_Smed_v4_4073_0_1;sc_Smp_045950',\n",
       "        'pl_dd_Smed_v4_19710_0_1;sc_Smp_062020',\n",
       "        'pl_dd_Smed_v4_227_0_1;sc_Smp_032760',\n",
       "        'pl_dd_Smed_v4_52_0_1;sc_Smp_097750',\n",
       "        'pl_dd_Smed_v4_794_0_1;sc_Smp_127120',\n",
       "        'pl_dd_Smed_v4_2469_0_1;sc_Smp_024360',\n",
       "        'pl_dd_Smed_v4_334_0_1;sc_Smp_024880',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_131770',\n",
       "        'pl_dd_Smed_v4_1760_0_1;sc_Smp_113620',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_2220_0_1;sc_Smp_099300',\n",
       "        'pl_dd_Smed_v4_2799_0_1;sc_Smp_138050',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_1235_0_1;sc_Smp_207080',\n",
       "        'pl_dd_Smed_v4_4969_0_1;sc_Smp_046740',\n",
       "        'pl_dd_Smed_v4_1377_0_1;sc_Smp_267310',\n",
       "        'pl_dd_Smed_v4_15869_0_1;sc_Smp_172700',\n",
       "        'pl_dd_Smed_v4_151_0_1;sc_Smp_347110',\n",
       "        'pl_dd_Smed_v4_2522_0_1;sc_Smp_246230',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_095920',\n",
       "        'pl_dd_Smed_v4_3781_0_1;sc_Smp_050920',\n",
       "        'pl_dd_Smed_v4_6347_0_1;sc_Smp_341370',\n",
       "        'pl_dd_Smed_v4_1167_0_1;sc_Smp_243210',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_127120',\n",
       "        'pl_dd_Smed_v4_1887_0_1;sc_Smp_033970',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_131360',\n",
       "        'pl_dd_Smed_v4_172_0_1;sc_Smp_201220',\n",
       "        'pl_dd_Smed_v4_424_0_1;sc_Smp_303400',\n",
       "        'pl_dd_Smed_v4_3089_0_1;sc_Smp_000980',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_967_0_1;sc_Smp_093320',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_141580',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_424_0_1;sc_Smp_069130',\n",
       "        'pl_dd_Smed_v4_7507_0_1;sc_Smp_064150',\n",
       "        'pl_dd_Smed_v4_1059_0_1;sc_Smp_001830',\n",
       "        'pl_dd_Smed_v4_1025_0_1;sc_Smp_004990',\n",
       "        'pl_dd_Smed_v4_6297_0_1;sc_Smp_145620',\n",
       "        'pl_dd_Smed_v4_2417_0_1;sc_Smp_094810',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_099030',\n",
       "        'pl_dd_Smed_v4_6803_0_1;sc_Smp_335340',\n",
       "        'pl_dd_Smed_v4_622_0_1;sc_Smp_038680',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_090440',\n",
       "        'pl_dd_Smed_v4_525_0_1;sc_Smp_004990',\n",
       "        'pl_dd_Smed_v4_325_0_1;sc_Smp_130170',\n",
       "        'pl_dd_Smed_v4_6850_0_1;sc_Smp_303400',\n",
       "        'pl_dd_Smed_v4_1327_0_1;sc_Smp_051530',\n",
       "        'pl_dd_Smed_v4_9065_0_1;sc_Smp_336830',\n",
       "        'pl_dd_Smed_v4_516_0_1;sc_Smp_089430',\n",
       "        'pl_dd_Smed_v4_186_0_1;sc_Smp_096450',\n",
       "        'pl_dd_Smed_v4_9064_0_1;sc_Smp_309670',\n",
       "        'pl_dd_Smed_v4_6850_0_1;sc_Smp_069130',\n",
       "        'pl_dd_Smed_v4_1771_0_1;sc_Smp_121430',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_095920',\n",
       "        'pl_dd_Smed_v4_1600_0_1;sc_Smp_176230',\n",
       "        'pl_dd_Smed_v4_2977_0_1;sc_Smp_154690',\n",
       "        'pl_dd_Smed_v4_4994_0_1;sc_Smp_163030',\n",
       "        'pl_dd_Smed_v4_2199_0_1;sc_Smp_015720',\n",
       "        'pl_dd_Smed_v4_3382_0_1;sc_Smp_046980',\n",
       "        'pl_dd_Smed_v4_2924_0_1;sc_Smp_007630',\n",
       "        'pl_dd_Smed_v4_276_0_1;sc_Smp_101450',\n",
       "        'pl_dd_Smed_v4_903_0_1;sc_Smp_017360',\n",
       "        'pl_dd_Smed_v4_360_0_1;sc_Smp_307550',\n",
       "        'pl_dd_Smed_v4_8119_0_1;sc_Smp_172850',\n",
       "        'pl_dd_Smed_v4_1115_0_1;sc_Smp_335300',\n",
       "        'pl_dd_Smed_v4_3243_0_1;sc_Smp_102450',\n",
       "        'pl_dd_Smed_v4_16174_0_1;sc_Smp_064320',\n",
       "        'pl_dd_Smed_v4_1278_0_1;sc_Smp_092280',\n",
       "        'pl_dd_Smed_v4_2747_0_1;sc_Smp_034700',\n",
       "        'pl_dd_Smed_v4_2082_0_1;sc_Smp_131360',\n",
       "        'pl_dd_Smed_v4_1580_0_1;sc_Smp_002080',\n",
       "        'pl_dd_Smed_v4_1763_0_1;sc_Smp_307410',\n",
       "        'pl_dd_Smed_v4_1642_0_1;sc_Smp_026630',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_11265_0_1;sc_Smp_103950',\n",
       "        'pl_dd_Smed_v4_16174_0_1;sc_Smp_019260',\n",
       "        'pl_dd_Smed_v4_172_0_1;sc_Smp_335710',\n",
       "        'pl_dd_Smed_v4_1273_0_1;sc_Smp_003970',\n",
       "        'pl_dd_Smed_v4_4073_0_1;sc_Smp_050920',\n",
       "        'pl_dd_Smed_v4_5419_0_1;sc_Smp_002410',\n",
       "        'pl_dd_Smed_v4_1285_0_1;sc_Smp_314650',\n",
       "        'pl_dd_Smed_v4_6177_0_1;sc_Smp_102490',\n",
       "        'pl_dd_Smed_v4_1750_0_1;sc_Smp_042090',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_123410',\n",
       "        'pl_dd_Smed_v4_1466_0_1;sc_Smp_318980',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_156_0_1;sc_Smp_340900',\n",
       "        'pl_dd_Smed_v4_335_0_1;sc_Smp_204770',\n",
       "        'pl_dd_Smed_v4_1302_0_1;sc_Smp_070340',\n",
       "        'pl_dd_Smed_v4_2325_0_1;sc_Smp_105860',\n",
       "        'pl_dd_Smed_v4_2592_0_1;sc_Smp_176810',\n",
       "        'pl_dd_Smed_v4_1332_0_1;sc_Smp_303400',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_013790',\n",
       "        'pl_dd_Smed_v4_458_0_1;sc_Smp_089430',\n",
       "        'pl_dd_Smed_v4_93_0_1;sc_Smp_129720',\n",
       "        'pl_dd_Smed_v4_1924_0_1;sc_Smp_163300',\n",
       "        'pl_dd_Smed_v4_435_0_1;sc_Smp_004600',\n",
       "        'pl_dd_Smed_v4_1828_0_1;sc_Smp_037120',\n",
       "        'pl_dd_Smed_v4_1332_0_1;sc_Smp_069130',\n",
       "        'pl_dd_Smed_v4_724_0_1;sc_Smp_133120',\n",
       "        'pl_dd_Smed_v4_5375_0_1;sc_Smp_311050',\n",
       "        'pl_dd_Smed_v4_2144_0_1;sc_Smp_015860',\n",
       "        'pl_dd_Smed_v4_2807_0_1;sc_Smp_074260',\n",
       "        'pl_dd_Smed_v4_294_0_1;sc_Smp_006320',\n",
       "        'pl_dd_Smed_v4_1045_0_1;sc_Smp_003970',\n",
       "        'pl_dd_Smed_v4_6820_1_1;sc_Smp_043100',\n",
       "        'pl_dd_Smed_v4_3250_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_986_0_1;sc_Smp_043100',\n",
       "        'pl_dd_Smed_v4_7797_0_1;sc_Smp_172030',\n",
       "        'pl_dd_Smed_v4_3816_0_1;sc_Smp_154320',\n",
       "        'pl_dd_Smed_v4_4273_0_1;sc_Smp_181490',\n",
       "        'pl_dd_Smed_v4_4061_0_1;sc_Smp_093100',\n",
       "        'pl_dd_Smed_v4_1887_0_1;sc_Smp_016200',\n",
       "        'pl_dd_Smed_v4_268_0_1;sc_Smp_060090',\n",
       "        'pl_dd_Smed_v4_318_0_1;sc_Smp_061200',\n",
       "        'pl_dd_Smed_v4_799_0_1;sc_Smp_163110',\n",
       "        'pl_dd_Smed_v4_50_0_1;sc_Smp_267120',\n",
       "        'pl_dd_Smed_v4_1710_0_1;sc_Smp_016130',\n",
       "        'pl_dd_Smed_v4_1566_1_1;sc_Smp_314650',\n",
       "        'pl_dd_Smed_v4_596_0_1;sc_Smp_211330',\n",
       "        'pl_dd_Smed_v4_836_0_1;sc_Smp_060180',\n",
       "        'pl_dd_Smed_v4_5257_0_1;sc_Smp_045950',\n",
       "        'pl_dd_Smed_v4_5677_0_1;sc_Smp_144180',\n",
       "        'pl_dd_Smed_v4_4035_0_1;sc_Smp_090920'], dtype='<U37'),\n",
       " array(['pl_dd_Smed_v4_659_0_1', 'pl_dd_Smed_v4_2787_0_1',\n",
       "        'pl_dd_Smed_v4_648_0_1', 'pl_dd_Smed_v4_2614_0_1',\n",
       "        'pl_dd_Smed_v4_15019_0_1', 'pl_dd_Smed_v4_5764_0_1',\n",
       "        'pl_dd_Smed_v4_7837_0_1', 'pl_dd_Smed_v4_1484_0_1',\n",
       "        'pl_dd_Smed_v4_10594_0_1', 'pl_dd_Smed_v4_4273_0_1',\n",
       "        'pl_dd_Smed_v4_3708_0_1', 'pl_dd_Smed_v4_11487_0_1',\n",
       "        'pl_dd_Smed_v4_756_0_1', 'pl_dd_Smed_v4_12365_0_1',\n",
       "        'pl_dd_Smed_v4_4712_0_1', 'pl_dd_Smed_v4_1651_0_1',\n",
       "        'pl_dd_Smed_v4_1920_0_1', 'pl_dd_Smed_v4_2682_0_1',\n",
       "        'pl_dd_Smed_v4_2286_0_1', 'pl_dd_Smed_v4_5956_0_1',\n",
       "        'pl_dd_Smed_v4_469_0_1', 'pl_dd_Smed_v4_1258_0_1',\n",
       "        'pl_dd_Smed_v4_13700_0_1', 'pl_dd_Smed_v4_8630_0_1',\n",
       "        'pl_dd_Smed_v4_8626_0_1', 'pl_dd_Smed_v4_601_0_1',\n",
       "        'pl_dd_Smed_v4_6734_0_1', 'pl_dd_Smed_v4_1000_0_1',\n",
       "        'pl_dd_Smed_v4_9065_0_1', 'pl_dd_Smed_v4_12789_0_1',\n",
       "        'pl_dd_Smed_v4_5862_0_1', 'pl_dd_Smed_v4_532_0_1',\n",
       "        'pl_dd_Smed_v4_11717_0_1', 'pl_dd_Smed_v4_5663_0_1',\n",
       "        'pl_dd_Smed_v4_2921_0_1', 'pl_dd_Smed_v4_13580_0_1',\n",
       "        'pl_dd_Smed_v4_9585_0_1', 'pl_dd_Smed_v4_10335_0_1',\n",
       "        'pl_dd_Smed_v4_13347_0_1', 'pl_dd_Smed_v4_1985_0_1',\n",
       "        'pl_dd_Smed_v4_2312_0_1', 'pl_dd_Smed_v4_9628_0_1',\n",
       "        'pl_dd_Smed_v4_13972_0_1', 'pl_dd_Smed_v4_7575_0_1',\n",
       "        'pl_dd_Smed_v4_10217_0_1', 'pl_dd_Smed_v4_11265_0_1',\n",
       "        'pl_dd_Smed_v4_928_0_1', 'pl_dd_Smed_v4_1194_0_1',\n",
       "        'pl_dd_Smed_v4_152_0_1', 'pl_dd_Smed_v4_10338_0_1',\n",
       "        'pl_dd_Smed_v4_12125_0_1', 'pl_dd_Smed_v4_1228_0_1',\n",
       "        'pl_dd_Smed_v4_2592_0_1', 'pl_dd_Smed_v4_12599_0_1',\n",
       "        'pl_dd_Smed_v4_6154_0_1', 'pl_dd_Smed_v4_7227_0_1',\n",
       "        'pl_dd_Smed_v4_9624_0_1', 'pl_dd_Smed_v4_1260_0_1',\n",
       "        'pl_dd_Smed_v4_3061_0_1', 'pl_dd_Smed_v4_15869_0_1',\n",
       "        'pl_dd_Smed_v4_13269_0_1', 'pl_dd_Smed_v4_8732_0_1',\n",
       "        'pl_dd_Smed_v4_13844_0_1', 'pl_dd_Smed_v4_5816_0_1',\n",
       "        'pl_dd_Smed_v4_12818_0_1', 'pl_dd_Smed_v4_5120_0_1',\n",
       "        'pl_dd_Smed_v4_5865_0_1', 'pl_dd_Smed_v4_2595_0_1',\n",
       "        'pl_dd_Smed_v4_3904_0_1', 'pl_dd_Smed_v4_976_0_1',\n",
       "        'pl_dd_Smed_v4_1302_0_1', 'pl_dd_Smed_v4_1646_0_1',\n",
       "        'pl_dd_Smed_v4_8206_0_1', 'pl_dd_Smed_v4_4994_0_1',\n",
       "        'pl_dd_Smed_v4_5156_0_1', 'pl_dd_Smed_v4_2811_0_1',\n",
       "        'pl_dd_Smed_v4_1204_0_1', 'pl_dd_Smed_v4_13140_0_1',\n",
       "        'pl_dd_Smed_v4_3342_0_1', 'pl_dd_Smed_v4_2016_0_1',\n",
       "        'pl_dd_Smed_v4_16174_0_1', 'pl_dd_Smed_v4_14547_0_1',\n",
       "        'pl_dd_Smed_v4_2297_0_1', 'pl_dd_Smed_v4_900_0_1',\n",
       "        'pl_dd_Smed_v4_2065_0_1', 'pl_dd_Smed_v4_3776_0_1',\n",
       "        'pl_dd_Smed_v4_4341_0_1', 'pl_dd_Smed_v4_7808_0_1',\n",
       "        'pl_dd_Smed_v4_11417_0_1', 'pl_dd_Smed_v4_9695_0_1',\n",
       "        'pl_dd_Smed_v4_673_0_1', 'pl_dd_Smed_v4_7507_0_1',\n",
       "        'pl_dd_Smed_v4_11558_0_1', 'pl_dd_Smed_v4_4808_0_1',\n",
       "        'pl_dd_Smed_v4_1807_0_1', 'pl_dd_Smed_v4_1089_0_1',\n",
       "        'pl_dd_Smed_v4_10019_0_1', 'pl_dd_Smed_v4_4087_0_1',\n",
       "        'pl_dd_Smed_v4_799_0_1', 'pl_dd_Smed_v4_802_0_1',\n",
       "        'pl_dd_Smed_v4_5611_0_1', 'pl_dd_Smed_v4_7612_0_1',\n",
       "        'pl_dd_Smed_v4_794_0_1', 'pl_dd_Smed_v4_3239_0_1',\n",
       "        'pl_dd_Smed_v4_1982_0_1', 'pl_dd_Smed_v4_2082_0_1',\n",
       "        'pl_dd_Smed_v4_6075_0_1', 'pl_dd_Smed_v4_5140_0_1',\n",
       "        'pl_dd_Smed_v4_2086_0_1', 'pl_dd_Smed_v4_1095_0_1',\n",
       "        'pl_dd_Smed_v4_904_0_1', 'pl_dd_Smed_v4_1535_0_1',\n",
       "        'pl_dd_Smed_v4_1609_0_1', 'pl_dd_Smed_v4_5055_0_1',\n",
       "        'pl_dd_Smed_v4_4679_0_1', 'pl_dd_Smed_v4_6009_0_1',\n",
       "        'pl_dd_Smed_v4_6610_0_1', 'pl_dd_Smed_v4_4379_0_1',\n",
       "        'pl_dd_Smed_v4_1038_0_1', 'pl_dd_Smed_v4_4471_0_1',\n",
       "        'pl_dd_Smed_v4_1134_0_1', 'pl_dd_Smed_v4_1516_0_1',\n",
       "        'pl_dd_Smed_v4_11475_0_1', 'pl_dd_Smed_v4_5644_0_1',\n",
       "        'pl_dd_Smed_v4_6301_0_1', 'pl_dd_Smed_v4_1983_0_1',\n",
       "        'pl_dd_Smed_v4_13890_0_1', 'pl_dd_Smed_v4_10739_0_1',\n",
       "        'pl_dd_Smed_v4_1416_0_1', 'pl_dd_Smed_v4_1928_0_1',\n",
       "        'pl_dd_Smed_v4_8634_0_1', 'pl_dd_Smed_v4_3517_0_1',\n",
       "        'pl_dd_Smed_v4_1682_0_1', 'pl_dd_Smed_v4_12569_0_1',\n",
       "        'pl_dd_Smed_v4_764_0_1', 'pl_dd_Smed_v4_737_0_1',\n",
       "        'pl_dd_Smed_v4_1222_0_1', 'pl_dd_Smed_v4_887_0_1',\n",
       "        'pl_dd_Smed_v4_1056_0_1', 'pl_dd_Smed_v4_13727_0_1',\n",
       "        'pl_dd_Smed_v4_4362_0_1', 'pl_dd_Smed_v4_443_0_1',\n",
       "        'pl_dd_Smed_v4_997_0_1', 'pl_dd_Smed_v4_998_0_1',\n",
       "        'pl_dd_Smed_v4_565_0_1', 'pl_dd_Smed_v4_1192_0_1',\n",
       "        'pl_dd_Smed_v4_1121_0_1', 'pl_dd_Smed_v4_3584_0_1',\n",
       "        'pl_dd_Smed_v4_4443_0_1', 'pl_dd_Smed_v4_4263_0_1',\n",
       "        'pl_dd_Smed_v4_2166_0_1', 'pl_dd_Smed_v4_1253_0_1',\n",
       "        'pl_dd_Smed_v4_420_0_1', 'pl_dd_Smed_v4_1755_0_1',\n",
       "        'pl_dd_Smed_v4_449_0_1', 'pl_dd_Smed_v4_8491_0_1',\n",
       "        'pl_dd_Smed_v4_5677_0_1', 'pl_dd_Smed_v4_560_0_1',\n",
       "        'pl_dd_Smed_v4_12608_0_1', 'pl_dd_Smed_v4_5573_0_1',\n",
       "        'pl_dd_Smed_v4_1801_0_1', 'pl_dd_Smed_v4_5312_0_1',\n",
       "        'pl_dd_Smed_v4_2102_0_1', 'pl_dd_Smed_v4_596_0_1',\n",
       "        'pl_dd_Smed_v4_8720_0_1', 'pl_dd_Smed_v4_805_0_1',\n",
       "        'pl_dd_Smed_v4_3731_0_1', 'pl_dd_Smed_v4_3431_0_1',\n",
       "        'pl_dd_Smed_v4_3155_0_1', 'pl_dd_Smed_v4_970_0_1',\n",
       "        'pl_dd_Smed_v4_8919_0_1', 'pl_dd_Smed_v4_3250_0_1',\n",
       "        'pl_dd_Smed_v4_3816_0_1', 'pl_dd_Smed_v4_2680_0_1',\n",
       "        'pl_dd_Smed_v4_4701_0_1', 'pl_dd_Smed_v4_721_0_1',\n",
       "        'pl_dd_Smed_v4_860_0_1', 'pl_dd_Smed_v4_724_0_1',\n",
       "        'pl_dd_Smed_v4_995_0_1', 'pl_dd_Smed_v4_9007_0_1',\n",
       "        'pl_dd_Smed_v4_1541_0_1', 'pl_dd_Smed_v4_5890_0_1',\n",
       "        'pl_dd_Smed_v4_1695_0_1', 'pl_dd_Smed_v4_3605_0_1',\n",
       "        'pl_dd_Smed_v4_1508_0_1', 'pl_dd_Smed_v4_592_0_1',\n",
       "        'pl_dd_Smed_v4_1279_0_1', 'pl_dd_Smed_v4_15228_0_1',\n",
       "        'pl_dd_Smed_v4_2505_0_1', 'pl_dd_Smed_v4_872_0_1',\n",
       "        'pl_dd_Smed_v4_516_0_1', 'pl_dd_Smed_v4_4240_0_1',\n",
       "        'pl_dd_Smed_v4_695_0_1', 'pl_dd_Smed_v4_1190_0_1',\n",
       "        'pl_dd_Smed_v4_3025_0_1', 'pl_dd_Smed_v4_3681_0_1',\n",
       "        'pl_dd_Smed_v4_1345_0_1', 'pl_dd_Smed_v4_676_0_1',\n",
       "        'pl_dd_Smed_v4_903_0_1', 'pl_dd_Smed_v4_4855_0_1',\n",
       "        'pl_dd_Smed_v4_1336_0_1', 'pl_dd_Smed_v4_6668_0_1',\n",
       "        'pl_dd_Smed_v4_514_0_1', 'pl_dd_Smed_v4_9993_0_1',\n",
       "        'pl_dd_Smed_v4_4005_0_1', 'pl_dd_Smed_v4_458_0_1',\n",
       "        'pl_dd_Smed_v4_2274_0_1', 'pl_dd_Smed_v4_3389_0_1',\n",
       "        'pl_dd_Smed_v4_621_0_1', 'pl_dd_Smed_v4_1445_0_1',\n",
       "        'pl_dd_Smed_v4_691_0_1', 'pl_dd_Smed_v4_3326_0_1',\n",
       "        'pl_dd_Smed_v4_11494_0_1', 'pl_dd_Smed_v4_14121_0_1',\n",
       "        'pl_dd_Smed_v4_818_0_1', 'pl_dd_Smed_v4_3675_0_1',\n",
       "        'pl_dd_Smed_v4_1154_0_1', 'pl_dd_Smed_v4_7488_0_1',\n",
       "        'pl_dd_Smed_v4_3781_0_1', 'pl_dd_Smed_v4_825_0_1',\n",
       "        'pl_dd_Smed_v4_349_0_1', 'pl_dd_Smed_v4_1012_0_1',\n",
       "        'pl_dd_Smed_v4_1331_0_1', 'pl_dd_Smed_v4_1125_0_1',\n",
       "        'pl_dd_Smed_v4_1169_0_1', 'pl_dd_Smed_v4_1172_0_1',\n",
       "        'pl_dd_Smed_v4_2444_0_1', 'pl_dd_Smed_v4_2720_0_1',\n",
       "        'pl_dd_Smed_v4_668_0_1', 'pl_dd_Smed_v4_1957_0_1',\n",
       "        'pl_dd_Smed_v4_10784_0_1', 'pl_dd_Smed_v4_633_0_1',\n",
       "        'pl_dd_Smed_v4_4364_0_1', 'pl_dd_Smed_v4_5889_0_1',\n",
       "        'pl_dd_Smed_v4_5268_0_1', 'pl_dd_Smed_v4_1771_0_1',\n",
       "        'pl_dd_Smed_v4_3849_0_1', 'pl_dd_Smed_v4_2139_0_1',\n",
       "        'pl_dd_Smed_v4_6820_1_1', 'pl_dd_Smed_v4_2057_0_1',\n",
       "        'pl_dd_Smed_v4_206_0_1', 'pl_dd_Smed_v4_1396_0_1',\n",
       "        'pl_dd_Smed_v4_667_0_1', 'pl_dd_Smed_v4_3556_0_1',\n",
       "        'pl_dd_Smed_v4_19210_0_1', 'pl_dd_Smed_v4_4136_0_1',\n",
       "        'pl_dd_Smed_v4_237_0_1', 'pl_dd_Smed_v4_1459_0_1',\n",
       "        'pl_dd_Smed_v4_6624_0_1', 'pl_dd_Smed_v4_5987_0_1',\n",
       "        'pl_dd_Smed_v4_986_0_1', 'pl_dd_Smed_v4_526_0_1',\n",
       "        'pl_dd_Smed_v4_2384_0_1', 'pl_dd_Smed_v4_3872_0_1',\n",
       "        'pl_dd_Smed_v4_174_0_1', 'pl_dd_Smed_v4_4442_0_1',\n",
       "        'pl_dd_Smed_v4_2417_0_1', 'pl_dd_Smed_v4_2799_0_1',\n",
       "        'pl_dd_Smed_v4_1327_0_1', 'pl_dd_Smed_v4_1044_0_1',\n",
       "        'pl_dd_Smed_v4_2060_0_1', 'pl_dd_Smed_v4_7797_0_1',\n",
       "        'pl_dd_Smed_v4_3736_0_1', 'pl_dd_Smed_v4_11258_0_1',\n",
       "        'pl_dd_Smed_v4_14261_0_1', 'pl_dd_Smed_v4_133_0_1',\n",
       "        'pl_dd_Smed_v4_7703_0_1', 'pl_dd_Smed_v4_424_0_1',\n",
       "        'pl_dd_Smed_v4_3735_0_1', 'pl_dd_Smed_v4_6352_0_1',\n",
       "        'pl_dd_Smed_v4_6850_0_1', 'pl_dd_Smed_v4_980_0_1',\n",
       "        'pl_dd_Smed_v4_330_0_1', 'pl_dd_Smed_v4_4172_0_1',\n",
       "        'pl_dd_Smed_v4_3346_0_1', 'pl_dd_Smed_v4_3246_0_1',\n",
       "        'pl_dd_Smed_v4_1077_0_1', 'pl_dd_Smed_v4_3715_0_1',\n",
       "        'pl_dd_Smed_v4_1760_0_1', 'pl_dd_Smed_v4_530_0_1',\n",
       "        'pl_dd_Smed_v4_879_0_1', 'pl_dd_Smed_v4_7132_0_1',\n",
       "        'pl_dd_Smed_v4_1642_0_1', 'pl_dd_Smed_v4_4873_0_1',\n",
       "        'pl_dd_Smed_v4_435_0_1', 'pl_dd_Smed_v4_2656_0_1',\n",
       "        'pl_dd_Smed_v4_4580_0_1', 'pl_dd_Smed_v4_2857_0_1',\n",
       "        'pl_dd_Smed_v4_1332_0_1', 'pl_dd_Smed_v4_1675_0_1',\n",
       "        'pl_dd_Smed_v4_3809_0_1', 'pl_dd_Smed_v4_6243_0_1',\n",
       "        'pl_dd_Smed_v4_421_0_1', 'pl_dd_Smed_v4_2695_0_1',\n",
       "        'pl_dd_Smed_v4_6828_0_1', 'pl_dd_Smed_v4_3539_0_1',\n",
       "        'pl_dd_Smed_v4_200_0_1', 'pl_dd_Smed_v4_2924_0_1',\n",
       "        'pl_dd_Smed_v4_527_0_1', 'pl_dd_Smed_v4_1608_0_1',\n",
       "        'pl_dd_Smed_v4_9510_0_1', 'pl_dd_Smed_v4_1025_0_1',\n",
       "        'pl_dd_Smed_v4_562_0_1', 'pl_dd_Smed_v4_296_0_1',\n",
       "        'pl_dd_Smed_v4_37_0_1', 'pl_dd_Smed_v4_525_0_1',\n",
       "        'pl_dd_Smed_v4_779_0_1', 'pl_dd_Smed_v4_856_0_1',\n",
       "        'pl_dd_Smed_v4_141_0_1', 'pl_dd_Smed_v4_1181_0_1',\n",
       "        'pl_dd_Smed_v4_3866_0_1', 'pl_dd_Smed_v4_1359_0_1',\n",
       "        'pl_dd_Smed_v4_4126_0_1', 'pl_dd_Smed_v4_6178_0_1',\n",
       "        'pl_dd_Smed_v4_4830_0_1', 'pl_dd_Smed_v4_2118_0_1',\n",
       "        'pl_dd_Smed_v4_3867_0_1', 'pl_dd_Smed_v4_4139_0_1',\n",
       "        'pl_dd_Smed_v4_2647_0_1', 'pl_dd_Smed_v4_732_0_1',\n",
       "        'pl_dd_Smed_v4_11724_0_1', 'pl_dd_Smed_v4_6148_0_1',\n",
       "        'pl_dd_Smed_v4_6452_0_1', 'pl_dd_Smed_v4_892_0_1',\n",
       "        'pl_dd_Smed_v4_276_0_1', 'pl_dd_Smed_v4_265_0_1',\n",
       "        'pl_dd_Smed_v4_1220_0_1', 'pl_dd_Smed_v4_2395_0_1',\n",
       "        'pl_dd_Smed_v4_1909_1_1', 'pl_dd_Smed_v4_5622_0_1',\n",
       "        'pl_dd_Smed_v4_231_0_1', 'pl_dd_Smed_v4_1059_0_1',\n",
       "        'pl_dd_Smed_v4_1353_0_1', 'pl_dd_Smed_v4_741_0_1',\n",
       "        'pl_dd_Smed_v4_15527_0_1', 'pl_dd_Smed_v4_2145_0_1',\n",
       "        'pl_dd_Smed_v4_3092_0_1', 'pl_dd_Smed_v4_1276_0_1',\n",
       "        'pl_dd_Smed_v4_2383_0_1', 'pl_dd_Smed_v4_1594_0_1',\n",
       "        'pl_dd_Smed_v4_1839_0_1', 'pl_dd_Smed_v4_672_0_1',\n",
       "        'pl_dd_Smed_v4_1763_0_1', 'pl_dd_Smed_v4_4969_0_1',\n",
       "        'pl_dd_Smed_v4_1235_0_1', 'pl_dd_Smed_v4_6813_0_1',\n",
       "        'pl_dd_Smed_v4_147_0_1', 'pl_dd_Smed_v4_2558_1_1',\n",
       "        'pl_dd_Smed_v4_2546_0_1', 'pl_dd_Smed_v4_3186_0_1',\n",
       "        'pl_dd_Smed_v4_360_0_1', 'pl_dd_Smed_v4_1986_0_1',\n",
       "        'pl_dd_Smed_v4_1641_0_1', 'pl_dd_Smed_v4_1788_0_1',\n",
       "        'pl_dd_Smed_v4_178_0_1', 'pl_dd_Smed_v4_171_0_1',\n",
       "        'pl_dd_Smed_v4_308_0_1', 'pl_dd_Smed_v4_5027_0_1',\n",
       "        'pl_dd_Smed_v4_611_0_1', 'pl_dd_Smed_v4_1993_0_1',\n",
       "        'pl_dd_Smed_v4_2547_0_1', 'pl_dd_Smed_v4_1872_0_1',\n",
       "        'pl_dd_Smed_v4_1036_0_1', 'pl_dd_Smed_v4_1390_0_1',\n",
       "        'pl_dd_Smed_v4_153_0_1', 'pl_dd_Smed_v4_3490_0_1',\n",
       "        'pl_dd_Smed_v4_1618_0_1', 'pl_dd_Smed_v4_2440_0_1',\n",
       "        'pl_dd_Smed_v4_1211_0_1', 'pl_dd_Smed_v4_1094_0_1',\n",
       "        'pl_dd_Smed_v4_3836_0_1', 'pl_dd_Smed_v4_201_0_1',\n",
       "        'pl_dd_Smed_v4_457_0_1', 'pl_dd_Smed_v4_2126_0_1',\n",
       "        'pl_dd_Smed_v4_45208_0_1', 'pl_dd_Smed_v4_9839_0_1',\n",
       "        'pl_dd_Smed_v4_2211_0_1', 'pl_dd_Smed_v4_1681_0_1',\n",
       "        'pl_dd_Smed_v4_7968_0_1', 'pl_dd_Smed_v4_1278_0_1',\n",
       "        'pl_dd_Smed_v4_93_0_1', 'pl_dd_Smed_v4_403_0_1',\n",
       "        'pl_dd_Smed_v4_1264_0_1', 'pl_dd_Smed_v4_10618_0_1',\n",
       "        'pl_dd_Smed_v4_3094_0_1', 'pl_dd_Smed_v4_113_0_1',\n",
       "        'pl_dd_Smed_v4_4073_0_1', 'pl_dd_Smed_v4_19710_0_1',\n",
       "        'pl_dd_Smed_v4_227_0_1', 'pl_dd_Smed_v4_52_0_1',\n",
       "        'pl_dd_Smed_v4_2469_0_1', 'pl_dd_Smed_v4_334_0_1',\n",
       "        'pl_dd_Smed_v4_2220_0_1', 'pl_dd_Smed_v4_1377_0_1',\n",
       "        'pl_dd_Smed_v4_151_0_1', 'pl_dd_Smed_v4_2522_0_1',\n",
       "        'pl_dd_Smed_v4_6347_0_1', 'pl_dd_Smed_v4_1167_0_1',\n",
       "        'pl_dd_Smed_v4_1887_0_1', 'pl_dd_Smed_v4_172_0_1',\n",
       "        'pl_dd_Smed_v4_3089_0_1', 'pl_dd_Smed_v4_967_0_1',\n",
       "        'pl_dd_Smed_v4_6297_0_1', 'pl_dd_Smed_v4_6803_0_1',\n",
       "        'pl_dd_Smed_v4_622_0_1', 'pl_dd_Smed_v4_325_0_1',\n",
       "        'pl_dd_Smed_v4_186_0_1', 'pl_dd_Smed_v4_9064_0_1',\n",
       "        'pl_dd_Smed_v4_1600_0_1', 'pl_dd_Smed_v4_2977_0_1',\n",
       "        'pl_dd_Smed_v4_2199_0_1', 'pl_dd_Smed_v4_3382_0_1',\n",
       "        'pl_dd_Smed_v4_8119_0_1', 'pl_dd_Smed_v4_1115_0_1',\n",
       "        'pl_dd_Smed_v4_3243_0_1', 'pl_dd_Smed_v4_2747_0_1',\n",
       "        'pl_dd_Smed_v4_1580_0_1', 'pl_dd_Smed_v4_1273_0_1',\n",
       "        'pl_dd_Smed_v4_5419_0_1', 'pl_dd_Smed_v4_1285_0_1',\n",
       "        'pl_dd_Smed_v4_6177_0_1', 'pl_dd_Smed_v4_1750_0_1',\n",
       "        'pl_dd_Smed_v4_1466_0_1', 'pl_dd_Smed_v4_156_0_1',\n",
       "        'pl_dd_Smed_v4_335_0_1', 'pl_dd_Smed_v4_2325_0_1',\n",
       "        'pl_dd_Smed_v4_1924_0_1', 'pl_dd_Smed_v4_1828_0_1',\n",
       "        'pl_dd_Smed_v4_5375_0_1', 'pl_dd_Smed_v4_2144_0_1',\n",
       "        'pl_dd_Smed_v4_2807_0_1', 'pl_dd_Smed_v4_294_0_1',\n",
       "        'pl_dd_Smed_v4_1045_0_1', 'pl_dd_Smed_v4_4061_0_1',\n",
       "        'pl_dd_Smed_v4_268_0_1', 'pl_dd_Smed_v4_318_0_1',\n",
       "        'pl_dd_Smed_v4_50_0_1', 'pl_dd_Smed_v4_1710_0_1',\n",
       "        'pl_dd_Smed_v4_1566_1_1', 'pl_dd_Smed_v4_836_0_1',\n",
       "        'pl_dd_Smed_v4_5257_0_1', 'pl_dd_Smed_v4_4035_0_1'], dtype='<U23'),\n",
       " array(['sc_Smp_179320', 'sc_Smp_086860', 'sc_Smp_027920', 'sc_Smp_032500',\n",
       "        'sc_Smp_143490', 'sc_Smp_172530', 'sc_Smp_094140', 'sc_Smp_082490',\n",
       "        'sc_Smp_054840', 'sc_Smp_162370', 'sc_Smp_009600', 'sc_Smp_245030',\n",
       "        'sc_Smp_037590', 'sc_Smp_301290', 'sc_Smp_171370', 'sc_Smp_009030',\n",
       "        'sc_Smp_210630', 'sc_Smp_046500', 'sc_Smp_055890', 'sc_Smp_034670',\n",
       "        'sc_Smp_345010', 'sc_Smp_247600', 'sc_Smp_047620', 'sc_Smp_040020',\n",
       "        'sc_Smp_018290', 'sc_Smp_124230', 'sc_Smp_335730', 'sc_Smp_337870',\n",
       "        'sc_Smp_080730', 'sc_Smp_334160', 'sc_Smp_058320', 'sc_Smp_079240',\n",
       "        'sc_Smp_167540', 'sc_Smp_004060', 'sc_Smp_151560', 'sc_Smp_128200',\n",
       "        'sc_Smp_047110', 'sc_Smp_157800', 'sc_Smp_179710', 'sc_Smp_137390',\n",
       "        'sc_Smp_128210', 'sc_Smp_162710', 'sc_Smp_346310', 'sc_Smp_054610',\n",
       "        'sc_Smp_146340', 'sc_Smp_179650', 'sc_Smp_309960', 'sc_Smp_094020',\n",
       "        'sc_Smp_158970', 'sc_Smp_311740', 'sc_Smp_145810', 'sc_Smp_033710',\n",
       "        'sc_Smp_157300', 'sc_Smp_342620', 'sc_Smp_046910', 'sc_Smp_076600',\n",
       "        'sc_Smp_347010', 'sc_Smp_200410', 'sc_Smp_115290', 'sc_Smp_346320',\n",
       "        'sc_Smp_128870', 'sc_Smp_097660', 'sc_Smp_020200', 'sc_Smp_066260',\n",
       "        'sc_Smp_333940', 'sc_Smp_115340', 'sc_Smp_052390', 'sc_Smp_033020',\n",
       "        'sc_Smp_173700', 'sc_Smp_315010', 'sc_Smp_148930', 'sc_Smp_172700',\n",
       "        'sc_Smp_140530', 'sc_Smp_084620', 'sc_Smp_029240', 'sc_Smp_123410',\n",
       "        'sc_Smp_147970', 'sc_Smp_013790', 'sc_Smp_000980', 'sc_Smp_309950',\n",
       "        'sc_Smp_172030', 'sc_Smp_331790', 'sc_Smp_048660', 'sc_Smp_137330',\n",
       "        'sc_Smp_090440', 'sc_Smp_044280', 'sc_Smp_074390', 'sc_Smp_095920',\n",
       "        'sc_Smp_058990', 'sc_Smp_121800', 'sc_Smp_124300', 'sc_Smp_089860',\n",
       "        'sc_Smp_016410', 'sc_Smp_096360', 'sc_Smp_137090', 'sc_Smp_137660',\n",
       "        'sc_Smp_133120', 'sc_Smp_100090', 'sc_Smp_078090', 'sc_Smp_055800',\n",
       "        'sc_Smp_212260', 'sc_Smp_063110', 'sc_Smp_332200', 'sc_Smp_154320',\n",
       "        'sc_Smp_248170', 'sc_Smp_080530', 'sc_Smp_321180', 'sc_Smp_051320',\n",
       "        'sc_Smp_129260', 'sc_Smp_211330', 'sc_Smp_004080', 'sc_Smp_160190',\n",
       "        'sc_Smp_108390', 'sc_Smp_167200', 'sc_Smp_341240', 'sc_Smp_050820',\n",
       "        'sc_Smp_145620', 'sc_Smp_031720', 'sc_Smp_245240', 'sc_Smp_068920',\n",
       "        'sc_Smp_157250', 'sc_Smp_057020', 'sc_Smp_340420', 'sc_Smp_337740',\n",
       "        'sc_Smp_098850', 'sc_Smp_341530', 'sc_Smp_169320', 'sc_Smp_103950',\n",
       "        'sc_Smp_077180', 'sc_Smp_247570', 'sc_Smp_051010', 'sc_Smp_310770',\n",
       "        'sc_Smp_342910', 'sc_Smp_066610', 'sc_Smp_175230', 'sc_Smp_064590',\n",
       "        'sc_Smp_167810', 'sc_Smp_124840', 'sc_Smp_028440', 'sc_Smp_021070',\n",
       "        'sc_Smp_061560', 'sc_Smp_127050', 'sc_Smp_068950', 'sc_Smp_094920',\n",
       "        'sc_Smp_104980', 'sc_Smp_169340', 'sc_Smp_172480', 'sc_Smp_337590',\n",
       "        'sc_Smp_175590', 'sc_Smp_131770', 'sc_Smp_127120', 'sc_Smp_332290',\n",
       "        'sc_Smp_050360', 'sc_Smp_104660', 'sc_Smp_179180', 'sc_Smp_004470',\n",
       "        'sc_Smp_151030', 'sc_Smp_075210', 'sc_Smp_133350', 'sc_Smp_003770',\n",
       "        'sc_Smp_300810', 'sc_Smp_337860', 'sc_Smp_131360', 'sc_Smp_026770',\n",
       "        'sc_Smp_060910', 'sc_Smp_079050', 'sc_Smp_074230', 'sc_Smp_087760',\n",
       "        'sc_Smp_333660', 'sc_Smp_015530', 'sc_Smp_317550', 'sc_Smp_136090',\n",
       "        'sc_Smp_016990', 'sc_Smp_035280', 'sc_Smp_055960', 'sc_Smp_178740',\n",
       "        'sc_Smp_066740', 'sc_Smp_045510', 'sc_Smp_158110', 'sc_Smp_163050',\n",
       "        'sc_Smp_163110', 'sc_Smp_267120', 'sc_Smp_321380', 'sc_Smp_020040',\n",
       "        'sc_Smp_157290', 'sc_Smp_027300', 'sc_Smp_336830', 'sc_Smp_012800',\n",
       "        'sc_Smp_053470', 'sc_Smp_025130', 'sc_Smp_017660', 'sc_Smp_161160',\n",
       "        'sc_Smp_150690', 'sc_Smp_038840', 'sc_Smp_145470', 'sc_Smp_096010',\n",
       "        'sc_Smp_330330', 'sc_Smp_030180', 'sc_Smp_087810', 'sc_Smp_098330',\n",
       "        'sc_Smp_078520', 'sc_Smp_331920', 'sc_Smp_055310', 'sc_Smp_044160',\n",
       "        'sc_Smp_169140', 'sc_Smp_340600', 'sc_Smp_179010', 'sc_Smp_103710',\n",
       "        'sc_Smp_006000', 'sc_Smp_005260', 'sc_Smp_074610', 'sc_Smp_148110',\n",
       "        'sc_Smp_008500', 'sc_Smp_176710', 'sc_Smp_032580', 'sc_Smp_140950',\n",
       "        'sc_Smp_078240', 'sc_Smp_209050', 'sc_Smp_036780', 'sc_Smp_084740',\n",
       "        'sc_Smp_044020', 'sc_Smp_175610', 'sc_Smp_130170', 'sc_Smp_148540',\n",
       "        'sc_Smp_005210', 'sc_Smp_101450', 'sc_Smp_331870', 'sc_Smp_129720',\n",
       "        'sc_Smp_343570', 'sc_Smp_163430', 'sc_Smp_046090', 'sc_Smp_158910',\n",
       "        'sc_Smp_335300', 'sc_Smp_047720', 'sc_Smp_042560', 'sc_Smp_069870',\n",
       "        'sc_Smp_141330', 'sc_Smp_092280', 'sc_Smp_099320', 'sc_Smp_155740',\n",
       "        'sc_Smp_010400', 'sc_Smp_303380', 'sc_Smp_079730', 'sc_Smp_031010',\n",
       "        'sc_Smp_093630', 'sc_Smp_069760', 'sc_Smp_173880', 'sc_Smp_319100',\n",
       "        'sc_Smp_073500', 'sc_Smp_006850', 'sc_Smp_168800', 'sc_Smp_207080',\n",
       "        'sc_Smp_002550', 'sc_Smp_171660', 'sc_Smp_080980', 'sc_Smp_303610',\n",
       "        'sc_Smp_054120', 'sc_Smp_213800', 'sc_Smp_136970', 'sc_Smp_328230',\n",
       "        'sc_Smp_062790', 'sc_Smp_083200', 'sc_Smp_063420', 'sc_Smp_132030',\n",
       "        'sc_Smp_034190', 'sc_Smp_012900', 'sc_Smp_133820', 'sc_Smp_072970',\n",
       "        'sc_Smp_179270', 'sc_Smp_001500', 'sc_Smp_013360', 'sc_Smp_054600',\n",
       "        'sc_Smp_045750', 'sc_Smp_180170', 'sc_Smp_072340', 'sc_Smp_077020',\n",
       "        'sc_Smp_305800', 'sc_Smp_001360', 'sc_Smp_179070', 'sc_Smp_087440',\n",
       "        'sc_Smp_003970', 'sc_Smp_010940', 'sc_Smp_304720', 'sc_Smp_076530',\n",
       "        'sc_Smp_174390', 'sc_Smp_118220', 'sc_Smp_076740', 'sc_Smp_106130',\n",
       "        'sc_Smp_074500', 'sc_Smp_180060', 'sc_Smp_064640', 'sc_Smp_197180',\n",
       "        'sc_Smp_165550', 'sc_Smp_043970', 'sc_Smp_246300', 'sc_Smp_347190',\n",
       "        'sc_Smp_006440', 'sc_Smp_053220', 'sc_Smp_064600', 'sc_Smp_020060',\n",
       "        'sc_Smp_332240', 'sc_Smp_070930', 'sc_Smp_247820', 'sc_Smp_064360',\n",
       "        'sc_Smp_083450', 'sc_Smp_212920', 'sc_Smp_085480', 'sc_Smp_136320',\n",
       "        'sc_Smp_312360', 'sc_Smp_048760', 'sc_Smp_022660', 'sc_Smp_067420',\n",
       "        'sc_Smp_095150', 'sc_Smp_330250', 'sc_Smp_051230', 'sc_Smp_172510',\n",
       "        'sc_Smp_340580', 'sc_Smp_076230', 'sc_Smp_002010', 'sc_Smp_073080',\n",
       "        'sc_Smp_050700', 'sc_Smp_007450', 'sc_Smp_099030', 'sc_Smp_042270',\n",
       "        'sc_Smp_078320', 'sc_Smp_029900', 'sc_Smp_011720', 'sc_Smp_073690',\n",
       "        'sc_Smp_098980', 'sc_Smp_073620', 'sc_Smp_093690', 'sc_Smp_309430',\n",
       "        'sc_Smp_002820', 'sc_Smp_097380', 'sc_Smp_072370', 'sc_Smp_320620',\n",
       "        'sc_Smp_070760', 'sc_Smp_073570', 'sc_Smp_247420', 'sc_Smp_063350',\n",
       "        'sc_Smp_015840', 'sc_Smp_193440', 'sc_Smp_077590', 'sc_Smp_164650',\n",
       "        'sc_Smp_068420', 'sc_Smp_063040', 'sc_Smp_210310', 'sc_Smp_067900',\n",
       "        'sc_Smp_092750', 'sc_Smp_067890', 'sc_Smp_129070', 'sc_Smp_181870',\n",
       "        'sc_Smp_008545', 'sc_Smp_006560', 'sc_Smp_092380', 'sc_Smp_090150',\n",
       "        'sc_Smp_046750', 'sc_Smp_113620', 'sc_Smp_004600', 'sc_Smp_200380',\n",
       "        'sc_Smp_067090', 'sc_Smp_166430', 'sc_Smp_151180', 'sc_Smp_186560',\n",
       "        'sc_Smp_213360', 'sc_Smp_000190', 'sc_Smp_012480', 'sc_Smp_084940',\n",
       "        'sc_Smp_194870', 'sc_Smp_093100', 'sc_Smp_241860', 'sc_Smp_030690',\n",
       "        'sc_Smp_051300', 'sc_Smp_070130', 'sc_Smp_210570', 'sc_Smp_094630',\n",
       "        'sc_Smp_056640', 'sc_Smp_025340', 'sc_Smp_050380', 'sc_Smp_019050',\n",
       "        'sc_Smp_153950', 'sc_Smp_105960', 'sc_Smp_205570', 'sc_Smp_342520',\n",
       "        'sc_Smp_336360', 'sc_Smp_133110', 'sc_Smp_157780', 'sc_Smp_004780',\n",
       "        'sc_Smp_006780', 'sc_Smp_048430', 'sc_Smp_340370', 'sc_Smp_079430',\n",
       "        'sc_Smp_088250', 'sc_Smp_058650', 'sc_Smp_091990', 'sc_Smp_328180',\n",
       "        'sc_Smp_045040', 'sc_Smp_307540', 'sc_Smp_058790', 'sc_Smp_019690',\n",
       "        'sc_Smp_078640', 'sc_Smp_076880', 'sc_Smp_053130', 'sc_Smp_083410',\n",
       "        'sc_Smp_082050', 'sc_Smp_004990', 'sc_Smp_326430', 'sc_Smp_066760',\n",
       "        'sc_Smp_329680', 'sc_Smp_132690', 'sc_Smp_162310', 'sc_Smp_020800',\n",
       "        'sc_Smp_036590', 'sc_Smp_064050', 'sc_Smp_078040', 'sc_Smp_024140',\n",
       "        'sc_Smp_045950', 'sc_Smp_307050', 'sc_Smp_157680', 'sc_Smp_315490',\n",
       "        'sc_Smp_094810', 'sc_Smp_210380', 'sc_Smp_320630', 'sc_Smp_102840',\n",
       "        'sc_Smp_150320', 'sc_Smp_130850', 'sc_Smp_083030', 'sc_Smp_213550',\n",
       "        'sc_Smp_071890', 'sc_Smp_303400', 'sc_Smp_145800', 'sc_Smp_072640',\n",
       "        'sc_Smp_019440', 'sc_Smp_053780', 'sc_Smp_069130', 'sc_Smp_103990',\n",
       "        'sc_Smp_017010', 'sc_Smp_037290', 'sc_Smp_050280', 'sc_Smp_062020',\n",
       "        'sc_Smp_032760', 'sc_Smp_097750', 'sc_Smp_024360', 'sc_Smp_024880',\n",
       "        'sc_Smp_099300', 'sc_Smp_138050', 'sc_Smp_046740', 'sc_Smp_267310',\n",
       "        'sc_Smp_347110', 'sc_Smp_246230', 'sc_Smp_050920', 'sc_Smp_341370',\n",
       "        'sc_Smp_243210', 'sc_Smp_033970', 'sc_Smp_201220', 'sc_Smp_093320',\n",
       "        'sc_Smp_141580', 'sc_Smp_064150', 'sc_Smp_001830', 'sc_Smp_335340',\n",
       "        'sc_Smp_038680', 'sc_Smp_051530', 'sc_Smp_089430', 'sc_Smp_096450',\n",
       "        'sc_Smp_309670', 'sc_Smp_121430', 'sc_Smp_176230', 'sc_Smp_154690',\n",
       "        'sc_Smp_163030', 'sc_Smp_015720', 'sc_Smp_046980', 'sc_Smp_007630',\n",
       "        'sc_Smp_017360', 'sc_Smp_307550', 'sc_Smp_172850', 'sc_Smp_102450',\n",
       "        'sc_Smp_064320', 'sc_Smp_034700', 'sc_Smp_002080', 'sc_Smp_307410',\n",
       "        'sc_Smp_026630', 'sc_Smp_019260', 'sc_Smp_335710', 'sc_Smp_002410',\n",
       "        'sc_Smp_314650', 'sc_Smp_102490', 'sc_Smp_042090', 'sc_Smp_318980',\n",
       "        'sc_Smp_340900', 'sc_Smp_204770', 'sc_Smp_070340', 'sc_Smp_105860',\n",
       "        'sc_Smp_176810', 'sc_Smp_163300', 'sc_Smp_037120', 'sc_Smp_311050',\n",
       "        'sc_Smp_015860', 'sc_Smp_074260', 'sc_Smp_006320', 'sc_Smp_043100',\n",
       "        'sc_Smp_181490', 'sc_Smp_016200', 'sc_Smp_060090', 'sc_Smp_061200',\n",
       "        'sc_Smp_016130', 'sc_Smp_060180', 'sc_Smp_144180', 'sc_Smp_090920'],\n",
       "       dtype='<U13'))"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Gp,G1,G2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get a table of enriched gene pairs from all cell type mappings that have an alignment score above some threshold (`thr`), you can use:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating gene pairs for the mapping: pl;Cathepsin+ cells: 10 to sc;Cathepsin\n",
      "Calculating gene pairs for the mapping: pl;Cathepsin+ cells: 28 to sc;Cathepsin\n",
      "Calculating gene pairs for the mapping: pl;Cathepsin+ cells: 4 to sc;Cathepsin\n",
      "Calculating gene pairs for the mapping: pl;Epidermal: 11 to sc;Tegument\n",
      "Calculating gene pairs for the mapping: pl;Epidermal: 2 to sc;Tegument_prog\n",
      "Calculating gene pairs for the mapping: pl;Epidermal: 24 to sc;Tegument\n",
      "Calculating gene pairs for the mapping: pl;Epidermal: 3 to sc;Tegument_prog\n",
      "Calculating gene pairs for the mapping: pl;Epidermal: 35 to sc;Tegument\n",
      "Calculating gene pairs for the mapping: pl;Intestine: 30 to sc;Intestine\n",
      "Calculating gene pairs for the mapping: pl;Intestine: 6 to sc;Tegument_prog\n",
      "Calculating gene pairs for the mapping: pl;Muscle: 13 to sc;Muscle\n",
      "Calculating gene pairs for the mapping: pl;Muscle: 14 to sc;Muscle\n",
      "Calculating gene pairs for the mapping: pl;Muscle: 16 to sc;Muscle\n"
     ]
    }
   ],
   "source": [
    "gene_pairs = gpf.find_all(thr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saving/Loading SAMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from samap.utils import save_samap, load_samap\n",
    "# save to a .pkl file\n",
    "save_samap(sm,'path/to/file') #including the file ending (.pkl) is optional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load\n",
    "sm = load_samap('path/to/file') #including the file ending (.pkl) is optional"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualizing SAMap results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launching an interactive GUI (requires SAMGUI, see the README instructions in the [SAM](https://github.com/atarashansky/self-assembling-manifold) github repo):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sm.gui()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create a sankey plot, use `samap.analysis.sankey_plot`. This requires `holoviews` to be installed (`pip install holoviews`). It should already be installed if you're using the Docker image."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from samap.analysis import sankey_plot\n",
    "k1 = 'cluster' #cell types annotation key in `sam1.adata.obs`\n",
    "k2 = 'tissue' #cell types annotation key in `sam2.adata.obs`\n",
    "sankey_plot(sm,k1,k2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
