{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file1 = 'path/to/transcriptome1.fasta'\n",
    "type1 = 'nucl' #or 'prot' if file1 is a proteome\n",
    "id1 = 'hu' #2-character ID (e.g. 'hu' for human)\n",
    "\n",
    "file2 = 'path/to/transcriptome2.fasta'\n",
    "type2 = 'nucl'\n",
    "id2 = 'mo' #2-character ID (e.g. 'mo' for mouse)\n",
    "!bash map_genes.sh --tr1 {file1} --t1 {type1} --n1 {id1} --tr2 {file2} --t2 {type2} --n2 {id2}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run SAMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from samap.mapping import SAMAP\n",
    "from samap.analysis import get_mapping_scores, GenePairFinder\n",
    "from samalg import SAM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SAMap accepts file paths to unprocessed, raw `.h5ad` files. Alternatively, if you already have a processed `SAM` object, you can load them in directly.\n",
    "\n",
    "\n",
    "Prior to running SAMap, you should have run `map_genes.sh`, which expects 2-character identifiers describing each species. For example, human and mouse might get `hu` and `mo` identifiers, respectively. `map_genes.sh` generates a `maps/` directory with the transcriptome mapping BLAST results deposited. The input species IDs and path to the `maps/` directory should be input into SAMap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id1 = 'hu'\n",
    "id2 = 'mo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# passing in file names (SAMap will process the data with SAM and save the resulting objects to two `.h5ad` files.)\n",
    "fn1 = '/path/to/file1.h5ad' #processed data will be automatically saved to `/path/to/file/file1_pr.h5ad`\n",
    "fn2 = '/path/to/file2.h5ad' #processed data will be automatically saved to `/path/to/file/file2_pr.h5ad`\n",
    "# runs SAMAP \n",
    "sm = SAMAP(fn1,fn2,id1,id2,f_maps = 'maps/')\n",
    "samap = sm.run()\n",
    "\n",
    "\"\"\"\n",
    "# passing in already-processed SAM objects\n",
    "sam1=SAM()\n",
    "sam2=SAM()\n",
    "sam1.load_data('/path/to/file1_pr.h5ad')\n",
    "sam2.load_data('/path/to/file2_pr.h5ad')\n",
    "\n",
    "sm = SAMAP(sam1,sam2,id1,id2,f_maps = 'maps/')\n",
    "samap = sm.run()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To calculate alignment scores between cell types, we can use `get_mapping_scores`. This function will use the combined SAM object produced by SAMap to calculate alignment scores between cell types in the provided cell type annotation columns of `sam.adata.obs`. If no cell type annotations exist, the leiden clusters generated by SAM can be used (`k1=k2='leiden_clusters'`).\n",
    "\n",
    "The resulting tables show the highest-scoring alignment scores for each cell type in organism 1 (`D1`) and organism 2 (`D2`), respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "k1 = 'cell_types' #cell types annotation key in `sam1.adata.obs`\n",
    "k2 = 'cell_types' #cell types annotation key in `sam2.adata.obs`\n",
    "D1,D2 = get_mapping_scores(sam1,sam2,samap,k1,k2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SAMap provides a class to find gene pairs enriched in different cell type pairs. The method entails finding gene pairs that contribute positively to the cross-species correlation between cell types and are differentially expressed in their respective mapped cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gpf = GenePairFinder(sam1,sam2,samap,k1=k1,k2=k2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`gpf.find_genes` can now be used to find gene pairs enriched in a cell type mapping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n1 = 'cell_type1' #cell type ID from organism 1 (must be present in `sam1.adata.obs[k1]`)\n",
    "n2 = 'cell_type2' #cell type ID from organism 2 (must be present in `sam2.adata.obs[k1]`)\n",
    "Gp,G1,G2 = gpf.find_genes(n1,n2)\n",
    "#Gp are the gene pairs, G1 are the genes from organism 1, G2 are the genes from organism 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saving/Loading SAMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from samap.utils import save_samap, load_samap\n",
    "# save to a .pkl file\n",
    "save_samap(sm,'path/to/file') #no need to include the file ending (.pkl)\n",
    "# load\n",
    "sm = load_samap('path/to/file') #no need to include the file ending (.pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualizing SAMap results\n",
    "\n",
    "Displaying a heatmap:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sm.display_heatmap(key1='leiden_clusters',key2='leiden_clusters')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launching an interactive GUI (requires SAMGUI, see the README instructions in the [SAM](https://github.com/atarashansky/self-assembling-manifold) github repo):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sm.gui()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
